# scikit-bio-tutorials

Interactive tutorials for using scikit-bio in biological research.

This repository hosts tutorial materials for the ISMB 2024 conference.

- https://www.iscb.org/ismb2024/programme-schedule/tutorials#ip3

The tutorial is broken down as follows, Google Colab notebooks can be launched via the following links

1. [Section 01: Basic bioinformatics using scikit-bio](https://colab.research.google.com/github/scikit-bio/scikit-bio-tutorials/blob/main/01-basic-bioinfo/01-basic-bioinfo.ipynb)
2. [Section 02: Embeddings and Vectors with Scikit-Bio](https://colab.research.google.com/github/scikit-bio/scikit-bio-tutorials/blob/main/02-language-model/02-language-model.ipynb)
3. [Section 03: Working with various omic data types](https://colab.research.google.com/github/scikit-bio/scikit-bio-tutorials/blob/main/03-omic-data-types/03-omic-data-types.ipynb)
4. [Section 04: Analyzing microbial communities](https://colab.research.google.com/github/scikit-bio/scikit-bio-tutorials/blob/main/04-community-structure/04-community-structure.ipynb)
5. [Section 05: Comparing microbial community structure](https://colab.research.google.com/github/scikit-bio/scikit-bio-tutorials/blob/main/05-community-diversity/05-community-diversity.ipynb)
6. [Section 06: Inferring and associating critical features](https://colab.research.google.com/github/scikit-bio/scikit-bio-tutorials/blob/main/06-marker-inference/06-marker-inference.ipynb)
7. [Section 07: Predicting host and environmental traits](https://colab.research.google.com/github/scikit-bio/scikit-bio-tutorials/blob/main/07-outcome-prediction/07-outcome-prediction.ipynb)

The example dataset(s) are available for download at Dropbox:

- https://www.dropbox.com/scl/fo/euir18o3jb4pglljaxslo/AKNF1Lk3sQ52ZTn_A-z2sqc?rlkey=u72814al272kvu17rnozsr5jp&dl=0

Alternatively, the tutorial materials (notebooks, slides, and datasets) are shared at Google Drive:

- https://drive.google.com/drive/folders/1YSpEBHrGM_fjmiN5NpwdgLL3WtT0VTsy?usp=sharing


# References
1. [Protein remote homology detection and structural alignment using deep learning](https://www.nature.com/articles/s41587-023-01917-2)
2. [ProtTrans: Towards Cracking the Language of Lifeâ€™s Code Through Self-Supervised Learning](https://www.biorxiv.org/content/10.1101/2020.07.12.199554v3)
3. [UniFrac: a New Phylogenetic Method for Comparing Microbial Communities ](https://journals.asm.org/doi/10.1128/AEM.71.12.8228-8235.2005)
4. [Establishing microbial composition measurement standards with reference frames](https://www.nature.com/articles/s41467-019-10656-5)
5. [Unifying the analysis of high-throughput sequencing datasets: characterizing RNA-seq, 16S rRNA gene sequencing and selective growth experiments by compositional data analysis](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4030730/)
6. [Balances: a New Perspective for Microbiome Analysis](https://pubmed.ncbi.nlm.nih.gov/30035234/)
7. [Compositionally Aware Phylogenetic Beta-Diversity Measures Better Resolve Microbiomes Associated with Phenotype](https://journals.asm.org/doi/10.1128/msystems.00050-22)
