# scikit-bio-tutorials

Interactive tutorials for using [scikit-bio](https://scikit.bio/) in biological research.

This repository hosts tutorial materials that were initially taught during the [ISMB 2024](https://www.iscb.org/ismb2024/programme-schedule/tutorials#ip3) conference. They were latter updated and expanded to reflect the new features of scikit-bio. The current tutorials are up-to-date with scikit-bio [0.7.0](https://github.com/scikit-bio/scikit-bio/releases/tag/0.7.0).


## Walk-through tutorials

The tutorials are broken down into eight sections. They can be directly launched in Google Colab via the following links.

### Basic bioinformatics

1. [Basic bioinformatics using scikit-bio](https://colab.research.google.com/github/scikit-bio/scikit-bio-tutorials/blob/main/01-basic-bioinfo/01-basic-bioinfo.ipynb)

### Microbiome data analysis

2. [Working with various omic data types](https://colab.research.google.com/github/scikit-bio/scikit-bio-tutorials/blob/main/02-multi-omic-data/02-multi-omic-data.ipynb)
3. [Analyzing microbial communities](https://colab.research.google.com/github/scikit-bio/scikit-bio-tutorials/blob/main/03-community-structure/03-community-structure.ipynb)
4. [Comparing microbial community structure](https://colab.research.google.com/github/scikit-bio/scikit-bio-tutorials/blob/main/04-community-diversity/04-community-diversity.ipynb)
5. [Inferring and associating critical features](https://colab.research.google.com/github/scikit-bio/scikit-bio-tutorials/blob/main/05-marker-inference/05-marker-inference.ipynb)
6. [Predicting host and environmental traits](https://colab.research.google.com/github/scikit-bio/scikit-bio-tutorials/blob/main/06-outcome-prediction/06-outcome-prediction.ipynb)

These sections use the [EMP500 dataset](https://www.nature.com/articles/s41564-022-01266-x), available for download at this [Dropbox link](https://www.dropbox.com/scl/fi/7sepykr3vhss60647osc8/emp500.tar.gz?rlkey=6q5ac48noeax5n2htaqslvo8g&dl=0).

### Advanced topics

7. [Single-cell data analysis](https://colab.research.google.com/github/scikit-bio/scikit-bio-tutorials/blob/main/07-single-cell-omics/07-single-cell-omics.ipynb)
8. [Protein language modeling](https://colab.research.google.com/github/scikit-bio/scikit-bio-tutorials/blob/main/08-language-model/08-language-model.ipynb)


## References

1. [Protein remote homology detection and structural alignment using deep learning](https://www.nature.com/articles/s41587-023-01917-2)
2. [ProtTrans: Towards Cracking the Language of Lifeâ€™s Code Through Self-Supervised Learning](https://www.biorxiv.org/content/10.1101/2020.07.12.199554v3)
3. [UniFrac: a New Phylogenetic Method for Comparing Microbial Communities ](https://journals.asm.org/doi/10.1128/AEM.71.12.8228-8235.2005)
4. [Establishing microbial composition measurement standards with reference frames](https://www.nature.com/articles/s41467-019-10656-5)
5. [Unifying the analysis of high-throughput sequencing datasets: characterizing RNA-seq, 16S rRNA gene sequencing and selective growth experiments by compositional data analysis](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4030730/)
6. [Balances: a New Perspective for Microbiome Analysis](https://pubmed.ncbi.nlm.nih.gov/30035234/)
7. [Compositionally Aware Phylogenetic Beta-Diversity Measures Better Resolve Microbiomes Associated with Phenotype](https://journals.asm.org/doi/10.1128/msystems.00050-22)
