{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/atharvved09/examples-scikit-bio/blob/main/02-multi-omic-data/02-multi-omic-data.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "648eb8f4-afb3-4aac-a7f9-1f30dad840cf",
      "metadata": {
        "id": "648eb8f4-afb3-4aac-a7f9-1f30dad840cf"
      },
      "source": [
        "# Section 02: Working with various omic data types\n",
        "\n",
        "- Author: Qiyun Zhu\n",
        "\n",
        "---"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "b1a093af-8a55-4e54-bbc0-e1df08254f84",
      "metadata": {
        "id": "b1a093af-8a55-4e54-bbc0-e1df08254f84"
      },
      "source": [
        "[**Omics**](https://en.wikipedia.org/wiki/Omics) refer to the studies of the totality of biological entities within a system. They usually employ high-throughput techniques, such as DNA sequencing and mass spectrometry, to generate information about all molecules, genes, cells, organisms, and more. This information can be analyzed using bioinformatics methods to inform the structure, function and dynamics of the biological entities.\n",
        "\n",
        "In microbiome research, because isolating and cultivating all microbes in a habitat is usually infeasible, omics are essential tools to characterize the microbiome (the name of which already indicates this nature). The prefix \"**meta**-\" is usually added to an omic to indicate the study of biological entities from all microorganisms within the environment (e.g., [metagenomics](https://en.wikipedia.org/wiki/Metagenomics)). Omics for microbiota can be employed in combination with omics for their host or the environment to profile a more comprehensive image of the biological system.\n",
        "\n",
        "Depending on the biological molecules it studies, the most common omics for host and microbiota include:\n",
        "\n",
        "- Molecule: DNA - mRNA - protein - metabolite\n",
        "- (Host) ome: genome - transcriptome - proteome - metabolome\n",
        "- (Microbiota) ome: metagenome - metatranscriptome - metaproteome - (meta)metabolome\n",
        "\n",
        "In this section, we will use a real-world study -- the **Earth Microbiome Project 500** ([EMP500](https://earthmicrobiome.org/emp500/)), as an example to explore various omic data types. We will navigate the data tables and relevant data files associated with individual omics, discuss the unique challenges and solutions in working with omic data, and compare between omic data types. This process will support the analyses in all subsequent sections."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "495a8b05-1956-4bf4-b376-927e93aa8c7d",
      "metadata": {
        "id": "495a8b05-1956-4bf4-b376-927e93aa8c7d"
      },
      "source": [
        "## Preparation"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "5108ecf9-a855-47c9-8db2-4d2bc07727c1",
      "metadata": {
        "id": "5108ecf9-a855-47c9-8db2-4d2bc07727c1"
      },
      "source": [
        "Install the latest version of scikit-bio if it hasn't been (needed for every Google Colab instance)."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "id": "317563d7-f081-4673-a233-62c0c7480d00",
      "metadata": {
        "id": "317563d7-f081-4673-a233-62c0c7480d00"
      },
      "outputs": [],
      "source": [
        "from importlib.util import find_spec"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "id": "62e25faa-b274-4472-a09b-c7267255d71a",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "62e25faa-b274-4472-a09b-c7267255d71a",
        "outputId": "9a9474e7-2b2c-497b-f68e-4960e8c4135b"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m10.9/10.9 MB\u001b[0m \u001b[31m47.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m58.6/58.6 kB\u001b[0m \u001b[31m2.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m12.3/12.3 MB\u001b[0m \u001b[31m48.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h"
          ]
        }
      ],
      "source": [
        "if find_spec('skbio') is None:\n",
        "    !pip install -q scikit-bio"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "id": "03bcda18-b5f0-414f-b201-db5526a82169",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 35
        },
        "id": "03bcda18-b5f0-414f-b201-db5526a82169",
        "outputId": "75a1c6c3-acaa-4699-a606-6b1142b3fcad"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'0.7.1.post1'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 3
        }
      ],
      "source": [
        "import skbio\n",
        "skbio.__version__"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "6bf93f88-a626-4e65-849f-6ca362e918ec",
      "metadata": {
        "id": "6bf93f88-a626-4e65-849f-6ca362e918ec"
      },
      "source": [
        "Import common libraries."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "id": "9673877e-af2d-46cb-a984-80cead29cec2",
      "metadata": {
        "id": "9673877e-af2d-46cb-a984-80cead29cec2"
      },
      "outputs": [],
      "source": [
        "import numpy as np\n",
        "import pandas as pd\n",
        "import matplotlib.pyplot as plt"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "cd3ff408-b8a2-44df-b73d-59581ec1937b",
      "metadata": {
        "id": "cd3ff408-b8a2-44df-b73d-59581ec1937b"
      },
      "source": [
        "Download the EMP500 dataset and unzip it. (If you have done it already, skip the following code.)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "id": "07cab6c7-a6e6-4648-ac7a-5e5e70d4122f",
      "metadata": {
        "id": "07cab6c7-a6e6-4648-ac7a-5e5e70d4122f"
      },
      "outputs": [],
      "source": [
        "!wget -q -O emp500.tar.gz \"https://www.dropbox.com/scl/fi/7sepykr3vhss60647osc8/emp500.tar.gz?rlkey=6q5ac48noeax5n2htaqslvo8g&dl=0\"\n",
        "!tar zxf emp500.tar.gz"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "id": "09097a5b-8225-4641-9193-2a1278dd312c",
      "metadata": {
        "id": "09097a5b-8225-4641-9193-2a1278dd312c"
      },
      "outputs": [],
      "source": [
        "HOME = 'emp500'"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "f9c3cb6d-9b71-42fe-a47d-a1cc91d66940",
      "metadata": {
        "id": "f9c3cb6d-9b71-42fe-a47d-a1cc91d66940"
      },
      "source": [
        "Check if the directory exists by listing its content."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "id": "0ba11d77-fef6-4701-bf30-7062fe32c803",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0ba11d77-fef6-4701-bf30-7062fe32c803",
        "outputId": "de9b810a-cbad-42cf-c3f9-24192d786812"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "amplicon  assembly  masspec  README.md\tsample.tsv  shotgun\n"
          ]
        }
      ],
      "source": [
        "!ls $HOME"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "a6168df3-ae6b-4690-b43f-7401243bdb9f",
      "metadata": {
        "id": "a6168df3-ae6b-4690-b43f-7401243bdb9f"
      },
      "source": [
        "## The EMP500 study"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "4416cd76-cedd-4d3c-a9a6-6b1811fe6824",
      "metadata": {
        "id": "4416cd76-cedd-4d3c-a9a6-6b1811fe6824"
      },
      "source": [
        "A `README.md` file within the data directory provides basic information about the study. Take a quick look at it:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "id": "107a9170-6842-4bb1-9abc-130bf336e638",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "107a9170-6842-4bb1-9abc-130bf336e638",
        "outputId": "466e1baa-d332-4239-96f6-764900fefe0f"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "# The EMP500 Dataset\n",
            "\n",
            "The Earth Microbiome Project 500 (EMP500) was a multi-omic study of microbial\n",
            "communities sampled from various biomes and habitats of the planet. The study\n",
            "was published in:\n",
            "\n",
            "- Shaffer, J. P., Nothias, L. F., Thompson, L. R., Sanders, J. G., Salido, R.\n",
            "  A., Couvillion, S. P., ... & Knight, R. (2022). Standardized multi-omics of\n",
            "  Earth’s microbiomes reveals microbial and metabolite diversity. Nature\n",
            "  Microbiology, 7(12), 2128-2150.\n"
          ]
        }
      ],
      "source": [
        "!head $HOME/README.md"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "c05d2da3-9af0-4338-9f0b-f81dcdd97200",
      "metadata": {
        "id": "c05d2da3-9af0-4338-9f0b-f81dcdd97200"
      },
      "source": [
        "The EMP500 project was described in the following paper:\n",
        "\n",
        "- https://www.nature.com/articles/s41564-022-01266-x\n",
        "\n",
        "In this work, the researchers performed multi-omic analyses of **880** samples collected from various free-living and host-associated environments across the Earth. They conducted amplicon sequencing (16S, 18S, ITS), shotgun metagenomics (and subsequent assembly), and untargeted metabolomics (LC-MS and GC-MS). Through the integration of these omic data types and sample traits, they identified the generic and environment-specific associations between microbes and metabolites."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "a8b170e7-5977-4bf2-81e8-2063c3c46277",
      "metadata": {
        "id": "a8b170e7-5977-4bf2-81e8-2063c3c46277"
      },
      "source": [
        "![image.png](attachment:cc8e1e8a-f016-4856-8b46-097f6bea875c.png)\n",
        "\n",
        "_**Fig. 1**: The 880 EMP500 samples: environmental classification and geographical locations. Linked from Fig. 1 of [Shaffer et al. (2022)](https://www.nature.com/articles/s41564-022-01266-x)._"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "cc96737b-1831-4c91-adc4-70188c1a2119",
      "metadata": {
        "id": "cc96737b-1831-4c91-adc4-70188c1a2119"
      },
      "source": [
        "## Sample metadata"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "ef46d044-5087-4154-a4dc-88e83100e790",
      "metadata": {
        "id": "ef46d044-5087-4154-a4dc-88e83100e790"
      },
      "source": [
        "[**Metadata**](https://en.wikipedia.org/wiki/Metadata) provides information about other data. In an omic study, the properties of the biological samples, specimens, subjects, etc. are usually provided by the **sample metadata**. Examples including demographics, morphology, behaviors, clinical measurements, etc. Sample metadata is the most essential component of an omic study, and should be treated in priority over any omic data. Without metadata, one is left with numbers but without any biological question to address.\n",
        "\n",
        "- Meanwhile, there are **feature metadata** -- the information about biological entities studied in each omic. They will be discussed below."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "62f5361b-367c-4f79-8658-560830af45c2",
      "metadata": {
        "id": "62f5361b-367c-4f79-8658-560830af45c2"
      },
      "source": [
        "### Metadata table"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "024f45a1-bd4f-42d0-98b7-852fb3f61537",
      "metadata": {
        "id": "024f45a1-bd4f-42d0-98b7-852fb3f61537"
      },
      "source": [
        "The sample metadata table, `sample.tsv`, is a tab-separated values ([TSV](https://en.wikipedia.org/wiki/Tab-separated_values)) file with samples as rows and properties as columns. Let's take a peek at the table:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "id": "57e769a7-1d13-4ace-9a66-3868e3265ff1",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "57e769a7-1d13-4ace-9a66-3868e3265ff1",
        "outputId": "9c0c0b92-7001-4937-ef30-bc7aeb131d3f"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "sample_name\tsample_name_original\tsample_type\tcollection_timestamp\tcountry\tdepth_sample\tdescription\te\n",
            "13114.angenent.65.s001\tAngenent65.misc.104\tbioreactor sludge\t08/09/2011 0:00\tUSA - New York\tnot appl\n",
            "13114.angenent.65.s002\tAngenent65.misc.519\tbioreactor sludge\t9/27/12 0:00\tUSA - New York\tnot applica\n",
            "13114.angenent.65.s003\tAngenent65.misc.946\tbioreactor sludge\t11/28/13 0:00\tUSA - New York\tnot applic\n",
            "13114.angenent.65.s004\tAngenent65.misc.1022\tbioreactor sludge\t02/12/2014 0:00\tUSA - New York\tnot app\n",
            "13114.angenent.65.s005\tAngenent65.misc.1538\tbioreactor sludge\t7/13/15 0:00\tUSA - New York\tnot applic\n",
            "13114.angenent.65.s006\tAngenent65.misc.1622\tbioreactor sludge\t10/05/2015 0:00\tUSA - New York\tnot app\n",
            "13114.angenent.65.s007\tAngenent65.misc.1722\tbioreactor sludge\t1/13/15 0:00\tUSA - New York\tnot applic\n",
            "13114.angenent.65.s008\tAngenent65.misc.1818\tbioreactor sludge\t4/18/16 0:00\tUSA - New York\tnot applic\n",
            "13114.angenent.65.s009\tAngenent65.misc.1888\tbioreactor sludge\t6/27/16 0:00\tUSA - New York\tnot applic\n"
          ]
        }
      ],
      "source": [
        "!head $HOME/sample.tsv | cut -c1-100"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "ac5469fd-e4c2-422f-897a-a838235bce9f",
      "metadata": {
        "id": "ac5469fd-e4c2-422f-897a-a838235bce9f"
      },
      "source": [
        "This table can be loaded using Pandas, and scikit-bio fully supports the analysis of Pandas dataframes."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "id": "48cb0340-c95d-4061-a85b-779164c40ffb",
      "metadata": {
        "id": "48cb0340-c95d-4061-a85b-779164c40ffb"
      },
      "outputs": [],
      "source": [
        "# meta = pd.read_table(f'{HOME}/sample.tsv', index_col=0)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "a320fe0d-f4e4-406d-ab77-b581d06af92b",
      "metadata": {
        "id": "a320fe0d-f4e4-406d-ab77-b581d06af92b"
      },
      "source": [
        "Meanwhile, scikit-bio provides a specific class: [`SampleMetadata`](https://scikit.bio/docs/latest/generated/skbio.metadata.SampleMetadata.html), for handling sample metadata. This class was ported from [QIIME 2](https://qiime2.org/), a widely adopted platform for microbiome data analysis. If you are already familiar with QIIME 2, you may find this class familiar as well. Our team is working to expand its functionality beyond what was already offered."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "id": "770bbbfe-096d-449b-9cea-d463e0dc6c5c",
      "metadata": {
        "id": "770bbbfe-096d-449b-9cea-d463e0dc6c5c"
      },
      "outputs": [],
      "source": [
        "from skbio.metadata import SampleMetadata"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "35c9d3bc-2891-47cb-80cf-0ca74919d1c9",
      "metadata": {
        "id": "35c9d3bc-2891-47cb-80cf-0ca74919d1c9"
      },
      "source": [
        "An advantage of `SampleMetadata` is that it implements pre-defined rules for parsing biological metadata. For example, [missing data](https://en.wikipedia.org/wiki/Missing_data) are prevalent in biological research. They can occur for multiple reasons, which should be noted rather than leaving the cells empty. The International Nucleotide Database Collaboration (INSDC, https://www.insdc.org/) defined terms representing different types of missing data, such as \"not applicable\", \"not collected\", \"not provided\" and \"restricted access\" (linked below).\n",
        "\n",
        "- https://www.insdc.org/submitting-standards/missing-value-reporting/"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "id": "0746cfcf-834d-4eee-a99b-b88fb7d5c540",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0746cfcf-834d-4eee-a99b-b88fb7d5c540",
        "outputId": "b865974d-56d4-46a0-af67-6e04ee1a35b9"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "SampleMetadata\n",
              "--------------\n",
              "880 IDs x 33 columns\n",
              "sample_name_original:          ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "sample_type:                   ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "collection_timestamp:          ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "country:                       ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "depth_sample:                  ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "description:                   ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "elevation:                     ColumnProperties(type='numeric', missing_scheme='INSDC:missing')\n",
              "emp500_principal_investigator: ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "emp500_study_id:               ColumnProperties(type='numeric', missing_scheme='INSDC:missing')\n",
              "emp500_pi_study_id:            ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "emp500_title:                  ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "empo_0:                        ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "empo_1:                        ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "empo_2:                        ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "empo_3:                        ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "empo_salinity:                 ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "empo_v2_2:                     ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "empo_v2_3a:                    ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "empo_v2_3b:                    ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "empo_v2_4a:                    ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "empo_v2_4b:                    ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "empo_4:                        ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "env_biome:                     ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "env_feature:                   ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "env_material:                  ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "env_package:                   ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "geo_loc_name:                  ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "host_subject_id:               ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "host_scientific_name:          ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "latitude:                      ColumnProperties(type='numeric', missing_scheme='INSDC:missing')\n",
              "longitude:                     ColumnProperties(type='numeric', missing_scheme='INSDC:missing')\n",
              "project_name:                  ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "scientific_name:               ColumnProperties(type='categorical', missing_scheme='INSDC:missing')\n",
              "\n",
              "Call to_dataframe() for a tabular representation."
            ]
          },
          "metadata": {},
          "execution_count": 12
        }
      ],
      "source": [
        "meta = SampleMetadata.load(f'{HOME}/sample.tsv', default_missing_scheme='INSDC:missing')\n",
        "meta"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "eadb65d3-b900-4495-a38c-e68c68f38aee",
      "metadata": {
        "id": "eadb65d3-b900-4495-a38c-e68c68f38aee"
      },
      "source": [
        "A `SampleMetadata` object can be converted into a Pandas dataframe with:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "id": "36a4d4d0-baeb-44ef-8e92-3871194c6d10",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 380
        },
        "id": "36a4d4d0-baeb-44ef-8e92-3871194c6d10",
        "outputId": "cd810528-4244-48a1-94e1-97909f0e7f5d"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                       sample_name_original        sample_type  \\\n",
              "sample_name                                                      \n",
              "13114.angenent.65.s001  Angenent65.misc.104  bioreactor sludge   \n",
              "13114.angenent.65.s002  Angenent65.misc.519  bioreactor sludge   \n",
              "13114.angenent.65.s003  Angenent65.misc.946  bioreactor sludge   \n",
              "\n",
              "                       collection_timestamp         country depth_sample  \\\n",
              "sample_name                                                                \n",
              "13114.angenent.65.s001      08/09/2011 0:00  USA - New York          NaN   \n",
              "13114.angenent.65.s002         9/27/12 0:00  USA - New York          NaN   \n",
              "13114.angenent.65.s003        11/28/13 0:00  USA - New York          NaN   \n",
              "\n",
              "                                                      description  elevation  \\\n",
              "sample_name                                                                    \n",
              "13114.angenent.65.s001  Corn Beer Fermentation Bioreactor Biomass      123.0   \n",
              "13114.angenent.65.s002  Corn Beer Fermentation Bioreactor Biomass      123.0   \n",
              "13114.angenent.65.s003  Corn Beer Fermentation Bioreactor Biomass      123.0   \n",
              "\n",
              "                       emp500_principal_investigator  emp500_study_id  \\\n",
              "sample_name                                                             \n",
              "13114.angenent.65.s001                      Angenent             65.0   \n",
              "13114.angenent.65.s002                      Angenent             65.0   \n",
              "13114.angenent.65.s003                      Angenent             65.0   \n",
              "\n",
              "                       emp500_pi_study_id  ...           env_feature  \\\n",
              "sample_name                                ...                         \n",
              "13114.angenent.65.s001         Angenent65  ...  anaerobic bioreactor   \n",
              "13114.angenent.65.s002         Angenent65  ...  anaerobic bioreactor   \n",
              "13114.angenent.65.s003         Angenent65  ...  anaerobic bioreactor   \n",
              "\n",
              "                            env_material      env_package  \\\n",
              "sample_name                                                 \n",
              "13114.angenent.65.s001  anaerobic sludge  host-associated   \n",
              "13114.angenent.65.s002  anaerobic sludge  host-associated   \n",
              "13114.angenent.65.s003  anaerobic sludge  host-associated   \n",
              "\n",
              "                                                             geo_loc_name  \\\n",
              "sample_name                                                                 \n",
              "13114.angenent.65.s001  United States of America:State of New York:Cit...   \n",
              "13114.angenent.65.s002  United States of America:State of New York:Cit...   \n",
              "13114.angenent.65.s003  United States of America:State of New York:Cit...   \n",
              "\n",
              "                       host_subject_id host_scientific_name latitude  \\\n",
              "sample_name                                                            \n",
              "13114.angenent.65.s001    LTA.misc.104                  NaN    42.44   \n",
              "13114.angenent.65.s002    LTA.misc.519                  NaN    42.44   \n",
              "13114.angenent.65.s003    LTA.misc.104                  NaN    42.44   \n",
              "\n",
              "                       longitude                       project_name  \\\n",
              "sample_name                                                           \n",
              "13114.angenent.65.s001     -76.5  Corn Beer Fermentation Bioreactor   \n",
              "13114.angenent.65.s002     -76.5  Corn Beer Fermentation Bioreactor   \n",
              "13114.angenent.65.s003     -76.5  Corn Beer Fermentation Bioreactor   \n",
              "\n",
              "                              scientific_name  \n",
              "sample_name                                    \n",
              "13114.angenent.65.s001  bioreactor metagenome  \n",
              "13114.angenent.65.s002  bioreactor metagenome  \n",
              "13114.angenent.65.s003  bioreactor metagenome  \n",
              "\n",
              "[3 rows x 33 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-3654836c-4b26-4e9d-bcb8-dd413532dbbf\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>sample_name_original</th>\n",
              "      <th>sample_type</th>\n",
              "      <th>collection_timestamp</th>\n",
              "      <th>country</th>\n",
              "      <th>depth_sample</th>\n",
              "      <th>description</th>\n",
              "      <th>elevation</th>\n",
              "      <th>emp500_principal_investigator</th>\n",
              "      <th>emp500_study_id</th>\n",
              "      <th>emp500_pi_study_id</th>\n",
              "      <th>...</th>\n",
              "      <th>env_feature</th>\n",
              "      <th>env_material</th>\n",
              "      <th>env_package</th>\n",
              "      <th>geo_loc_name</th>\n",
              "      <th>host_subject_id</th>\n",
              "      <th>host_scientific_name</th>\n",
              "      <th>latitude</th>\n",
              "      <th>longitude</th>\n",
              "      <th>project_name</th>\n",
              "      <th>scientific_name</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>sample_name</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>13114.angenent.65.s001</th>\n",
              "      <td>Angenent65.misc.104</td>\n",
              "      <td>bioreactor sludge</td>\n",
              "      <td>08/09/2011 0:00</td>\n",
              "      <td>USA - New York</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Corn Beer Fermentation Bioreactor Biomass</td>\n",
              "      <td>123.0</td>\n",
              "      <td>Angenent</td>\n",
              "      <td>65.0</td>\n",
              "      <td>Angenent65</td>\n",
              "      <td>...</td>\n",
              "      <td>anaerobic bioreactor</td>\n",
              "      <td>anaerobic sludge</td>\n",
              "      <td>host-associated</td>\n",
              "      <td>United States of America:State of New York:Cit...</td>\n",
              "      <td>LTA.misc.104</td>\n",
              "      <td>NaN</td>\n",
              "      <td>42.44</td>\n",
              "      <td>-76.5</td>\n",
              "      <td>Corn Beer Fermentation Bioreactor</td>\n",
              "      <td>bioreactor metagenome</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13114.angenent.65.s002</th>\n",
              "      <td>Angenent65.misc.519</td>\n",
              "      <td>bioreactor sludge</td>\n",
              "      <td>9/27/12 0:00</td>\n",
              "      <td>USA - New York</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Corn Beer Fermentation Bioreactor Biomass</td>\n",
              "      <td>123.0</td>\n",
              "      <td>Angenent</td>\n",
              "      <td>65.0</td>\n",
              "      <td>Angenent65</td>\n",
              "      <td>...</td>\n",
              "      <td>anaerobic bioreactor</td>\n",
              "      <td>anaerobic sludge</td>\n",
              "      <td>host-associated</td>\n",
              "      <td>United States of America:State of New York:Cit...</td>\n",
              "      <td>LTA.misc.519</td>\n",
              "      <td>NaN</td>\n",
              "      <td>42.44</td>\n",
              "      <td>-76.5</td>\n",
              "      <td>Corn Beer Fermentation Bioreactor</td>\n",
              "      <td>bioreactor metagenome</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13114.angenent.65.s003</th>\n",
              "      <td>Angenent65.misc.946</td>\n",
              "      <td>bioreactor sludge</td>\n",
              "      <td>11/28/13 0:00</td>\n",
              "      <td>USA - New York</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Corn Beer Fermentation Bioreactor Biomass</td>\n",
              "      <td>123.0</td>\n",
              "      <td>Angenent</td>\n",
              "      <td>65.0</td>\n",
              "      <td>Angenent65</td>\n",
              "      <td>...</td>\n",
              "      <td>anaerobic bioreactor</td>\n",
              "      <td>anaerobic sludge</td>\n",
              "      <td>host-associated</td>\n",
              "      <td>United States of America:State of New York:Cit...</td>\n",
              "      <td>LTA.misc.104</td>\n",
              "      <td>NaN</td>\n",
              "      <td>42.44</td>\n",
              "      <td>-76.5</td>\n",
              "      <td>Corn Beer Fermentation Bioreactor</td>\n",
              "      <td>bioreactor metagenome</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>3 rows × 33 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-3654836c-4b26-4e9d-bcb8-dd413532dbbf')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-3654836c-4b26-4e9d-bcb8-dd413532dbbf button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-3654836c-4b26-4e9d-bcb8-dd413532dbbf');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-04890fb1-30ec-4fba-b5e7-6f8a4fbb6ab7\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-04890fb1-30ec-4fba-b5e7-6f8a4fbb6ab7')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-04890fb1-30ec-4fba-b5e7-6f8a4fbb6ab7 button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "df_"
            }
          },
          "metadata": {},
          "execution_count": 13
        }
      ],
      "source": [
        "df_ = meta.to_dataframe()\n",
        "df_.head(3)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "46069d92-5d4c-4a87-aae0-6dc7f32d637c",
      "metadata": {
        "id": "46069d92-5d4c-4a87-aae0-6dc7f32d637c"
      },
      "source": [
        "### Metadata column"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "96043977-0f47-486f-b4d0-d2a4801c1896",
      "metadata": {
        "id": "96043977-0f47-486f-b4d0-d2a4801c1896"
      },
      "source": [
        "One can extract a metadata column by its name. The output is a [`MetadataColumn`](https://scikit.bio/docs/latest/generated/skbio.metadata.MetadataColumn.html) object.\n",
        "\n",
        "For example, column `empo_2` describes the level 2 EMP Ontology ([EMPO](https://earthmicrobiome.org/protocols-and-standards/empo/)) of the environment from which the sample was collected."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "id": "ce9a3d83-5462-4aed-960d-fbc39c44e847",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ce9a3d83-5462-4aed-960d-fbc39c44e847",
        "outputId": "f100e802-8fb2-4ccd-fcd7-363e8719657f"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<CategoricalMetadataColumn name='empo_2' id_count=880>"
            ]
          },
          "metadata": {},
          "execution_count": 14
        }
      ],
      "source": [
        "empo_2 = meta.get_column('empo_2')\n",
        "empo_2"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "3e612e28-0a1a-4239-9985-28cd14b26490",
      "metadata": {
        "id": "3e612e28-0a1a-4239-9985-28cd14b26490"
      },
      "source": [
        "Likewise, a metadata column can be converted into a Pandas series."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "id": "54010b99-13c6-4302-858a-9bdd923ca26a",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 490
        },
        "id": "54010b99-13c6-4302-858a-9bdd923ca26a",
        "outputId": "c43be64a-8d50-4a7f-edf5-11418d2a8802"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "sample_name\n",
              "13114.angenent.65.s001    Animal\n",
              "13114.angenent.65.s002    Animal\n",
              "13114.angenent.65.s003    Animal\n",
              "13114.angenent.65.s004    Animal\n",
              "13114.angenent.65.s005    Animal\n",
              "                           ...  \n",
              "13114.zaneveld.9.s018     Animal\n",
              "13114.zaneveld.9.s019     Animal\n",
              "13114.zaneveld.9.s020     Animal\n",
              "13114.zaneveld.9.s021     Saline\n",
              "13114.zaneveld.9.s022     Saline\n",
              "Name: empo_2, Length: 880, dtype: object"
            ],
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>empo_2</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>sample_name</th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>13114.angenent.65.s001</th>\n",
              "      <td>Animal</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13114.angenent.65.s002</th>\n",
              "      <td>Animal</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13114.angenent.65.s003</th>\n",
              "      <td>Animal</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13114.angenent.65.s004</th>\n",
              "      <td>Animal</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13114.angenent.65.s005</th>\n",
              "      <td>Animal</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13114.zaneveld.9.s018</th>\n",
              "      <td>Animal</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13114.zaneveld.9.s019</th>\n",
              "      <td>Animal</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13114.zaneveld.9.s020</th>\n",
              "      <td>Animal</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13114.zaneveld.9.s021</th>\n",
              "      <td>Saline</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13114.zaneveld.9.s022</th>\n",
              "      <td>Saline</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>880 rows × 1 columns</p>\n",
              "</div><br><label><b>dtype:</b> object</label>"
            ]
          },
          "metadata": {},
          "execution_count": 15
        }
      ],
      "source": [
        "s_ = empo_2.to_series()\n",
        "s_"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "febfb64b-feee-4bab-92d6-dc5d1ae98406",
      "metadata": {
        "id": "febfb64b-feee-4bab-92d6-dc5d1ae98406"
      },
      "source": [
        "Using Pandas, one can further explore the distribution of values."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "id": "206be48c-cac4-4c42-8e20-6be7dbe22521",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 272
        },
        "id": "206be48c-cac4-4c42-8e20-6be7dbe22521",
        "outputId": "c85374b1-4a82-40f9-afe1-db87ee7fe37a"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "empo_2\n",
              "Non-saline    343\n",
              "Animal        301\n",
              "Plant         115\n",
              "Saline        109\n",
              "Fungus         12\n",
              "Name: count, dtype: int64"
            ],
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>count</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>empo_2</th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>Non-saline</th>\n",
              "      <td>343</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Animal</th>\n",
              "      <td>301</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Plant</th>\n",
              "      <td>115</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Saline</th>\n",
              "      <td>109</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Fungus</th>\n",
              "      <td>12</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div><br><label><b>dtype:</b> int64</label>"
            ]
          },
          "metadata": {},
          "execution_count": 16
        }
      ],
      "source": [
        "empo_2_freqs = s_.value_counts()\n",
        "empo_2_freqs"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "id": "95699b5e-3883-4d9c-95bb-ae5b162fb9e8",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 428
        },
        "id": "95699b5e-3883-4d9c-95bb-ae5b162fb9e8",
        "outputId": "060365c1-abc2-4344-dfa8-92f35ce0704a"
      },
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<Figure size 640x480 with 1 Axes>"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAbQAAAGbCAYAAACszmWlAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjAsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvlHJYcgAAAAlwSFlzAAAPYQAAD2EBqD+naQAAUU9JREFUeJzt3XlcFPXjBvBndpdlgeUGuRQQAVNEzau877zLMrW01MwOz8xMMzPNSkuzUruszCPLO820S03LKy3v+0AUVBZQOeSG3fn9wbf9iYAusLuf3eV5v16+kmF25llSn52Zz3xGkmVZBhERkZ1TiA5ARERkDiw0IiJyCCw0IiJyCCw0IiJyCCw0IiJyCCw0IiJyCCw0IiJyCCw0IiJyCCw0IiJyCCw0ogqQJAkzZswQHaNCdu7cCUmSsHPnTtFRbCoLOR4WWjW1dOlSSJJU7q+///7buO5/y0aMGFHmtqZOnWpc5/r168blw4YNK7FNDw8PNGrUCPPmzUN+fr7F3+Odvv/+e3z88cdW3y9V3D///IMxY8YgJiYGbm5uCA0NxYABA3Du3DnR0ciGqUQHILFmzpyJ2rVrl1oeGRlZ4muNRoP169fjs88+g1qtLvG9lStXQqPRIC8vr9R2nJ2d8fXXXwMA0tPTsX79ekycOBH//PMPVq1aZcZ3cm/ff/89Tpw4gfHjx1t1v1Rx77//Pvbs2YP+/fujYcOG0Ol0+OSTT9CkSRP8/fffaNCggeiIZINYaNVcjx490KxZs3uu1717d2zatAm//PILHnnkEePyvXv3Ij4+Hv369cP69etLvU6lUuGpp54yfj1q1Cg88MADWL16NT788EMEBweb542QQ5kwYQK+//77Eh+eBg4ciNjYWLz33ntYsWKFwHRkq3jKkUwSEhKCdu3a4fvvvy+x/LvvvkNsbKzJn5gVCgU6dOgAALh06dJd17148SL69+8PHx8fuLq64sEHH8SWLVtKrPPfNZk1a9bg3XffRc2aNaHRaNC5c2dcuHDBuF6HDh2wZcsWXL582XgKNDw83Pj9lJQUPPvsswgICIBGo0GjRo2wbNkyk97T4cOH0aNHD3h4eECr1aJz584lTtn+59ixY2jfvj1cXFxQs2ZNvPPOO1iyZAkkSTL+LIYOHQo/Pz8UFhaWev1DDz2EunXrmpTJFPv370f37t3h6ekJV1dXtG/fHnv27DF+f926dZAkCX/++Wep1y5atAiSJOHEiRPGZWfOnMHjjz8OHx8faDQaNGvWDJs2bapUtlatWpU6ExAVFYWYmBicPn26Utskx8cjtGouIyOjxHUvoPiama+vb6l1Bw0ahJdeeglZWVnQarUoKirC2rVrMWHChDJPN5YnLi4OAMrcx3+Sk5PRqlUr5OTkYNy4cfD19cWyZcvw8MMPY926dXj00UdLrP/ee+9BoVBg4sSJyMjIwJw5czB48GDs378fQPF1voyMDFy5cgUfffQRAECr1QIAcnNz0aFDB1y4cAFjxoxB7dq1sXbtWgwbNgzp6el46aWXys158uRJtG3bFh4eHpg0aRKcnJywaNEidOjQAX/++SceeOABAMDVq1fRsWNHSJKEKVOmwM3NDV9//TWcnZ1LbO/pp5/G8uXL8dtvv6F3797G5TqdDn/88QemT59u6o/5rv744w/06NEDTZs2xfTp06FQKLBkyRJ06tQJu3btQosWLdCrVy9otVqsWbMG7du3L/H61atXIyYmxvhB5uTJk2jdujVCQkLw2muvwc3NDWvWrEHfvn2xfv36Uv+/KkOWZSQnJyMmJqbK2yIHJVO1tGTJEhlAmb+cnZ1LrAtAHj16tHzz5k1ZrVbL3377rSzLsrxlyxZZkiT50qVL8vTp02UAcmpqqvF1Q4cOld3c3OTU1FQ5NTVVvnDhgjxr1ixZkiS5YcOGd803fvx4GYC8a9cu47Jbt27JtWvXlsPDw2W9Xi/Lsizv2LFDBiDXq1dPzs/PN647f/58GYB8/Phx47JevXrJYWFhpfb18ccfywDkFStWGJcVFBTILVu2lLVarZyZmVniZzF9+nTj13379pXVarUcFxdnXHbt2jXZ3d1dbteunXHZ2LFjZUmS5MOHDxuX3bhxQ/bx8ZEByPHx8bIsy7Jer5dr1qwpDxw4sETGDz/8UJYkSb548eJdf25l+e9ntGPHDlmWZdlgMMhRUVFyt27dZIPBYFwvJydHrl27tty1a1fjsieffFKuUaOGXFRUZFyWlJQkKxQKeebMmcZlnTt3lmNjY+W8vDzjMoPBILdq1UqOiooqN0tFfPvttzIAefHixRV+LVUPPOVYzX366afYunVriV+//PJLmet6e3uje/fuWLlyJYDiQRatWrVCWFhYudvPzs6Gv78//P39ERkZiddffx0tW7bEhg0b7prr559/RosWLdCmTRvjMq1Wi+effx6XLl3CqVOnSqz/zDPPlDhF1bZtWwDFpy3v5eeff0ZgYCCefPJJ4zInJyeMGzcOWVlZZZ5yAwC9Xo/ff/8dffv2RUREhHF5UFAQBg0ahN27dyMzMxMA8Ouvv6Jly5Zo3LixcT0fHx8MHjy4xDYVCgUGDx6MTZs24datW8bl3333HVq1alXmAJ6KOnLkCM6fP49Bgwbhxo0buH79Oq5fv47s7Gx07twZf/31FwwGA4Di61YpKSklhtmvW7cOBoMBAwcOBADcvHkTf/zxBwYMGIBbt24Zt3fjxg1069YN58+fx9WrV6uU+cyZMxg9ejRatmyJoUOHVmlb5Lh4yrGaa9GihUmDQv4zaNAgPP3000hISMDGjRsxZ86cu66v0Wjw008/ASge8Vi7dm3UrFnznvu5fPmy8XTd7erVq2f8/u3X7UJDQ0us5+3tDQBIS0szaV9RUVFQKEp+vrt9X2VJTU1FTk5Omde16tWrB4PBgMTERMTExODy5cto2bJlqfXuHE0KAEOGDMH777+PDRs2YMiQITh79iwOHjyIL7744p7vxRTnz58HgLsWQ0ZGhvEDjKenJ1avXo3OnTsDKD7d2LhxY0RHRwMALly4AFmWMW3aNEybNq3M7aWkpCAkJKRSeXU6HXr16gVPT0+sW7cOSqWyUtshx8dCowp5+OGH4ezsjKFDhyI/Px8DBgy46/pKpRJdunSxeK7y/pGTZdni+za3+vXro2nTplixYgWGDBmCFStWQK1W3/Nnbar/jr7mzp1b4ojxdv9dX3R2dkbfvn2xYcMGfPbZZ0hOTsaePXswa9asUtubOHEiunXrVub2yipuU2RkZKBHjx5IT0/Hrl27OCqW7oqFRhXi4uKCvn37YsWKFejRowf8/Pwssp+wsDCcPXu21PIzZ84Yv19RkiSVu69jx47BYDCUOEq71778/f3h6upabk6FQoFatWoZt3H7qMv/lLUMKD5KmzBhApKSkvD999+jV69exqPOqqpTpw4AwMPDw6QPGwMHDsSyZcuwfft2nD59GrIsG083AjCebnVycjLrh5e8vDz06dMH586dw7Zt21C/fn2zbZscE6+hUYVNnDgR06dPL/f0kjn07NkTBw4cwL59+4zLsrOz8eWXXyI8PLxS/7i5ubkhIyOjzH3pdDqsXr3auKyoqAgLFy6EVqstNcLvP0qlEg899BB+/PHHErcgJCcn4/vvv0ebNm3g4eEBAOjWrRv27duHI0eOGNe7efMmvvvuuzK3/eSTT0KSJLz00ku4ePFiiXv5qqpp06aoU6cOPvjgA2RlZZX6fmpqaomvu3TpAh8fH6xevRqrV69GixYtSlzLq1GjBjp06IBFixYhKSnpntszhV6vx8CBA7Fv3z6sXbu2zNO1RHfiEVo198svvxiPRG7XqlWrEgMdbteoUSM0atTIorlee+01rFy5Ej169MC4cePg4+ODZcuWIT4+HuvXry91vcsUTZs2xerVqzFhwgQ0b94cWq0Wffr0wfPPP49FixZh2LBhOHjwIMLDw7Fu3Trs2bMHH3/8Mdzd3cvd5jvvvIOtW7eiTZs2GDVqFFQqFRYtWoT8/PwS1xcnTZqEFStWoGvXrhg7dqxx2H5oaChu3rxZ6ujR398f3bt3x9q1a+Hl5YVevXqV2veMGTPw1ltvYceOHcZ7+0yhUCjw9ddfo0ePHoiJicEzzzyDkJAQXL16FTt27ICHh4fxuidQfOT12GOPYdWqVcjOzsYHH3xQapuffvop2rRpg9jYWDz33HOIiIhAcnIy9u3bhytXruDo0aMm5wOAV155BZs2bUKfPn1w8+bNUjdSm7PgyYEIHmVJgtxt2D4AecmSJcZ18b9h+3dzt2H7lRUXFyc//vjjspeXl6zRaOQWLVrImzdvLrHOf8PA165dW2J5fHx8qfeRlZUlDxo0SPby8pIBlBjCn5ycLD/zzDOyn5+frFar5djY2BKv/Q/uGLYvy7J86NAhuVu3brJWq5VdXV3ljh07ynv37i312sOHD8tt27aVnZ2d5Zo1a8qzZ8+WFyxYIAOQdTpdqfXXrFkjA5Cff/75Mn8+r7zyiixJknz69Okyv/+f8obKHz58WH7sscdkX19f2dnZWQ4LC5MHDBggb9++vdQ2tm7dKgOQJUmSExMTy9xPXFycPGTIEDkwMFB2cnKSQ0JC5N69e8vr1q27Z5Y7tW/f/q5/PonKIsmyHV41J3IQ48ePx6JFi5CVlVVqYMuPP/6Ivn374q+//jLehnC7Fi1aICwsDGvXrrVWXCKbxkIjspLc3Fy4uLgYv75x4waio6PRpEkTbN26tdT6vXv3xunTp3HhwoVSpyQzMzPh7++PI0eOGG8vIKrueA2NyEpatmyJDh06oF69ekhOTsbixYuRmZlZanDNqlWrcOzYMWzZsgXz588vc3Smh4eHkEfwENkyHqERWcnrr7+OdevW4cqVK5AkCU2aNMH06dNLDXWXJAlarRYDBw7EF198AZWKnzuJTMFCIyIih8D70IiIyCGw0IiIyCGw0IiIyCGw0IiIyCGw0IiIyCGw0IiIyCGw0IiIyCGw0IiIyCGw0IiIyCGw0IiIyCGw0IiIyCGw0IiIyCGw0IiIyCGw0IiIyCGw0IiIyCGw0IiIyCGw0IiIyCGw0IiIyCGw0IiIyCGw0IiIyCGw0IiIyCGw0IiIyCGw0IiIyCGw0IiIyCGw0IiIyCGw0IiIyCGw0IiIyCGw0IiIyCGw0Iio2pkxYwYaN25s/HrYsGHo27evsDxkHiw0IqqUYcOGQZIkvPfeeyWWb9y4EZIkCUpVOfPnz8fSpUtFx6AqYqERUaVpNBq8//77SEtLEx2lSjw9PeHl5SU6BlURC42IKq1Lly4IDAzE7Nmzy11n/fr1iImJgbOzM8LDwzFv3rwS3w8PD8esWbMwfPhwuLu7IzQ0FF9++eVd95uWlobBgwfD398fLi4uiIqKwpIlS4zfnzx5MqKjo+Hq6oqIiAhMmzYNhYWF5W7vzlOOHTp0wLhx4zBp0iT4+PggMDAQM2bMKPGa9PR0jBgxAv7+/vDw8ECnTp1w9OjRu+Ymy2KhEVGlKZVKzJo1CwsXLsSVK1dKff/gwYMYMGAAnnjiCRw/fhwzZszAtGnTSp3emzdvHpo1a4bDhw9j1KhRGDlyJM6ePVvufqdNm4ZTp07hl19+wenTp/H555/Dz8/P+H13d3csXboUp06dwvz58/HVV1/ho48+qtB7W7ZsGdzc3LB//37MmTMHM2fOxNatW43f79+/P1JSUvDLL7/g4MGDaNKkCTp37oybN29WaD9kRjIRUSUMHTpUfuSRR2RZluUHH3xQHj58uCzLsrxhwwb5v39aBg0aJHft2rXE61599VW5fv36xq/DwsLkp556yvi1wWCQa9SoIX/++efl7rtPnz7yM888Y3LWuXPnyk2bNjV+PX36dLlRo0ZlvhdZluX27dvLbdq0KbGN5s2by5MnT5ZlWZZ37dole3h4yHl5eSXWqVOnjrxo0SKTc5F58QiNiKrs/fffx7Jly3D69OkSy0+fPo3WrVuXWNa6dWucP38eer3euKxhw4bG30uShMDAQKSkpAAAevToAa1WC61Wi5iYGADAyJEjsWrVKjRu3BiTJk3C3r17S+xj9erVaN26NQIDA6HVavHGG28gISGhQu/p9kwAEBQUZMx09OhRZGVlwdfX15hNq9UiPj4ecXFxFdoPmY9KdAAisn/t2rVDt27dMGXKFAwbNqzCr3dycirxtSRJMBgMAICvv/4aubm5Jdbr0aMHLl++jJ9//hlbt25F586dMXr0aHzwwQfYt28fBg8ejLfeegvdunWDp6cnVq1aVeraXVUyZWVlISgoCDt37iz1Og4uEYeFRkRm8d5776Fx48aoW7eucVm9evWwZ8+eEuvt2bMH0dHRUCqVJm03JCSkzOX+/v4YOnQohg4dirZt2+LVV1/FBx98gL179yIsLAxTp041rnv58uVKvKPyNWnSBDqdDiqVCuHh4WbdNlUeC42IzCI2NhaDBw/GggULjMteeeUVNG/eHG+//TYGDhyIffv24ZNPPsFnn31WpX29+eabaNq0KWJiYpCfn4/NmzejXr16AICoqCgkJCRg1apVaN68ObZs2YINGzZUaX936tKlC1q2bIm+fftizpw5iI6OxrVr17BlyxY8+uijaNasmVn3R6bhNTQiMpuZM2caT8sBxUcya9aswapVq9CgQQO8+eabmDlzZqVOS95OrVZjypQpaNiwIdq1awelUolVq1YBAB5++GG8/PLLGDNmDBo3boy9e/di2rRpVdrfnSRJws8//4x27drhmWeeQXR0NJ544glcvnwZAQEBZt0XmU6SZVkWHYLIlt3KK0RSRh5SMvORU1CE3EI98gsNyCvSI69Qj9yC//99XqEB+YV65BcZoFJKcFUroXFSwsVJ+f+/V5f82lWtgo+bGiFeLnBRm3YajohK4ylHqtZyC/S4lpGLpPQ843+TMnJxLSMPSem5SMrIQ1Z+kdXyeLs6IdjLBcFeLgjxckGwl6bE1zXcne1uWikia+ERGlUbSRm5OHk1EyevZeLktQycvJaJq+m5omNViFqpQKivK2KCPdAg2BMxIR5oEOIJD43TvV9M5OBYaORwDAYZ8TeyjcV16lpxid3MLhAdzSIkCajl7YoGIR6ICfZEgxBPNAj2gK/WWXQ0IqtioZHd0xtkHL2Sjn1xN7DnwnUcTUxHdoH+3i90cEGeGjQJ9UbbKD+0i/ZHsJeL6EhEFsVCI7t0ISULf55LxZ4L1/FP/E3csuJ1LntVx98N7aL90S7KHw9G+HIACjkcFhrZhdwCPfbGXceOsynYeTYVV9Ls69qXrVGrFGge7o22UcUFVz/YQ3QkoipjoZHNyivUY/vpFGw8chV/nktFQZHh3i+iSvF3d0bn+2qgT6NgtIzwhULBkZRkf1hoZFMMBhn7Lt7AxsNX8esJHU8lCuDv7oxesUHo0ygYTcO8RcchMhkLjWzCyWsZ2Hj4Kn46mgRdZp7oOPQ/tXxc8Oj9NdG/aU3U8nEVHYforlhoJMyVtBz8eOQafjxyFeeSs0THobuQJKB5uA8eb1oTvWKD4ObMORnI9rDQyOr2X7yBr3fHY/vpZBj4p8/uuKmV6Ne0Jp5tUxthvm6i4xAZsdDIKor0Bmw5noTFu+Nx7EqG6DhkBgoJeKh+IJ5rVxtNw3xExyFioZFlZeYVYuX+BCzbewnXMnhtzFE1CfXCc20j0C0mkCMkSRgWGllE4s0cLN4dj7X/JnLWjmokzNcVw1vXRv9mNeGq5nU2si4WGpnVqWuZWPjHefx2UsfrY9WYp4sTBj8QiuFtasOPc0qSlbDQyCwSb+Zg3u9n8ePRa+CfKPqPm1qJEW0j8EL7CB6xkcWx0KhK0rILsPCPC1ix/zJn8qBy+bs746XOUXiieS2olArRcchBsdCoUnIL9PhmTzy++DMOt/I4mweZpo6/GyZ1vw/dYgJFRyEHxEKjCtEbZKz5NxEfbzuH5Mx80XHITjUL88aUnvU4tRaZFQuNTPb7SR3m/HYWF1I4qweZR/eYQEzucR9q+/EGbao6Fhrd09X0XLyx4Th2nE0VHYUckEohYXib2pjQNRoaJz6jjSqPhUblkmUZy/ddxpxfz/BeMrK4cF9XvNevIR6M8BUdhewUC43KdCHlFiavP46Dl9NER6FqRJKAJ1uEYkqP++CucRIdh+wMC41KKNQb8NmOOHy68wKH4ZMwQZ4avNO3ATrXCxAdhewIC42MjiSmY/K6YzibfEt0FCIAwMONgjHj4Rj4uKlFRyE7wEIj5BQU4YPfzmHp3nhOV0U2x8dNjel96uORxiGio5CNY6FVc6eTMjH6u0O4eD1bdBSiu+paPwAfPN4Inq68tkZlY6FVY6v/ScD0TSeRV8hrZWQfanq74LPBTdCwppfoKGSDWGjVUG6BHm9sPIH1h66IjkJUYWqlAm/0rochLcNFRyEbw0KrZi6kZGHUdwdxLpmzfZB9e7hRMGY/Fgs3Z87iT8VYaNXIxsNX8fqG48jhTdLkIOr4u+GLp5oiKsBddBSyASy0aiCvUI+3fjqFlQcSREchMjtXtRLvPtoAj95fU3QUEoyF5uAu38jGqO8O4eS1TNFRiCxq0AOhmN6nPpxVnA+yumKhObBDCWkYsexf3MwuEB2FyCoa1fLCN0ObwVfrLDoKCcBCc1BbTyVj7MpDHJJP1U64ryuWDW+BMF8+kqa6YaE5oG//vowZm05Cz2k/qJry06rxzbDmvF+tmmGhORBZljHnt7P4fGec6ChEwrmqlfhscBN0qFtDdBSyEhaagyjUGzBp3TFsOHxVdBQim6FSSJj9WCz6N6slOgpZAQvNAdzKK8TIFYew+8J10VGIbNLEh6IxplOU6BhkYSw0O5ecmYdhS/7B6SQOyye6m6ceDMXMhxtAoZBERyELYaHZscSbOXjiy79xNT1XdBQiu/BQ/QAsHHQ/71VzUArRAahyrqXn4smvWGZEFfH7qWSM/u4QCvW8ncURsdDsUEpmHgZ/vR9X0lhmRBW17XQKxq8+wttaHBALzc7cyMrHoK/3I54P5CSqtC3HkvDquqPgFRfHwkKzI+k5BRj89X5cSOGjX4iq6odDVzF14wnRMciMWGh2IjOvEE8vPoAzuluioxA5jO/3J2DmT6dExyAzYaHZgez8Igz95gCOX80QHYXI4XyzJx5zfj0jOgaZAQvNxuUW6PHM0n9wOCFddBQih/XZzjgs3H5edAyqIhaaDSvUG/D8t//iQPxN0VGIHN68refw9a6LomNQFbDQbNgbG05g13lOZ0VkLe/+fBq/nkgSHYMqiYVmoz7fGYfV/yaKjkFUrcgyMGHNUZziE97tEgvNBv1yPAlzfuNFaiIRcgr0eG75v7ielS86ClUQC83GXLiSgpfXHAHv9yQS52p6Ll789iAKijhFlj1hodmSWzrUWdsRH4ftE52EqNr793Ia3th4XHQMqgAWmq0oKgBWPw0p4wq6X5mPbVHr4aLUi05FVK2t+fcKFu+OFx2DTMRCsxVbXgauHDB+GZm4HvtrfoJwlzyBoYho1s+n8de5VNExyAQsNFtw4Cvg8IpSiz2S92Ob+wx09eN9aESi6A0yxnx/CBdTOYeqrWOhiZZ0DPjt9XK/rcpMwJcFr2F8KG/4JBIlM68II5b/i1t5haKj0F2w0EQqyAHWPwvoC+66mlSQhZdS38TiKA4WIRLlYmo2pnF2fpvGQhPp18nA9XMmrSrJBnROXIgdkWvgpuRQYiIRNh65hh8OXREdg8rBQhPl1I/AoeUVflntKxuxP+QjRLnxadVEIrz540lcvsEH7NoiFpoIGVeATeMq/XJtykH84jodPfw5zyORtWXlF2HcqiMo1PNMia1hoVmbwQD88DyQl16lzahuXcFneVMwKYyPvCCytqOJ6Zi/jX/3bA0Lzdp2fQBc3mOWTUmF2RiZPAPLo3aZZXtEZLrP/4zD4YQ00THoNpIsc9ZAq0k8AHzTHZDNPwNIYs1e6HlpIG4Vqcy+bSIqW4S/G34e1xYaJ6XoKAQeoVlPUQHw4xiLlBkA1LqyBX8HfYh62hyLbJ+ISruYmo33fuGTMWwFC81a9swHrp+16C7cUo9gs+ZNPBKQYtH9ENH/W7bvEvZe4AAtW8BCs4YbccXXzqxAmXUNH+e8jjfCLVueRFRMloHXNxxHfhEnExeNhWYNWyYARdabZFgqzMGzuplYGbUDksRLpESWdulGDr76i9PTicZCs7Rja4CLO62+WwkyWiZ+hT0R38LTqcjq+yeqbj7dEYdr6ZzwQCQWmiXlpt114mFrCL76K/YFzEWsO2c2ILKk3EI93t1yWnSMao2FZklb3wSyxT9HyfX6cWxUv4HHA5NFRyFyaFuOJ3GAiEAsNEu5vA849K3oFEbK7GTMzZqCmbVPiY5C5NCmbzqJIk6LJQQLzRJkGfj5VQC2NSBDKsrDkKR3sD5qKweLEFnI+ZQsLN17SXSMaomFZgnH1wLJx0WnKFfTxCX4u/YS+Kr5sEIiS5i/7TxSb+WLjlHtsNDMragA+OMd0SnuKeDaNuzxfx9NPPlYeSJzu5VfhNm/cICItbHQzO3gEiD9sugUJtHcOIV1ytcxKChJdBQih7Ph8FUcvMzJi62JhWZO+VnAX3NFp6gQRc51vJs5BbMjbPcUKZE9kmXgw62csceaWGjmtO9TmximX1GSvgBPXpuNH6N/gVLi6Cwic9lz4Qb+vXRTdIxqg4VmLtnXgb0LRaeokkYJ32J/7a9Rw5mDRYjMZf52PgjUWlho5vLXB0DBLdEpqszv2k7s8pmFFl6ZoqMQOYRd56/jEB8EahUsNHNITwD+XSw6hdk4p53FKmkqhgVfER2FyCEs4FGaVbDQzGHPfEBfIDqFWSlyb2B6+lR8WOew6ChEdm/n2VQcTUwXHcPhsdCqKucmcOR70SksQjIU4rGrc/Fz1E9wUnBmEaKq4FGa5bHQquqfxUBhjugUFlU/cSX2h36OEA1nPiCqrO1nUnDiaoboGA6NhVYVRfnAgS9Fp7AKH91u7PB6B218+BeSqLI44tGyWGhVcWw1kJ0iOoXVqNPjsNwwBc/XTBAdhcgubT2VjJPX+KHQUlholSXLxTdSVzOKvHRMufkGFkYeFB2FyC4t40z8FsNCq6zzW4HUM6JTCCEZitDnyjz8FrURzgrOLEJUET8dTUJmHicvsAQWWmXtXSA6gXB1E9dgf+inCHXJEx2FyG7kFuqx8fBV0TEcEgutMpKOApd2iU5hE7x0+7Dd42108OFMCESm+n4/r0NbAgutMv79RnQCm+KUEY8l+ikYU+uS6ChEduGM7hYfLWMBLLSKKswDTmwQncLmSPmZeOX6NCyK3C86CpFd4FGa+bHQKurMZiCfw27LIsl6dLsyH9uj1sNFqRcdh8imbTl+DRm5HBxiTiy0inLQaa7MqU7ieuyvuRARrhwsQlSevEIDfjjECcDNiYVWEZlJwMUdolPYBY/kA/hdOwMP+fHhhkTlWXmApx3NiYVWEcdWATLvuzKVKjMBiwpew8uhF0VHIbJJ55Kz+ERrM2KhVcSRlaIT2B2pIAvjUt/EN1F7REchsklr/k0UHcFhsNBMdeUgcP2s6BR2SZIN6JT4KXZGroabioNFiG637XQK9AY+nskcWGimOsrBIFUVfuVH7A/+GNFuuaKjENmMm9kFOBDP047mwEIzhcEAnOS9Z+agTTmIn13fRG//66KjENmM307qREdwCCw0U1w5AOTcEJ3CYahuXcXCvCmYHMZnQxEBxY+VoapjoZni7C+iEzgcqTAbLybPwIqoP0VHIRLuanoun2ZtBiw0U5z7TXQChyRBRpvERdgd+R3cVUWi4xAJxdOOVcdCu5e0S0DqadEpHFrNK1uwP3Ae6mlzREchEoaFVnUstHvh0ZlVuF4/is2aaegbkCI6CpEQ55KzEH89W3QMu8ZCu5dzv4pOUG0os5LwUfYUTKtdPZ8ETsSjtKphod1NfhZwabfoFNWKVJSLZ5NmYnXUDkgSbzal6oWFVjUstLuJ+wPQF4hOUS09kPgV9kYsh7cTB4tQ9XEkMR0ZOXykTGWx0O7mPK+fiRR09TfsrTEHDT2yREchsgpZBg4mcNaQymKh3c0lTqgrmsuNE9jg9Ab6B/JUDFUP/1xKEx3BbrHQypOVAqTFi05BAJTZKZiT9TreiTgpOgqRxfFxMpXHQitP4n7RCeg2UlEenrr2Ln6I/p2DRcihHb2SgfwiPpWiMlho5WGh2aQmCUuxv/Y38Ffzwjk5poIiA6fBqiQWWnkSD4hOQOWocW07dvm9hyaet0RHIbIIXkerHBZaWYoKgGtHRKegu9DcPI11iql4Kuiq6ChEZsfraJXDQitL0hFAny86Bd2DIvc63s6civcijouOQmRWBy+nQZZ5rbiiWGhl4fUzuyHpC/DEtdnYFPULlJJBdBwis0jLKcSFFN5/WVEstLKw0OxOw8RvcSD8KwQ6c2YXcgy8jlZxLLSyXD0kOgFVgm/Sn/jTZxYe8MoUHYWoyk5c40jHimKh3Sk3HcjkQAN75Zx2Diul1zE8JFF0FKIqieMpxwpjod0plY8usXeK3JuYlvYGPqxzWHQUokq7yGejVRgL7U4pfDq1I5AMhXjs6lz8HPUTnBQcLUb2J/VWPjLzOIFARbDQ7sQjNIdSP3ElDoR+hhANb8Mg+3MxlUdpFcFCuxMLzeF46/Zgh9c7aOPDi+xkX3gdrWJYaHe6cVF0ArIAdXoclhum4IWaCaKjEJksLpWFVhEstNsV5QOZV0SnIAtR5KXjtZtv4JPIf0RHITIJTzlWDAvtdjfjAZmzTTgyyVCE3lc+wu9RG+Cs4P9rsm08QqsYFtrtbsaJTkBWEp24FgdqfYJQlzzRUYjKdflGDvQGjtI1FQvtdmmXRScgK/JM/hvbPWaiky+nGCLbVKA3IPFmjugYdoOFdrusZNEJyMqcMi5hcdFrGBsaLzoKUZkS01hopqpUoXXq1Anp6emllmdmZqJTp05VzSRO9nXRCUgAKf8WJqS+iS8j/xYdhaiUm9mccNtUlSq0nTt3oqCg9A85Ly8Pu3btqnIoYbJTRCcgQSRZj4euLMAfkWvhpuRgEbIdaSw0k6kqsvKxY8eMvz916hR0Op3xa71ej19//RUhISHmS2dt2amiE5BgEVc2YF/IFTx6/UXE5biIjkOEmzmc/spUFSq0xo0bQ5IkSJJU5qlFFxcXLFy40GzhrC6LhUaAR8o/+M1Dh1Guk/H7dR/RcaiaS8/hEZqpKlRo8fHxkGUZEREROHDgAPz9/Y3fU6vVqFGjBpRKpdlDWg2P0Oh/VJmJWKR+DZ+ETca8y3VEx6FqjNfQTFehQgsLCwMAGAwOeI0hLwPQcwJb+n9SQRbGpExHg6iReOZ8a9FxqJpK4xGaySpUaLc7f/48duzYgZSUlFIF9+abb1Y5mNVxhCOVQZIN6Jj4Kf6MTETPS/2RXWTHZyDILqVl8xqaqSpVaF999RVGjhwJPz8/BAYGQpIk4/ckSbLPQsviCEcqX9iVTdgflIjHbo7BuWwOFiHr4RGa6SRZlis8r0pYWBhGjRqFyZMnWyKTGGd/BVYOFJ2CbFyRewjGYzI2p/qJjkLVhMZJgTNv9xAdwy5U6j60tLQ09O/f39xZxNLzUxDdm+rWVSzMm4LXws6JjkLVRF6hATkFRaJj2IVKFVr//v3x+++/mzuLWCw0MpFUmI0Xkt/Cd1F/io5C1URGLq+jmaJS19AiIyMxbdo0/P3334iNjYWTk1OJ748bN84s4azKwE9AZDoJMlonLsLuyET0uvwEMgorPb6K6J6K9Jxx3xSVuoZWu3bt8jcoSbh40Q6f+nzoW2DTGNEpyA7l+DVC/4wxOHnLTXQUclA7JnZAbT/++bqXSn2sjI93wJnJecqRKsn1+lFscpuGVwNeww/JNUTHIQfEZ6KZho+P+Q9POVIVKLN1mJc9BW/WPi06CjkgQ8VPpFVLlTpCGz58+F2//80331QqjFB6XnSlqpGKcjE86W00iHoWAy90gixL934RkQl4Dc00lSq0tLSST/gtLCzEiRMnkJ6ebr/PQ+MpRzKTFomL8UmEhKA9+Sh0r4ECN18UuHiiUO2OAoUbCuCMQoMShYUKyA44ixyZn4+BH45MUalC27BhQ6llBoMBI0eORJ06djqRK085khl963ceIwK8EPLj9nLXkSUJBu8A6H1DUOQdAL2HPwpcfVDo7IkClRsKJA3y9U7IK5SQlysjP08P8IN6tcQxtKYx289JoVBgwoQJ6NChAyZNmmSuzVqPxE9AZD5J+WmYUC8eS240guvuo2WuI8kylDd1UN7UQW3CNg1qZxh8Q6D3DYbe0x+FWn8UunqhUO2BAqUr8mVn5OlVyMsH8nINKCrg4Z+jkKr479OwYcOwbNmyUsvPnz+PyMjIKm3blpi1+OPi4lBUZKdHOk4cEkvmUaB0Rlp+BmQJGNv6Ar6+EQnp9IUqb1dRkA9F0kWokky7Lcbg5gm9Xwj0PsEo8vRHoZsPCjVeKHDSokDhinyDE/KLlMjNA/Jz9DBwJJ3NqmqhAUD37t2xZMmSEstufwSYI6hUoU2YMKHE17IsIykpCVu2bMHQoUPNEszq1Cw0Mg+dVxBkFB8d3VLkY3LvdMxND4CclGzVHIrsDCiyM+B0+dQ91zWe/vSrCb1XAAo9/FHk5oMCtUfJ058FCuTmGVCQq7fCO6D/qNRVH5Du7OyMwMDAEsuGDRuG9PR0bNy40bhs/PjxOHLkCHbu3AkA6NChAxo2bAiNRoOvv/4aarUaL774ImbMmGF8zZkzZzBixAj8+++/iIiIwIIFC9C1a1ds2LABffv2xc6dO9GxY0ekpaXBy8sLAHDkyBHcf//9iI+PR3h4OC5fvowxY8Zg9+7dKCgoQHh4OObOnYuePXua/B4rVWiHDx8u8bVCoYC/vz/mzZt3zxGQNouFRmaic/cH8P/ldUmVjo+fqIXxX7lBzsoWF+wubj/9CQD3ep6ArNZA7xeCIt9g6D0DUKj1RZGLFwrU7sbTn/n/O/2Zm2OAvpCnP6tC6ST2Dqtly5ZhwoQJ2L9/P/bt24dhw4ahdevW6Nq1K/R6Pfr27YvQ0FDs378ft27dwiuvvFLhfYwePRoFBQX466+/4ObmhlOnTkGr1VZoG5UqtB07dlTmZbZNXbEfHFF5dC4eQF7Jo7E9mkSEDqmHRxedAvT2f3QjFeRBdS0OqmtxJq1v0HpD718Teu9AFHrWQJHx9Kcb8iUXFBjUyCtUIC8PyMst4ujPOzipq/4cvs2bN5coiB49esDNzbQP8g0bNsT06dMBAFFRUfjkk0+wfft2dO3aFVu3bkVcXBx27txpPAJ899130bVr1wrlS0hIQL9+/RAbGwsAiIiIqNDrgSpeQ0tNTcXZs2cBAHXr1rXv87E8QiMzSVI7A3mll6/0PI3wwU1w//J/rB9KMEVWGhRZaXCKPw7NPdaVFUoYfAJR5BcCvVcgitz9UOjqjQJnDxQq3ZAvaYqP/gqKR38W5Nn/B4R7UTlXvdA6duyIzz//3Pi1m5sbpkyZYtJrGzZsWOLroKAgpKQUP0Py7NmzqFWrVonTmS1atKhwvnHjxmHkyJH4/fff0aVLF/Tr16/Ufu+lUoWWnZ2NsWPHYvny5canVSuVSgwZMgQLFy6Eq6trZTYrFguNzESnLP8C/uyQw1jQuzkCN1e/UjOVZNBDef0qlNevmrS+7OwCvX8t6H2CUOgZgCKtLwpdvFCo1iJf4YoC2Rl5Rcri0Z85euiL7Gvwi0qtgEJR9UEhbm5upUY0KhQK3Dmdb2Fh6Ukm7pyAXpIk47/9plAoik+Z3r6vO/czYsQIdOvWDVu2bMHvv/+O2bNnY968eRg7dqzJ+6n0oJA///wTP/30E1q3bg0A2L17N8aNG4dXXnmlxKcAu8FTjmQmOvnus868HHsES240hGbfMSslcmxSfi5UV85BdeUcnE1Y3+DpV3z05x2EIo//jf509kSBkxsKJBfkG5yKT3/mysjP1UP0rFNOGsvdhebv748TJ06UWHbkyJFSBXY3devWRWJiIpKTkxEQEAAA+Oefkh/Y/jt7l5SUBG9vb+N+7lSrVi28+OKLePHFFzFlyhR89dVXli+09evXY926dejQoYNxWc+ePeHi4oIBAwbYaaHxCI3MQ1eUddfv6yFjbLuL+PJmBKSzdvhkCjunyLgOdcZ1AGXfH3g7WamC3i8YBp9gFHkH/u/ozxuFzh7IV7qhQCq+9y8/X0JurgGF+ea/+KdxtVyhderUCXPnzsXy5cvRsmVLrFixAidOnMD9999v8ja6du2KOnXqYOjQoZgzZw5u3bqFN954A8D/324QGRmJWrVqYcaMGXj33Xdx7tw5zJs3r8R2xo8fjx49eiA6OhppaWnYsWMH6tWrV6H3U6mfVE5OjrGJb1ejRg3k5ORUZpPiOfMIjcxDl592z3UyFHl4o082ZqX7Q05OtUIqqgxJXwRVcgKQnGDaze8at+LBLz5BKPKsgSK34tOfBU7uKFC6FI/+LFIiLw/IzdHDYMIcjRqt6UdLFdWtWzdMmzYNkyZNQl5eHoYPH44hQ4bg+PHjJm9DqVRi48aNGDFiBJo3b46IiAjMnTsXffr0gUZTfMXUyckJK1euxMiRI9GwYUM0b94c77zzDvr372/cjl6vx+jRo3HlyhV4eHige/fu+Oijjyr0fir1PLTOnTvD19cXy5cvNwbOzc3F0KFDcfPmTWzbtq2imxRPloF3anBOR6qSLI0HWgZ5mbx+h9wwjP4yCbK9fhCkKtF7+cPgG4Iin/+d/nT1QaHGAwUqLQokDfIMTvAP80T7Z0w/YrIFe/bsQZs2bXDhwgWrTodYqSO0jz/+GN27d0fNmjXRqFEjAMDRo0fh7OyM33//3awBrUaSAPcgIP2y6CRkx5I8gwDkmrz+TpfLCB0Sg95fnXCI4fxUMcr0VCjTU+EUd6TcdbwHPQnAtgttw4YN0Gq1iIqKwoULF/DSSy+hdevWVp/bt1KFFhsbi/Pnz+O7777DmTNnAABPPvkkBg8eDBeXe92SacM8a7LQqEp0Wh/AYNrovP8s9z6JsCebInbFAQulInumqmH7D429desWJk+ejISEBPj5+aFLly6lrpFZQ6UKbfbs2QgICMBzzz1XYvk333yD1NRUTJ482SzhrM4jRHQCsnNJGi1QibOHb9c6hE97Nof/zxzOTyWp/G2/0IYMGYIhQ4aIjlG5J1YvWrQI9913X6nlMTEx+OKLL6ocShhPFhpVjc7JlKEDZXup4VHkt2hgxjTkCOzhCM1WVKrQdDodgoKCSi339/dHUlJSlUMJwyM0qqLkKky5VyQZML5jAhAZbrY8ZP9YaKar1F+/WrVqYc+ePaWW79mzB8HBwVUOJYxnTdEJyM4lGcqY86oCbihy8GbfPCj8fM2UiOydUwALzVSVuob23HPPYfz48SgsLESnTp0AANu3b8ekSZMqNcuyzeARGlWRrvBWlbdxxuk6Pn8qHC8uyoacW7WCJPsmOTtD+b/HrdC9VarQXn31Vdy4cQOjRo1CQUHxfVsajQaTJ082ebJLm8QjNKoCGRKS826aZVvbXS4h7OkG6P71MaACc+aRY1GH1hIdwa5U6sbq/2RlZeH06dNwcXFBVFQUnJ1NmUnNxr0bBBTyJlequOtaf3T0N+9tK2/FN0G9VRzOX1159OmDkLlzRMewG1V6apxWq0Xz5s3RoEEDxygzAPCLEp2A7FSyZ+C9V6qg6bUP4Wa3ZmbfLtkHzX11RUewK2Ifg2qLAmJFJyA7leTqZZHtjrv/OAqbxVhk22TbnMu4PYrKx0K7UwD/4aDK0Wks88SGAkmPlztdBSJCLbJ9sl0aFlqFsNDuFMgbW6lydCrLPeYjRZmFmY/pIfl4W2wfZFuU/n5Q+fL2jYpgod0pgIVGlZMkWXZy4RNOyVj8VA1IjnK9mu5Kc1/FngVGLLTSXH0Adzu+OZyE0ektf8/Yr25x2DakfvHTIcihcUBIxbHQysLTjlQJuoIMq+xnkd9xnB/Q3Cr7InE4IKTiWGhl4WlHqqAihQrX881zU7UppkYcQnqXplbbH1kfB4RUHAutLDxCowpK8QyGQbbujB5jm51A0f28zuKIJI0G6vBw0THsDgutLIENRScgO5Pk7mf1feZLekzomgQpjFO2ORrnqChISqXoGHaHhVYWvyjAzV90CrIjOldPMftVZuHdxwHJ20vI/skyOCCkclho5QlvIzoB2RGdWiNs30fUOix7KhCSuvIPFyXbwgEhlcNCK094W9EJyI4kqcT+VdqsvYCdT/Par6NwadRYdAS7xEIrDwuNKiAZRaIj4NMax3Dx8RaiY1AVKf38oImpLzqGXWKhlcc/GtCaf/Z0cky6Itt45NBrUYeQ2bGJ6BhUBdo2bSDxxvlKYaHdDa+jkYmS8tNERzAa88Ap6BtyUIG90rZvJzqC3WKh3Q0LjUyQq3ZFRkGm6BhGeVIRJnZPhVSLU7jZHaUSbq1bi05ht1hod1Obn5To3nSeQaIjlHJVmYn3+6sgeXqIjkIV4NK4MZQe/H9WWSy0u/Gtw4mK6Z6StLb5iI9/na/h+6drAk5OoqOQibTt+CG6Klho9xLZSXQCsnHJLrb7iXqD+znsfYoz39gLbTuOrq4KFtq91HtYdAKycUlOtn1D88eBR5HwGIfz2zpVjRrQ1OPcnFXBQruXiI6As+1+AifxdErbH2L9avQhZLe7X3QMugu3thyEVlUstHtRqYHo7qJTkA3TyQWiI9yTLAFjW52FISZKdBQqh7Zde9ER7B4LzRT1edqRypdUeEt0BJNkSQWY1CsNUjAnDLA5Tk5wa91KdAq7x0IzRWQXwMlNdAqyUcl51nuwZ1UlKNPxwRPOkNy1oqPQbVwbN4ZSy/8nVcVCM4WTCxDVVXQKskEZrt7I1eeJjlEh+52vYu2QcEClEh2F/oezg5gHC81UPO1IZUjysL2bqk2xxuMM/nmqsegYBAAKBTx69BCdwiGw0EwV1Q1QiXvmFdkmndZHdIRKmxt0BFcfaS46RrXn1qoVnEJCRMdwCCw0Uzlri6+lEd0mSeMqOkKVTKh3GLmtG4mOUa15DegvOoLDYKFVRKMnRScgG6Oz82mlZAkY0+YC5Hp1REeplpT+fnDv2FF0DIfBQquI6O6ANkB0CrIhOkkWHaHKbinyMaV3JqTAGqKjVDtefR+FZOcfimwJC60ilCoepVEJOoN9jXAsz0VVGuY/6QZJy9tTrEaS4NX/cdEpHAoLraKaDBGdgGyIzoaeg1ZVuzWJ2DAkAlAqRUepFlwffADq0FDRMRyKTRTajBkz0LhxY/vYj28dIJwzYhNgkBRIsaObqk3xvedpHB7cRHSMasG7PweDmJvZCm3fvn1QKpXo1atXhV87ceJEbN++3VxRLK/ZcNEJyAakegSgSC4SHcPsZoccRnJvDue3JKWPD9y7cNS0uZmt0BYvXoyxY8fir7/+wrVr1yr0Wq1WC19f23xIYpnq9QG0nA+vutN5OO4givGxR5DXks9RsxTPvn0hqW37sUP2yCyFlpWVhdWrV2PkyJHo1asXli5davzezp07IUkStm/fjmbNmsHV1RWtWrXC2bNnjevceSpw2LBh6Nu3L2bNmoWAgAB4eXlh5syZKCoqwquvvgofHx/UrFkTS5YsKZFj8uTJiI6OhqurKyIiIjBt2jQUFhaa4y2WpHTitTRCkouX6AgWo4eMse0uAtG1RUdxSBwMYhlmKbQ1a9bgvvvuQ926dfHUU0/hm2++gSyXHM48depUzJs3D//++y9UKhWGD7/7abs//vgD165dw19//YUPP/wQ06dPR+/eveHt7Y39+/fjxRdfxAsvvIArV64YX+Pu7o6lS5fi1KlTmD9/Pr766it89NFH5niLpTV7BlBwLrzqLFnjIjqCRWUo8jD1kRxINfxER3Eors2bw7k2PyhYglkKbfHixXjqqacAAN27d0dGRgb+/PPPEuu8++67aN++PerXr4/XXnsNe/fuRV5e+UOefXx8sGDBAtStWxfDhw9H3bp1kZOTg9dffx1RUVGYMmUK1Go1du/ebXzNG2+8gVatWiE8PBx9+vTBxIkTsWbNGnO8xdI8goEG/SyzbbILumowue951Q18OsgTkotjl7c1cWYQy6lyoZ09exYHDhzAk08W35+lUqkwcOBALF68uMR6DRv+//n4oKDiCV1TUlLK3W5MTAwUiv+PFxAQgNjYWOPXSqUSvr6+JbaxevVqtG7dGoGBgdBqtXjjjTeQkJBQtTd4N21fAWD7Tysmy0hywAEhZdnpchmbh0YDCpsYFG3XlH5+cH/oIdExHFaV/4QuXrwYRUVFCA4Ohkqlgkqlwueff47169cjIyPDuJ7TbXfDS1JxCRgMhnK363TH3fOSJJW57L9t7Nu3D4MHD0bPnj2xefNmHD58GFOnTkVBgQWfJuxft3iACFVLOkOu6AhWs8z7JI4/2Ux0DLvnO3w4FM7OomM4rCoVWlFREZYvX4558+bhyJEjxl9Hjx5FcHAwVq5caa6c97R3716EhYVh6tSpaNasGaKionD58mXL77jdRMvvg2ySLj9ddASrejv0EFJ7cjh/ZSl9feH95BOiYzi0KhXa5s2bkZaWhmeffRYNGjQo8atfv36lTjtaUlRUFBISErBq1SrExcVhwYIF2LBhg+V3HNQIiOIphOqmQOmMm9Ws0ADgpYZHUdCigegYdsn3mWFQ8FqkRVWp0BYvXowuXbrA09Oz1Pf69euHf//9F8eOHavKLkz28MMP4+WXX8aYMWPQuHFj7N27F9OmTbPKvtHuVevsh2xGslcQZNj/xMQVVSQZ8FLHBCAyXHQUu6L09ob3oEGiYzg8Sb5zfD1VztLewKVdolOQlRwIb45npWTRMYS5r9AP7ywtgOG6Y039ZSn+L78MvxeeFx3D4XHYkrnwWlq1onPxEB1BqDNO1/HFYF9ILnyK+70oPT3hPXiw6BjVAgvNXCI6ADVbiE5BVqJTc6TaNtd4/DrkPg7nvwff55+DUuBjeZYuXQovLy/j19aaDF4E/kk0py4zRCcgK0lS8v5DAFjscwKnB3A4f3lUgYHw/t+kE5WVmpqKkSNHIjQ0FM7OzggMDES3bt2wZ8+eSm3P7iaDrwAWmjmFt+Z9adWETrbAHKF2anrtQ7j5EEutLP5jRlf5vrN+/frh8OHDWLZsGc6dO4dNmzahQ4cOuHHjRqW2Z3eTwVcAC83cur4NKHk6ytHpirJFR7Ap45ocR2HT+qJj2BR1nTrwfPTRKm0jPT0du3btwvvvv4+OHTsiLCwMLVq0wJQpU/Dwww8DAD788EPExsbCzc0NtWrVwqhRo5CVlVXuNsubDP6DDz5AUFAQfH19MXr06BITu+fn52PixIkICQmBm5sbHnjgAezcubNK780SWGjm5lMbeOAF0SnIwnT5HN13uwJJj5c7X4NUm09g/o//+JcgVfHp31qtFlqtFhs3bkR+fn6Z6ygUCixYsAAnT57EsmXL8Mcff2DSpEkV2s+OHTsQFxeHHTt2YNmyZVi6dGmJp6aMGTMG+/btw6pVq3Ds2DH0798f3bt3x/nz56vy9syOhWYJ7V4F3PxFpyALydJ4IKuQR2h3SlFm4a1+ekg+3qKjCOfSqBE8unat8nZUKhWWLl2KZcuWwcvLC61bt8brr79e4v7e8ePHo2PHjggPD0enTp3wzjvvVHhSdm9vb3zyySe477770Lt3b/Tq1ct4nS0hIQFLlizB2rVr0bZtW9SpUwcTJ05EmzZtSj3CSzQWmiVoPICOr4tOQRai8wwSHcFmnXBKxjdP1YBUnecrVCgQMOU1s22uX79+uHbtGjZt2oTu3btj586daNKkifEIatu2bejcuTNCQkLg7u6Op59+Gjdu3EBOTo7J+4iJiYHytqPJoKAg48Tvx48fh16vR3R0tPGIUavV4s8//0RcXJzZ3qc5sNAspclQoAavKTiiJK2P6Ag27Re3OGwfUh+QqudIUO/Bg+Fi5mHxGo0GXbt2xbRp07B3714MGzYM06dPx6VLl9C7d280bNgQ69evx8GDB/Hpp58CQIUmZr/bxO9ZWVlQKpU4ePBgiTl7T58+jfnz55vvTZoBC81SFEqg2yzRKcgCdBqt6Ag27wu/4zjfv/pNZOwUHIwaL4+3+H7q16+P7OxsHDx4EAaDAfPmzcODDz6I6OhoXLt2zaz7uv/++6HX65GSkoLIyMgSvwIDA826r6pioVlSnY5A3V6iU5CZJTmpRUewC1PrHEJGl6aiY1hV4FtvQeHqarbt3bhxA506dcKKFStw7NgxxMfHY+3atZgzZw4eeeQRREZGorCwEAsXLsTFixfx7bff4osvvjDb/gEgOjoagwcPxpAhQ/DDDz8gPj4eBw4cwOzZs7Flyxaz7quqWGiW1mseoCk9eTPZr2T+rTHZmGYnUNS4nugYVuH5yMPQtm1j1m1qtVo88MAD+Oijj9CuXTs0aNAA06ZNw3PPPYdPPvkEjRo1wocffoj3338fDRo0wHfffYfZs2ebNQMALFmyBEOGDMErr7yCunXrom/fvvjnn38QGmpbo1o5ObE1HF4B/DhadAoyk2cbd8GBjHOiY9iNQL0WC1drIV++IjqKxSh9fBCxZTNU3hzhKRI/a1rD/U8BkV1EpyAzSSrMFB3BruiUWXj3cUDyctwzFQFTX2eZ2QAWmrX0mQ84V+8Z2h2BDAnJebypuqKOqHX49ulgSGrHu/6o7dABnr14rdwWsNCsxbMm0HWm6BRURTe1figwmD4cmv7fJu15/Pm0Yz3tWqHVInDGdNEx6H9YaNbU7BmgdnvRKagKdB4BoiPYtU9qHEN8P8d5zFKNVybAycaGrldnLDRre+QTQM37mOyVzo3XSapqcvQh3Op4v+gYVebSrCm8nnhCdAy6DQvN2rxCga5viU5BlZSkMd89RtXZ2BZnoI+tKzpGpUlqNYJmvg2pms6GYqtYaCI0H8Hnptkpncrp3ivRPeUoCjGxRyqkmsGio1RKjVcmwDmitugYdAcWmiiPfAp48y+EvdFJetERHMZVZSbmDFBB8rCv0b8evXrBZ+hQ0TGoDCw0UTSewIBlgEojOglVQJI+V3QEh/KP8zWsHFITUKlERzGJc926CHrnbdExqBwsNJGCGgHd3xOdgipAV8Cbqs3tB/dz2Pd0I9Ex7knh6YmanyyEwsVFdBQqBwtNtGbPAA0Hik5BJihSqHCdT6q2iI8CjyLxURsezq9QIOSDuVDXqiU6Cd0FC80W9P4I8L9PdAq6hxTPIBhkg+gYDmti3UPIbtdYdIwy+Y8dA23btqJj0D2w0GyB2g0YsBxwchOdhO5C5+4vOoJDkyVgbKtzMMREiY5SgrZzZ/i++KLoGGQCFpqt8K8LPLxAdAq6iyRX+xqNZ4+ypAJM7pkGKdg2Zt9Q166N4Pff4/1mdoKFZktiHwc6TBGdgsqhU3MwgDVcVqXjgyecIbmLnVFH4eaGmp8shFLLmX3sBQvN1nR4DWg8WHQKKoNOpRQdodrY73wVa4eECx3OHzR7Fpzr1BG2f6o4Fpot6jMfiOggOgXdQScXio5QrazxOIN/BjcWsm/f556Dx0MPCdk3VR4LzRYpnYAB3wI1YkQnodvo9DmiI1Q7c4OP4NrDza26T2379vB/ebxV90nmwUKzVRoPYPBawN0+57pzRLr8dNERqqVXYo4gt7V1brx2ffBBhCyYD0nBfxrtEf+v2TLPEGDwGkDtLjpJtZerdkV6QYboGNWSHjLGtY2DfJ9lr2e5NG2KWp99CoWzs0X3Q5bDQrN1gbHAgKWAwj7munNUOs8g0RGqtQwpD1P6ZEIKqGGR7WtiY1Fr0RdQuPLxQPaMhWYPIrsAfb8AJI6yEyVJ6ys6QrV3UZWG+YPcILmZdwIC5/vuQ+jXX3F4vgNgodmLhv2BR1lqoiS78KZqW7Bbk4iNQyIApXn+Hqjr1EHoN4uh9PQ0y/ZILBaaPWk4gKUmiE7N6yq24juv0zg8qEmVt+MUForQJd9A5eNjhlRkC1ho9oalJkQS/6bYlNk1DyO5V+WH8zsFByNsyRI41bDMNTkSg39N7ZGx1Pi/z1p0coHoCHSH8Q2PIP/B2Aq/TlWjBkKXLoFTMG+JcTT8F9FeNRwAPLqIpWYlusIs0RHoDnrIGNM+HoiubfJrlL6+CF26BOrQUAsmI1H4r6E9Y6lZjY4P9rRJGYo8TH0kB1INv3uuq/TyQug338A5IsIKyUgE/kto7xoOAPovBVQa0UkcVoarN3KLckXHoHKcV93AZ4O8ILmU/zQEp5AQhH3/HTR1o62YjKyNheYI6j8CDNkEuHC0liXoPHhTta3b4XIJm4dGA2VMWaWpXx/hq1byyKwaYKE5itAHgBHbAG/TryeQaZK03qIjkAmWeZ/EiSealljm1rYtwr5dDpU/nzZeHbDQHIlvneJSC2kmOolD0WnMOzMFWc7MsMO43qP4z79nv8dQ6/PPoDDzzCJku1hojsbNDxi2Gbivt+gkDiNJ5SQ6AlXAS42Ow/mtSQh+911IAh8QStbHQnNETi7Fz1Nr8bzoJA5Bp5BFRyATOSud8U67WYgY+IzoKCQAC81RKRRAz7nAQ+9yVpEq0hnyREcgE/hqfLG422L0jOgpOgoJwkJzdK3GAE+tB1w5W3xl6QoyRUege2jk3wireq9CI3/rPAiUbBMLrTqo0xF44S8guOoTulY3BkmBlDzeVG3Lnqr3FJZ0X4JAt0DRUUgwSZZlXiCoLooKgF8mAQeXiE5iN1I8g9DZh4NCbJGbkxtmtJqB7uHdRUchG8EjtOpEpQb6fAz0Wwyo3UWnsQtJ7pyN3RZFekViZa+VLDMqgYVWHcU+DrzwJxBY8ZnKqxudm5foCHSHPhF98F3P71Dbk5MIUEkstOrKtw4wYjvQ/DnRSWyazrn8+QHJujydPTG3/VzMajsLrk6uouOQDWKhVWcqZ6DXB8DTGwFPPk6jLDremGsTWoe0xoaHN/AUI90VC42KR0GO2gc0HwFAEp3GpuigFx2hWnNRuWDag9PwRZcv4O/K+Rjp7lhoVMxZC/SaBwz9CfAOF53GZiTpc0RHqLYa+TfCuj7rMKDuANFRyE5w2D6VVpADbJ8JHFgEyAbRaYTqcF8j3MhPEx2jWtE6aTHm/jF4ou4TUCo4yw2ZjoVG5Uv4G/hxNHDjgugkQhQq1WgaGgQZ/CtiLd3Cu2Fy88k8vUiVwkKjuyvMBXZ/DOxdABRWr9Nvib5h6OnBvx7WUMu9FqY+MBWtQ1qLjkJ2jIVGpsm4CmybARxfC1STI5Z/wptjuJQsOoZDc1I44ZkGz+D5hs/DWeksOg7ZOQ4KIdN4hgD9viq+d61mC9FprCLJhbOpWIoECd3Du+PHvj9i7P1jWWZkFrzJhiqmZlNgxFbg+LriI7aMRNGJLEan1gB8cozZNQ1oileavoJYf85UQ+bFQqPKiX0cuK8XsPcTYPdHQGG26ERmp1PynjxzivCMwMtNX0aHWh1ERyEHxUKjynNyAdq/CjQZUjxo5N8lDlVsSXKh6AgOIdAtEM83fB6PRT7GYfhkURwUQuaTcxP4+zNg/5dAfoboNFX2aGxrXMhy3FOqllbLvRaebfAsHq7zMJyUfAQPWR4LjcwvLxM48GVxueXcEJ2m0lpF18etwizRMexOhGcERsSOQM/aPXlERlbFQiPLKcgpfpjo3oXArSTRaSokS+OBlkFeomPYlXo+9TAidgS6hnWFJPH6I1kfC40srygfOPI98M9iIPm46DQmuRBQF4+65oqOYfNUkgodQzviyfueRPPA5qLjUDXHQiPruvJv8eCRkz/Y9Mwju+u0xEjDVdExbJavxhf9ovthQPQABLgFiI5DBICFRqLkpgPHVheXW+pp0WlKWRvTFTNzzoqOYXMa+TfCE/c9gW5h3TjQg2wOh+2TGC5ewAMvFP9K+Lu42E5tBIps405mnZNadASbEaINQa+IXugd0Ru1PWuLjkNULhYaiRf6YPGvHu8BZ7YAp34ELu4E9AXCIukU1fvEhbvaHd3Cu6F3RG80qdGEgzzILvCUI9mmvAzg7C/AqU1A3HarH7k927gLDmScs+o+RXNzckPbkLZ4KPwhtK/ZHmolj1LJvrDQyPblZwHnfgVObwLOb7XKYJJeMQ8gIce+bjWojBouNdCuVjt0qtUJDwY9yOtiZNdYaGRfCnKA+L+AS7uK/5t8wuxP1ZYhoXlkJPL1+Wbdri1QKVSI9YtFq+BWaF+zPer51hMdichsWGhk33LTgEt7/ldwu4CUU6jq89puaP3Rwd/FPPkEUylUaODbAM0Dm6NZYDPcX+N+uKgc470R3YmFRo4l+0ZxuV3aBVw7DKScrvApypPBDfCEc6aFAlqWj8YH9XzrIcY3Bk0DmrLAqFphoZFjMxiAm3GA7njx6UndieL/ZpZ/0/T2qLYYX3TZiiErJ8A1APV866G+T33U862Hej71eJMzVWsctk+OTaEA/KKKfzV47P+X59wsLrmUU8DNeCA9AUi/DKQnQKdxA2xkTmIfjQ9C3UMR6hGKUPdQhHmEGX+vVWtFxyOyKTxCI7pDdl46ruQkIyUnBSk5Kbieex2ZBZm4VXDL+Mv4deEtFOgLoDfoYZAN0Mt6yHdcw5MgQaPSQKPUQKPSwEXlYvzaw9kDvhpf+Ln4wdfFt8Tv/Vz84ObkJuinQGR/WGhEZibLMvRyccHJkKFWqHljMpEVsNCIiMghKEQHICL7s3TpUnh5eYmOQVQCC42IyjRs2DBIkgRJkqBWqxEZGYmZM2eiqKjIIvu7dOkSJEnCkSNHLLJ9cnwc5UhE5erevTuWLFmC/Px8/Pzzzxg9ejScnJwQFBQkOhpRKTxCI6JyOTs7IzAwEGFhYRg5ciS6dOmCTZs2lVovLi4OjzzyCAICAqDVatG8eXNs27atxDrh4eGYNWsWhg8fDnd3d4SGhuLLL780fr927eJH09x///2QJAkdOnSw6Hsjx8NCIyKTubi4oKCg9GN9srKy0LNnT2zfvh2HDx9G9+7d0adPHyQkJJRYb968eWjWrBkOHz6MUaNGYeTIkTh7tvhBqgcOHAAAbNu2DUlJSfjhhx8s/4bIobDQiOieZFnGtm3b8Ntvv6FTp06lvt+oUSO88MILaNCgAaKiovD222+jTp06pY7mevbsiVGjRiEyMhKTJ0+Gn58fduzYAQDw9/cHAPj6+iIwMBA+Pj6Wf2PkUFhoRFSuzZs3Q6vVQqPRoEePHhg4cCBmzJhRar2srCxMnDgR9erVg5eXF7RaLU6fPl3qCK1hw4bG30uShMDAQKSkpFj6bVA1wUEhRFSujh074vPPP4darUZwcDBUqrL/yZg4cSK2bt2KDz74AJGRkXBxccHjjz9e6vSkk1PJ561JkgSDwbyP/6Hqi4VGROVyc3NDZGTkPdfbs2cPhg0bhkcffRRA8RHbpUuXKrQvtbr4Cdl6vb7COYkAnnIkIjOIiorCDz/8gCNHjuDo0aMYNGhQhY+8atSoARcXF/z6669ITk5GRkaGhdKSo2KhEVGVffjhh/D29karVq3Qp08fdOvWDU2aNKnQNlQqFRYsWIBFixYhODgYjzzyiIXSkqPiXI5EROQQeIRGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQOgYVGREQO4f8ANSftMyOUmSsAAAAASUVORK5CYII=\n"
          },
          "metadata": {}
        }
      ],
      "source": [
        "empo_2_freqs.plot.pie(title='EMP ontology, level 2');"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "37946f1b-2361-4343-a64b-f527a99eb24e",
      "metadata": {
        "id": "37946f1b-2361-4343-a64b-f527a99eb24e"
      },
      "source": [
        "## Amplicon sequencing"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "fe77f75a-1497-407f-92fe-499e20aa5ef8",
      "metadata": {
        "id": "fe77f75a-1497-407f-92fe-499e20aa5ef8"
      },
      "source": [
        "**[16S rRNA](https://en.wikipedia.org/wiki/16S_ribosomal_RNA) [amplicon](https://en.wikipedia.org/wiki/Amplicon) sequencing** is the classical approach for studying microbiomes. It usually involves high-throughput DNA sequencing of one or several variable regions of the 16S ribosomal RNA gene, which consitutes the backbone of the small subunit of the **prokaryotic** ribosome, and is universally present in all species. Since as early as 1977, researchers have been using 16S rRNA sequencing to identify unculturable microorganisms in the environment. Today, 16S rRNA sequencing remains an affordable, reliable and sensitive method for profiling the composition of microbial communities.\n",
        "\n",
        "- The EMP500 study also generated **18S rRNA** (for eukaryotes) and internal transcribed spacer (**ITS**, for fungi) amplicon sequencing data. They are not included in this demo dataset, but are available from Qiita in case you want to further practice the skills."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "7d65b4a5-a704-4742-ae2a-8f224ec52754",
      "metadata": {
        "id": "7d65b4a5-a704-4742-ae2a-8f224ec52754"
      },
      "source": [
        "The 16S rRNA data are provided under the `amplicon` directory."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "id": "2e7f3fbe-e0c2-4c24-a3b7-e6f9719493bb",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2e7f3fbe-e0c2-4c24-a3b7-e6f9719493bb",
        "outputId": "f502fc93-a5da-47b9-fedf-0571a1d3eff3"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "16s.biom  16s.fa  16s.nwk  16s.tax  16s.tsv  subset\n"
          ]
        }
      ],
      "source": [
        "!ls $HOME/amplicon"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "c949358a-881b-47d3-b20b-71472a3dd885",
      "metadata": {
        "id": "c949358a-881b-47d3-b20b-71472a3dd885"
      },
      "source": [
        "### Sequences"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "7b54413d-e67b-4164-bfc4-df321ed7fbc0",
      "metadata": {
        "id": "7b54413d-e67b-4164-bfc4-df321ed7fbc0"
      },
      "source": [
        "The processed sequencing data are stored in `16s.fa`. This is a [FASTA](https://en.wikipedia.org/wiki/FASTA_format)-formatted file."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 19,
      "id": "ea7d7001-c1d7-4c96-9c49-13d44862b2d7",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ea7d7001-c1d7-4c96-9c49-13d44862b2d7",
        "outputId": "d6c722eb-34c3-4779-902b-8073d8e37eda"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            ">00006499\n",
            "AAACCAGCACCTCAAGTGGTCAGGAGGATTATTGGGCCTAAAGCATCCGTAGCCGGCTCTGTAAGTTTTCGGTTAAATCCATGTGCTCAACGCATGGGCTGCCGAAAATACTGCATAGCTAGGGAGTGGGAGAGGTAGACGGTACTCAGT\n",
            ">00012890\n",
            "AAAGTAGGAGACGAGCGTTATCCGGATTTACTGGGCGTAAAGCGTGTGCAGGCGGTTCGGCAAGTTGGATGTGAAATCTCCCGGCTCAACTGGGAGGGGTCGTTCAATACTACCAAACTTGAGGGCATCAGAGGAAGGTGGAATTCCGGG\n",
            ">00012904\n",
            "AAAGTAGGATCCGAGCGTTATCCGAATTTACTGGGCGTAAAGCGCGTGCAGGCGGTTCGGTAAGTTGGATGTGAAAGCTCCCGGCTCAACTGGGAGAGGTCGTTCAATACTGCCGGACTCGAGGATGGTAGAGGGAGGTGGAATTCCCGG\n",
            ">00024522\n",
            "AACAGAACCCCCAAGCGTTATCCGGATTTATTGGGCGTAAAGGGTCCGTAGGTGGTTAAGTTAGTCCAAAGTCAAATACTTCAGGCTCAACCTGAGGGACGCTTTGGAGACGGCTTAACTTGAGGGCATTAGGGGTTGATGGAACGGTCG\n",
            ">00024634\n",
            "AACAGAACCCGCAAGCATTATCCGGATTTATTGGGCGTAAAGGGTCCGTAGGTGGTCGGGAAAGTCTAAAGTCAAATACCTTGAGCTTAACTTAAGGGACGCTTTAGAAACTCCCTGACTTGAGGGCGTTAAAGGCTGATAGAACGGTCG\n"
          ]
        }
      ],
      "source": [
        "!head -n10 $HOME/amplicon/16s.fa"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "8da35de6-b1a8-4362-840d-e325ccec5aa6",
      "metadata": {
        "id": "8da35de6-b1a8-4362-840d-e325ccec5aa6"
      },
      "source": [
        "The method to work with sequence data in scikit-bio was introduced in section 01. For example, let's read the first sequence from the file:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 20,
      "id": "38fce90c-50a6-4ed5-aca5-b4c7d5d1661d",
      "metadata": {
        "id": "38fce90c-50a6-4ed5-aca5-b4c7d5d1661d"
      },
      "outputs": [],
      "source": [
        "from skbio import DNA"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 21,
      "id": "6a3c8f6f-2581-43bc-8700-bc4323e1f4c7",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "6a3c8f6f-2581-43bc-8700-bc4323e1f4c7",
        "outputId": "4600c276-1d24-4403-c9fd-7297f7cdadef"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "DNA\n",
              "---------------------------------------------------------------------\n",
              "Metadata:\n",
              "    'description': ''\n",
              "    'id': '00006499'\n",
              "Stats:\n",
              "    length: 150\n",
              "    has gaps: False\n",
              "    has degenerates: False\n",
              "    has definites: True\n",
              "    GC-content: 50.67%\n",
              "---------------------------------------------------------------------\n",
              "0   AAACCAGCAC CTCAAGTGGT CAGGAGGATT ATTGGGCCTA AAGCATCCGT AGCCGGCTCT\n",
              "60  GTAAGTTTTC GGTTAAATCC ATGTGCTCAA CGCATGGGCT GCCGAAAATA CTGCATAGCT\n",
              "120 AGGGAGTGGG AGAGGTAGAC GGTACTCAGT"
            ]
          },
          "metadata": {},
          "execution_count": 21
        }
      ],
      "source": [
        "DNA.read(f'{HOME}/amplicon/16s.fa')"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "a1e72b6d-29fc-4613-9e10-d93058e3a846",
      "metadata": {
        "id": "a1e72b6d-29fc-4613-9e10-d93058e3a846"
      },
      "source": [
        "Each sequence is 150 bp in length, which is limited by the sequencing platform (Illumina). It covers a portion of the 16S rRNA V4 region (appr. 252 bp), which has been shown effective and widely adopted in discerning microorganisms. The standard protocol for generating 16S rRNA V4 amplicons is provided at the [EMP website](https://earthmicrobiome.org/protocols-and-standards/16s/)."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "0d44991f-a20f-4d17-b918-4cfa23ba9f12",
      "metadata": {
        "id": "0d44991f-a20f-4d17-b918-4cfa23ba9f12"
      },
      "source": [
        "The raw sequencing data were processed by [Deblur](https://github.com/biocore/deblur) to extract amplicon sequence variants ([ASVs](https://en.wikipedia.org/wiki/Amplicon_sequence_variant)), which represent putatively unique microorganisms in the community. The ASV method provides finer resolution than the traditional operational taxonomic unit ([OTU](https://en.wikipedia.org/wiki/Operational_taxonomic_unit)) method."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "c6ae089d-a76d-4049-984b-903db606e0dc",
      "metadata": {
        "id": "c6ae089d-a76d-4049-984b-903db606e0dc"
      },
      "source": [
        "The ASVs were then mapped against the **Greengenes 2** (GG2) reference database ([McDonald et al., 2024](https://www.nature.com/articles/s41587-023-01845-1)). The IDs of the ASVs correspond to IDs of reference sequences in the GG2 databases. You can access the database files at:\n",
        "\n",
        "- https://ftp.microbio.me/greengenes_release/2022.10/"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "a51eb06c-56e6-49c1-b18f-6353cf49b61e",
      "metadata": {
        "id": "a51eb06c-56e6-49c1-b18f-6353cf49b61e"
      },
      "source": [
        "### Data table"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "0c5f079b-b0df-4133-86b9-dabbaf872c29",
      "metadata": {
        "id": "0c5f079b-b0df-4133-86b9-dabbaf872c29"
      },
      "source": [
        "The abundances (i.e., counts) of individual ASVs in indivdual samples are stored in a data table, as provided by the file `16s.biom`. The extension name `.biom` suggests that it is in the Biological Observation Matrix Format ([**BIOM**](https://biom-format.org/)), a computationally efficient format for storing omic data (we will explain why later).\n",
        "\n",
        "scikit-bio has integrated BIOM in its [`Table`](https://scikit.bio/docs/latest/generated/skbio.table.Table.html) module. Therefore we can directly read a BIOM table."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 22,
      "id": "67ecf7e4-33b7-4571-8334-5ed111aaf590",
      "metadata": {
        "id": "67ecf7e4-33b7-4571-8334-5ed111aaf590"
      },
      "outputs": [],
      "source": [
        "from skbio import Table"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 23,
      "id": "560bcadc-b71c-4fc1-bea0-fce33760b02f",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "560bcadc-b71c-4fc1-bea0-fce33760b02f",
        "outputId": "b22ea396-1b12-4a03-8eb0-5b8e667703f7"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "427269 x 805 <class 'biom.table.Table'> with 847146 nonzero entries (0% dense)"
            ]
          },
          "metadata": {},
          "execution_count": 23
        }
      ],
      "source": [
        "table = Table.read(f'{HOME}/amplicon/16s.biom')\n",
        "table"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "5377b363-30cd-425e-89ba-ee2f8cb8dd4f",
      "metadata": {
        "id": "5377b363-30cd-425e-89ba-ee2f8cb8dd4f"
      },
      "source": [
        "As shown, the dimensionality of this table is 427269 x 805 (you can also get this by `table.shape`). This means 805 samples and 427,269 observations (a.k.a., features)."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "a4dfdc57-52ee-4ba0-b7f3-da13186d263b",
      "metadata": {
        "id": "a4dfdc57-52ee-4ba0-b7f3-da13186d263b"
      },
      "source": [
        "The sample IDs are consistent with the sample IDs in the metadata."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 24,
      "id": "7e01e815-2633-43fb-b63f-2e76d6d954f0",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7e01e815-2633-43fb-b63f-2e76d6d954f0",
        "outputId": "beec6b26-e125-4b65-8479-26878dce0790"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['13114.song.51.s009', '13114.minich.76.s006',\n",
              "       '13114.stegen.38.s015', '13114.song.52.s002',\n",
              "       '13114.mousseau.88.s056', '13114.mcmahon.21.s002',\n",
              "       '13114.myrold.5.s008', '13114.song.51.s012', '13114.king.27.s017',\n",
              "       '13114.stegen.36.s004'], dtype='<U48')"
            ]
          },
          "metadata": {},
          "execution_count": 24
        }
      ],
      "source": [
        "table.ids('sample')[:10]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 25,
      "id": "85822df0-d629-46d4-977b-cc143d29b2bb",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "85822df0-d629-46d4-977b-cc143d29b2bb",
        "outputId": "88e3acdd-7099-4430-bc03-9b4e0f92cc89"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "True"
            ]
          },
          "metadata": {},
          "execution_count": 25
        }
      ],
      "source": [
        "set(table.ids('sample')).issubset(meta.ids)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "22f40f93-067a-4be0-8645-66fe63c33c99",
      "metadata": {
        "id": "22f40f93-067a-4be0-8645-66fe63c33c99"
      },
      "source": [
        "Not all sample IDs in the metadata (_n_ = 880) are found in this data table (_n_ = 805). This difference can be attributed to sample loss, insufficient biological materials, inadequate sequencing yield or quality, etc. This is common and one should be cautious about it, especially during multi-omic integration."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "84cb2e67-4dcf-4a7d-8a57-a99805190c38",
      "metadata": {
        "id": "84cb2e67-4dcf-4a7d-8a57-a99805190c38"
      },
      "source": [
        "The observation IDs are units of microbial components in the community. As discussed above, each observation represents an ASV."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 26,
      "id": "536beadc-2ec0-41f9-8490-3a3d8d6b3f98",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "536beadc-2ec0-41f9-8490-3a3d8d6b3f98",
        "outputId": "3d2c1089-e3a6-4eef-e98e-f9ca6a7a42c5"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['08665176', '13062664', '03483654', '21781706', '07176695',\n",
              "       '18439252', '08434208', '07654326', '22343722', '13688331'],\n",
              "      dtype='<U8')"
            ]
          },
          "metadata": {},
          "execution_count": 26
        }
      ],
      "source": [
        "table.ids('observation')[:10]"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "8f2db1a2-62d1-4956-8a73-c2ddd0b257bf",
      "metadata": {
        "id": "8f2db1a2-62d1-4956-8a73-c2ddd0b257bf"
      },
      "source": [
        "In data science, a more generic term \"[**feature**](https://en.wikipedia.org/wiki/Feature_(machine_learning)\" is typically used to describe all independent variables, regardless of OTUs, ASVs, genes, metabolites, etc. A data table of samples by features like this one is usually referred to as a **feature table**."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "3cef063b-90ac-4162-b549-a0857ab545aa",
      "metadata": {
        "id": "3cef063b-90ac-4162-b549-a0857ab545aa"
      },
      "source": [
        "### Sparse data"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "9d1e81c8-bee7-4c85-8c20-2681791adf36",
      "metadata": {
        "id": "9d1e81c8-bee7-4c85-8c20-2681791adf36"
      },
      "source": [
        "A BIOM table can be converted into a Pandas dataframe. This allows one to take a quick look at the table content. What did you see?"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 27,
      "id": "d4783bf6-8d74-4307-b6e0-aec008bffddb",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 256
        },
        "id": "d4783bf6-8d74-4307-b6e0-aec008bffddb",
        "outputId": "0a5a4512-e7a3-41b9-e509-50c43b04e808"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "          13114.song.51.s009  13114.minich.76.s006  13114.stegen.38.s015  \\\n",
              "08665176                   0                     0                     0   \n",
              "13062664                   0                     0                     0   \n",
              "03483654                   0                     0                     0   \n",
              "21781706                   0                     0                     0   \n",
              "07176695                   0                     0                     0   \n",
              "\n",
              "          13114.song.52.s002  13114.mousseau.88.s056  13114.mcmahon.21.s002  \\\n",
              "08665176                   0                       0                      0   \n",
              "13062664                   0                       0                      0   \n",
              "03483654                   0                       0                      0   \n",
              "21781706                   0                       0                      0   \n",
              "07176695                   0                       0                      0   \n",
              "\n",
              "          13114.myrold.5.s008  13114.song.51.s012  13114.king.27.s017  \\\n",
              "08665176                    0                   0                   0   \n",
              "13062664                    0                   0                   0   \n",
              "03483654                    0                   0                   0   \n",
              "21781706                    0                   0                   0   \n",
              "07176695                    0                   0                   0   \n",
              "\n",
              "          13114.stegen.36.s004  ...  13114.thomas.19.s061  \\\n",
              "08665176                     0  ...                     0   \n",
              "13062664                     0  ...                     0   \n",
              "03483654                     0  ...                     0   \n",
              "21781706                     0  ...                     0   \n",
              "07176695                     0  ...                     0   \n",
              "\n",
              "          13114.angenent.65.s007  13114.mousseau.88.s011  13114.pinto.63.s025  \\\n",
              "08665176                       0                       0                    0   \n",
              "13062664                       0                       0                    0   \n",
              "03483654                       0                       0                    0   \n",
              "21781706                       0                       0                    0   \n",
              "07176695                       0                       0                    0   \n",
              "\n",
              "          13114.song.52.s005  13114.mousseau.88.s062  \\\n",
              "08665176                   0                       0   \n",
              "13062664                   0                       0   \n",
              "03483654                   0                       0   \n",
              "21781706                   0                       0   \n",
              "07176695                   0                       0   \n",
              "\n",
              "          13114.roy.chowdhury.45.s010  13114.rohwer.85.s016  \\\n",
              "08665176                            0                     0   \n",
              "13062664                            0                     0   \n",
              "03483654                            0                     0   \n",
              "21781706                            0                     0   \n",
              "07176695                            0                     0   \n",
              "\n",
              "          13114.jensen.43.s011  13114.Kshtrika1.misc.600  \n",
              "08665176                     0                         0  \n",
              "13062664                     0                         0  \n",
              "03483654                     0                         0  \n",
              "21781706                     0                         0  \n",
              "07176695                     0                         0  \n",
              "\n",
              "[5 rows x 805 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-3c4a1413-74ce-4a3c-b175-4e6823a95691\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>13114.song.51.s009</th>\n",
              "      <th>13114.minich.76.s006</th>\n",
              "      <th>13114.stegen.38.s015</th>\n",
              "      <th>13114.song.52.s002</th>\n",
              "      <th>13114.mousseau.88.s056</th>\n",
              "      <th>13114.mcmahon.21.s002</th>\n",
              "      <th>13114.myrold.5.s008</th>\n",
              "      <th>13114.song.51.s012</th>\n",
              "      <th>13114.king.27.s017</th>\n",
              "      <th>13114.stegen.36.s004</th>\n",
              "      <th>...</th>\n",
              "      <th>13114.thomas.19.s061</th>\n",
              "      <th>13114.angenent.65.s007</th>\n",
              "      <th>13114.mousseau.88.s011</th>\n",
              "      <th>13114.pinto.63.s025</th>\n",
              "      <th>13114.song.52.s005</th>\n",
              "      <th>13114.mousseau.88.s062</th>\n",
              "      <th>13114.roy.chowdhury.45.s010</th>\n",
              "      <th>13114.rohwer.85.s016</th>\n",
              "      <th>13114.jensen.43.s011</th>\n",
              "      <th>13114.Kshtrika1.misc.600</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>08665176</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13062664</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>03483654</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21781706</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>07176695</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows × 805 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-3c4a1413-74ce-4a3c-b175-4e6823a95691')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-3c4a1413-74ce-4a3c-b175-4e6823a95691 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-3c4a1413-74ce-4a3c-b175-4e6823a95691');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-af148ab1-e5a0-4d57-a431-02d1aa75f502\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-af148ab1-e5a0-4d57-a431-02d1aa75f502')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-af148ab1-e5a0-4d57-a431-02d1aa75f502 button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "df"
            }
          },
          "metadata": {},
          "execution_count": 27
        }
      ],
      "source": [
        "df = table.to_dataframe()\n",
        "df.head()"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "cfc4da5a-d0de-4e5a-8775-c3f5c932b7c7",
      "metadata": {
        "id": "cfc4da5a-d0de-4e5a-8775-c3f5c932b7c7"
      },
      "source": [
        "Most cell values are zeros. This means that while there are more than 400k ASVs discovered across the sample collection, each sample only contains a very small subset of them. This indicates an important statistical property of biological omic data that is distinct from many other fields: [**sparsity**](https://en.wikipedia.org/wiki/Sparse_matrix)."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "12e472cc-fe26-40bc-8a58-b9fc9adc87d9",
      "metadata": {
        "id": "12e472cc-fe26-40bc-8a58-b9fc9adc87d9"
      },
      "source": [
        "One can let scikit-bio report the density (i.e., fraction of non-zero values) of the table:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 28,
      "id": "fa23ec0a-7e3b-4ed3-88f2-131c5d693dc2",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "fa23ec0a-7e3b-4ed3-88f2-131c5d693dc2",
        "outputId": "edbced60-d99f-4628-bfa7-08ba671fd5b8"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "0.0024629806503703887"
            ]
          },
          "metadata": {},
          "execution_count": 28
        }
      ],
      "source": [
        "table.get_table_density()"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "91c024f9-0712-40c3-8404-1da0855f6fc5",
      "metadata": {
        "id": "91c024f9-0712-40c3-8404-1da0855f6fc5"
      },
      "source": [
        "That is, only 0.25% of the cell values are non-zero."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 29,
      "id": "8d3f7a44-37b5-4742-acbc-c4e9e9b240b5",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8d3f7a44-37b5-4742-acbc-c4e9e9b240b5",
        "outputId": "ff026cbf-cf59-43d1-f272-bbd11aeac002"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array([3.6743e+04, 3.6377e+04, 1.5897e+04, 3.7726e+04, 4.1853e+04,\n",
              "       4.8000e+01, 6.0000e+00, 3.2423e+04, 1.0679e+04, 2.1607e+04])"
            ]
          },
          "metadata": {},
          "execution_count": 29
        }
      ],
      "source": [
        "sums = table.sum(axis='sample')\n",
        "sums[:10]"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "4bc051d9-9c1f-47dc-a455-016eac9477cd",
      "metadata": {
        "id": "4bc051d9-9c1f-47dc-a455-016eac9477cd"
      },
      "source": [
        "To efficiently store sparse data, one would avoid storing numerous zeros, but instead only store non-zero cell values and their coordinates. This is referred to as the [sparse matrix](https://en.wikipedia.org/wiki/Sparse_matrix). This is the underlying data structure of BIOM, as in contrast to the dense matrix as used in e.g. NumPy."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "4c935d06-e508-46ca-a82a-b81115254860",
      "metadata": {
        "id": "4c935d06-e508-46ca-a82a-b81115254860"
      },
      "source": [
        "If needed, you can convert the BIOM table into a 2D NumPy array (a dense matrix). To avoid stressing the computer (because the data table is huge!), the following code only converts the first 5 rows and columns of the table."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 30,
      "id": "dd3ec97d-7e7c-4d43-816d-c287541c7118",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "dd3ec97d-7e7c-4d43-816d-c287541c7118",
        "outputId": "b1f7f08f-462e-4b55-88a5-b4e09beb4e8b"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array([[0., 0., 0., 0., 0.],\n",
              "       [0., 0., 0., 0., 0.],\n",
              "       [0., 0., 0., 0., 0.],\n",
              "       [0., 0., 0., 0., 0.],\n",
              "       [0., 0., 0., 0., 0.]])"
            ]
          },
          "metadata": {},
          "execution_count": 30
        }
      ],
      "source": [
        "table.head().matrix_data.toarray()"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "5a963d24-e0ce-40e4-b1f8-40dc869e9f8f",
      "metadata": {
        "id": "5a963d24-e0ce-40e4-b1f8-40dc869e9f8f"
      },
      "source": [
        "### Taxonomy"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "7edacca7-d18f-4a91-9ad3-de34b28ddfe1",
      "metadata": {
        "id": "7edacca7-d18f-4a91-9ad3-de34b28ddfe1"
      },
      "source": [
        "As shown above, there are over 400k features (ASVs) but merely ~800 samples in the 16S rRNA amplicon data table. Such a large number of features (referred to as [**high-dimensionality**](https://en.wikipedia.org/wiki/High-dimensional_statistics) in data science) imposes significant challenges to the subsequent analysis (you may read about the \"[curse of dimensionality](https://en.wikipedia.org/wiki/Curse_of_dimensionality)\" and intepretation. While there are pure mathematical approaches to mitigate high-dimensionality, biological sciences provide unique, knowledge-based solutions to it.\n",
        "\n",
        "These features are not a plain list of entities. They are (presumably) microorganisms, and organisms can be classified. The classification of organisms, a.k.a., [**taxonomy**](https://en.wikipedia.org/wiki/Taxonomy), provides a hierarchical system with (usually) some 7-8 fixed ranks to describe the relationships among organisms."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "edcfa000-0584-4a06-ae28-c48cb9374a41",
      "metadata": {
        "id": "edcfa000-0584-4a06-ae28-c48cb9374a41"
      },
      "source": [
        "The taxonomic assignments of the ASVs according to the GG2 reference is provided in the file `16s.tax`. It reads like:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 31,
      "id": "c38932c1-4b75-4335-8bc1-d1abb5372939",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "c38932c1-4b75-4335-8bc1-d1abb5372939",
        "outputId": "a46751e4-19a5-498d-e2bf-619328d932c0"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "00006499\td__Archaea; p__Thermoproteota; c__Nitrososphaeria_A; o__Nitrososphaerales; f__Nitrosopumilaceae; g__Nitrosopelagicus; s__Nitrosopelagicus sp000402075\n",
            "00012890\td__Bacteria; p__Chloroflexota; c__Anaerolineae; o__Anaerolineales; f__EnvOPS12; g__UBA8950; s__UBA8950 sp001872455\n",
            "00012904\td__Bacteria; p__Chloroflexota; c__Anaerolineae; o__Anaerolineales; f__; g__; s__\n",
            "00024522\td__Bacteria; p__Patescibacteria; c__Paceibacteria; o__Paceibacterales; f__; g__; s__\n",
            "00024634\td__Bacteria; p__Patescibacteria; c__Paceibacteria; o__Paceibacterales; f__; g__; s__\n",
            "00024664\td__Bacteria; p__Patescibacteria; c__Paceibacteria; o__Paceibacterales; f__UBA5738; g__UBA5738; s__UBA5738 sp002423555\n",
            "00024725\td__Bacteria; p__Patescibacteria; c__Paceibacteria; o__Paceibacterales; f__UBA5738; g__UBA5738; s__UBA5738 sp002423555\n",
            "00024797\td__Bacteria; p__Patescibacteria; c__Paceibacteria; o__Paceibacterales; f__UBA10102; g__; s__\n",
            "00025031\td__Bacteria; p__Patescibacteria; c__Paceibacteria; o__Paceibacterales; f__; g__; s__\n",
            "00025034\td__Bacteria; p__Patescibacteria; c__Paceibacteria; o__Paceibacterales; f__; g__; s__\n"
          ]
        }
      ],
      "source": [
        "!head $HOME/amplicon/16s.tax"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "d8e9b1a3-17d8-4d0c-851c-fc937f8e0f08",
      "metadata": {
        "id": "d8e9b1a3-17d8-4d0c-851c-fc937f8e0f08"
      },
      "source": [
        "It is a mapping of ASV IDs to taxonomic **lineage** strings. Each lineage string consists of [seven ranks](https://en.wikipedia.org/wiki/Taxonomic_rank), as indicated by their initials. The seven ranks are (from high to low):"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 32,
      "id": "14fb8c73-dc56-4202-9e5e-7fff24d2a7d8",
      "metadata": {
        "id": "14fb8c73-dc56-4202-9e5e-7fff24d2a7d8"
      },
      "outputs": [],
      "source": [
        "ranks = ['Domain', 'Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species']"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "43bbfe7a-e75a-45a8-b27b-72051add963f",
      "metadata": {
        "id": "43bbfe7a-e75a-45a8-b27b-72051add963f"
      },
      "source": [
        "The following code can read lineage strings into a Pandas dataframe with columns representing individual ranks."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 33,
      "id": "eb911e23-1634-447f-a305-d9f25d1f0641",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        },
        "id": "eb911e23-1634-447f-a305-d9f25d1f0641",
        "outputId": "898ee34f-eb0b-4cb6-f3ff-a5bc34fa24d5"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "            Domain           Phylum              Class              Order  \\\n",
              "00006499   Archaea   Thermoproteota  Nitrososphaeria_A  Nitrososphaerales   \n",
              "00012890  Bacteria    Chloroflexota       Anaerolineae     Anaerolineales   \n",
              "00012904  Bacteria    Chloroflexota       Anaerolineae     Anaerolineales   \n",
              "00024522  Bacteria  Patescibacteria      Paceibacteria    Paceibacterales   \n",
              "00024634  Bacteria  Patescibacteria      Paceibacteria    Paceibacterales   \n",
              "\n",
              "                     Family             Genus                       Species  \n",
              "00006499  Nitrosopumilaceae  Nitrosopelagicus  Nitrosopelagicus sp000402075  \n",
              "00012890           EnvOPS12           UBA8950           UBA8950 sp001872455  \n",
              "00012904                NaN               NaN                           NaN  \n",
              "00024522                NaN               NaN                           NaN  \n",
              "00024634                NaN               NaN                           NaN  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-8a48bc12-2517-406f-a7e9-ec489b67a352\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Domain</th>\n",
              "      <th>Phylum</th>\n",
              "      <th>Class</th>\n",
              "      <th>Order</th>\n",
              "      <th>Family</th>\n",
              "      <th>Genus</th>\n",
              "      <th>Species</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>00006499</th>\n",
              "      <td>Archaea</td>\n",
              "      <td>Thermoproteota</td>\n",
              "      <td>Nitrososphaeria_A</td>\n",
              "      <td>Nitrososphaerales</td>\n",
              "      <td>Nitrosopumilaceae</td>\n",
              "      <td>Nitrosopelagicus</td>\n",
              "      <td>Nitrosopelagicus sp000402075</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>00012890</th>\n",
              "      <td>Bacteria</td>\n",
              "      <td>Chloroflexota</td>\n",
              "      <td>Anaerolineae</td>\n",
              "      <td>Anaerolineales</td>\n",
              "      <td>EnvOPS12</td>\n",
              "      <td>UBA8950</td>\n",
              "      <td>UBA8950 sp001872455</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>00012904</th>\n",
              "      <td>Bacteria</td>\n",
              "      <td>Chloroflexota</td>\n",
              "      <td>Anaerolineae</td>\n",
              "      <td>Anaerolineales</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>00024522</th>\n",
              "      <td>Bacteria</td>\n",
              "      <td>Patescibacteria</td>\n",
              "      <td>Paceibacteria</td>\n",
              "      <td>Paceibacterales</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>00024634</th>\n",
              "      <td>Bacteria</td>\n",
              "      <td>Patescibacteria</td>\n",
              "      <td>Paceibacteria</td>\n",
              "      <td>Paceibacterales</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-8a48bc12-2517-406f-a7e9-ec489b67a352')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-8a48bc12-2517-406f-a7e9-ec489b67a352 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-8a48bc12-2517-406f-a7e9-ec489b67a352');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-9229976f-4a71-4368-915c-72ffb88e6889\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-9229976f-4a71-4368-915c-72ffb88e6889')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-9229976f-4a71-4368-915c-72ffb88e6889 button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "taxonomy"
            }
          },
          "metadata": {},
          "execution_count": 33
        }
      ],
      "source": [
        "taxonomy = {}\n",
        "with open(f'{HOME}/amplicon/16s.tax', 'r') as f:\n",
        "    for line in f:\n",
        "        id_, lineage = line.rstrip().split('\\t')\n",
        "        ranked = [x.split('__')[1] for x in lineage.split('; ')]\n",
        "        taxonomy[id_] = ranked\n",
        "taxonomy = pd.DataFrame.from_dict(taxonomy, orient='index', columns=ranks).replace('', np.nan)\n",
        "taxonomy.head()"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "88987ea1-3a30-4d8c-b9a6-241b659f16b4",
      "metadata": {
        "id": "88987ea1-3a30-4d8c-b9a6-241b659f16b4"
      },
      "source": [
        "Given the taxonomy, one can collapse the data table to a specific taxonomic rank. The collapsing process will aggregate (sum) cell values that belong to the same taxonomic group. This not only facilitates biologically meaningful description of community composition, but also serves as a knowledge-guided method of dimensionality reduction. The following example converts the features into the phylum level."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 34,
      "id": "675ee769-e119-4e80-b543-1fb9def16b68",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "675ee769-e119-4e80-b543-1fb9def16b68",
        "outputId": "a76f958d-8b63-4a37-87fe-d17893b2ffab"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "124 x 805 <class 'biom.table.Table'> with 19786 nonzero entries (19% dense)"
            ]
          },
          "metadata": {},
          "execution_count": 34
        }
      ],
      "source": [
        "table_phylum = table.collapse(taxonomy['Phylum'].to_dict(), norm=False, axis='observation')\n",
        "table_phylum"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "9c5c1c36-f866-4ae7-af8d-2d18546c98e4",
      "metadata": {
        "id": "9c5c1c36-f866-4ae7-af8d-2d18546c98e4"
      },
      "source": [
        "After collapsing, the data table becomes:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 35,
      "id": "2fbab5e7-8b38-4058-b2a1-8947abd91f35",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 256
        },
        "id": "2fbab5e7-8b38-4058-b2a1-8947abd91f35",
        "outputId": "2ff6f9e8-562f-408d-d641-f784e0668af7"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                    13114.song.51.s009  13114.minich.76.s006  \\\n",
              "Proteobacteria                   916.0               12636.0   \n",
              "Desulfobacterota_C                   0                     0   \n",
              "Planctomycetota                      0                 315.0   \n",
              "Thermoplasmatota                     0                     0   \n",
              "Gemmatimonadota                      0                     0   \n",
              "\n",
              "                    13114.stegen.38.s015  13114.song.52.s002  \\\n",
              "Proteobacteria                    4400.0                32.0   \n",
              "Desulfobacterota_C                     0                   0   \n",
              "Planctomycetota                   1292.0              5802.0   \n",
              "Thermoplasmatota                       0                   0   \n",
              "Gemmatimonadota                      2.0                   0   \n",
              "\n",
              "                    13114.mousseau.88.s056  13114.mcmahon.21.s002  \\\n",
              "Proteobacteria                      2126.0                    6.0   \n",
              "Desulfobacterota_C                       0                      0   \n",
              "Planctomycetota                          0                      0   \n",
              "Thermoplasmatota                         0                      0   \n",
              "Gemmatimonadota                          0                      0   \n",
              "\n",
              "                    13114.myrold.5.s008  13114.song.51.s012  \\\n",
              "Proteobacteria                        0               206.0   \n",
              "Desulfobacterota_C                    0                   0   \n",
              "Planctomycetota                     2.0                   0   \n",
              "Thermoplasmatota                      0                   0   \n",
              "Gemmatimonadota                       0                   0   \n",
              "\n",
              "                    13114.king.27.s017  13114.stegen.36.s004  ...  \\\n",
              "Proteobacteria                  1079.0                4302.0  ...   \n",
              "Desulfobacterota_C                   0                  21.0  ...   \n",
              "Planctomycetota                      0                 721.0  ...   \n",
              "Thermoplasmatota                     0                   2.0  ...   \n",
              "Gemmatimonadota                      0                1038.0  ...   \n",
              "\n",
              "                    13114.thomas.19.s061  13114.angenent.65.s007  \\\n",
              "Proteobacteria                    8162.0                  1014.0   \n",
              "Desulfobacterota_C                     0                       0   \n",
              "Planctomycetota                    179.0                    24.0   \n",
              "Thermoplasmatota                       0                       0   \n",
              "Gemmatimonadota                        0                       0   \n",
              "\n",
              "                    13114.mousseau.88.s011  13114.pinto.63.s025  \\\n",
              "Proteobacteria                      6954.0              26091.0   \n",
              "Desulfobacterota_C                       0                    0   \n",
              "Planctomycetota                      230.0                730.0   \n",
              "Thermoplasmatota                       8.0                    0   \n",
              "Gemmatimonadota                       15.0                364.0   \n",
              "\n",
              "                    13114.song.52.s005  13114.mousseau.88.s062  \\\n",
              "Proteobacteria                    30.0                   139.0   \n",
              "Desulfobacterota_C                   0                       0   \n",
              "Planctomycetota                   99.0                    12.0   \n",
              "Thermoplasmatota                     0                       0   \n",
              "Gemmatimonadota                      0                       0   \n",
              "\n",
              "                    13114.roy.chowdhury.45.s010  13114.rohwer.85.s016  \\\n",
              "Proteobacteria                              2.0               15152.0   \n",
              "Desulfobacterota_C                            0                     0   \n",
              "Planctomycetota                               0                     0   \n",
              "Thermoplasmatota                              0                     0   \n",
              "Gemmatimonadota                               0                     0   \n",
              "\n",
              "                    13114.jensen.43.s011  13114.Kshtrika1.misc.600  \n",
              "Proteobacteria                    5473.0                         0  \n",
              "Desulfobacterota_C                  90.0                         0  \n",
              "Planctomycetota                   2679.0                         0  \n",
              "Thermoplasmatota                    31.0                         0  \n",
              "Gemmatimonadota                    177.0                         0  \n",
              "\n",
              "[5 rows x 805 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-d46886d7-0bfe-4ba7-8b03-9f334e6805a8\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>13114.song.51.s009</th>\n",
              "      <th>13114.minich.76.s006</th>\n",
              "      <th>13114.stegen.38.s015</th>\n",
              "      <th>13114.song.52.s002</th>\n",
              "      <th>13114.mousseau.88.s056</th>\n",
              "      <th>13114.mcmahon.21.s002</th>\n",
              "      <th>13114.myrold.5.s008</th>\n",
              "      <th>13114.song.51.s012</th>\n",
              "      <th>13114.king.27.s017</th>\n",
              "      <th>13114.stegen.36.s004</th>\n",
              "      <th>...</th>\n",
              "      <th>13114.thomas.19.s061</th>\n",
              "      <th>13114.angenent.65.s007</th>\n",
              "      <th>13114.mousseau.88.s011</th>\n",
              "      <th>13114.pinto.63.s025</th>\n",
              "      <th>13114.song.52.s005</th>\n",
              "      <th>13114.mousseau.88.s062</th>\n",
              "      <th>13114.roy.chowdhury.45.s010</th>\n",
              "      <th>13114.rohwer.85.s016</th>\n",
              "      <th>13114.jensen.43.s011</th>\n",
              "      <th>13114.Kshtrika1.misc.600</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>Proteobacteria</th>\n",
              "      <td>916.0</td>\n",
              "      <td>12636.0</td>\n",
              "      <td>4400.0</td>\n",
              "      <td>32.0</td>\n",
              "      <td>2126.0</td>\n",
              "      <td>6.0</td>\n",
              "      <td>0</td>\n",
              "      <td>206.0</td>\n",
              "      <td>1079.0</td>\n",
              "      <td>4302.0</td>\n",
              "      <td>...</td>\n",
              "      <td>8162.0</td>\n",
              "      <td>1014.0</td>\n",
              "      <td>6954.0</td>\n",
              "      <td>26091.0</td>\n",
              "      <td>30.0</td>\n",
              "      <td>139.0</td>\n",
              "      <td>2.0</td>\n",
              "      <td>15152.0</td>\n",
              "      <td>5473.0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Desulfobacterota_C</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>21.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>90.0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Planctomycetota</th>\n",
              "      <td>0</td>\n",
              "      <td>315.0</td>\n",
              "      <td>1292.0</td>\n",
              "      <td>5802.0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>2.0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>721.0</td>\n",
              "      <td>...</td>\n",
              "      <td>179.0</td>\n",
              "      <td>24.0</td>\n",
              "      <td>230.0</td>\n",
              "      <td>730.0</td>\n",
              "      <td>99.0</td>\n",
              "      <td>12.0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>2679.0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Thermoplasmatota</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>2.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>8.0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>31.0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Gemmatimonadota</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>2.0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1038.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>15.0</td>\n",
              "      <td>364.0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>177.0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows × 805 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-d46886d7-0bfe-4ba7-8b03-9f334e6805a8')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-d46886d7-0bfe-4ba7-8b03-9f334e6805a8 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-d46886d7-0bfe-4ba7-8b03-9f334e6805a8');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-c4e80d72-9a6c-44ce-a175-8cceb66b835b\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-c4e80d72-9a6c-44ce-a175-8cceb66b835b')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-c4e80d72-9a6c-44ce-a175-8cceb66b835b button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe"
            }
          },
          "metadata": {},
          "execution_count": 35
        }
      ],
      "source": [
        "table_phylum.to_dataframe().head()"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "5e525c7c-4893-4be6-97bd-89acab2212fc",
      "metadata": {
        "id": "5e525c7c-4893-4be6-97bd-89acab2212fc"
      },
      "source": [
        "Note: If you work with a Pandas dataframe instead of a BIOM table, you can achieve this using the `groupby` function."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 36,
      "id": "ae3378e2-4e20-41de-bf1f-bcd5aa0d7eee",
      "metadata": {
        "id": "ae3378e2-4e20-41de-bf1f-bcd5aa0d7eee"
      },
      "outputs": [],
      "source": [
        "# table.to_dataframe().groupby(taxonomy['Phylum']).sum()"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "782f3746-b960-47ec-9218-d6e675b2a4e0",
      "metadata": {
        "id": "782f3746-b960-47ec-9218-d6e675b2a4e0"
      },
      "source": [
        "It should be noted that 16S rRNA sequencing, particularly V4 region sequencing, can reach a taxonomic resolution as low as **genus**. Despite that the taxonomic assignment also contains species, it is usually considered less reliable to base the analysis on a species-level profile inferred from 16S rRNA data."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "c14592b6-6545-41ea-8cab-5b032e0d7f79",
      "metadata": {
        "id": "c14592b6-6545-41ea-8cab-5b032e0d7f79"
      },
      "source": [
        "However, the emerging technique of **full-length rRNA operon sequencing** largely overcomes this limitation by targetting the entire 16S rRNA gene (about 1.5 kb) or even the entire rRNA operon (about 4.5 kb) with the [long-read sequencing technology](https://en.wikipedia.org/wiki/Third-generation_sequencing). The EMP500 study also generated full-length rRNA operon data, despite not included in this demo dataset. You may consider exploring it as a practice after the tutorial."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "a59bd0b9-7628-43e5-bf8d-35c60d1ed833",
      "metadata": {
        "id": "a59bd0b9-7628-43e5-bf8d-35c60d1ed833"
      },
      "source": [
        "## Shotgun metagenomics"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "90bb1253-c6fe-4297-aa09-e7b679c42bae",
      "metadata": {
        "id": "90bb1253-c6fe-4297-aa09-e7b679c42bae"
      },
      "source": [
        "[**Shotgun sequencing**](https://en.wikipedia.org/wiki/Shotgun_sequencing) randomly sequences all DNA fragments in a biological sample, rather than targeting a specific DNA region as amplicon sequencing does. [Shotgun metagenomics](https://en.wikipedia.org/wiki/Shotgun_sequencing#Metagenomic_shotgun_sequencing), or whole-metagenomic shotgun (WMS) sequencing, or simply \"[metagenomics](https://en.wikipedia.org/wiki/Metagenomics)\", is the application of shotgun sequencing to recover DNA sequences from all microorganisms in a habitat.\n",
        "\n",
        "Compared with 16S rRNA sequencing, shotgun metagenomics delivers higher resolution (usually down to species- or even strain-level), informs function in addition to composition, and includes all domains of life (including viruses and eukaryotes) rather than just prokaryotes, despite being more expensive, and more challenging in bioinformatic analysis.\n",
        "\n",
        "There are two strategies to analyze shotgun metagenomic data:\n",
        "\n",
        "1) **Read-based**, which usually involves mapping sequencing reads against a reference database to inform the composition of a metagenome.\n",
        "2) **Assembly-based**, which performs assembly of sequencing reads to recover the original microbial genomes in a (usually) reference-free way."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "ba81de71-4ee9-48d9-89ab-8b91aaa04515",
      "metadata": {
        "id": "ba81de71-4ee9-48d9-89ab-8b91aaa04515"
      },
      "source": [
        "The shotgun metagenomic data processed using the read-based method is provided within the `shotgun` directory."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 37,
      "id": "9ad02aa3-9c22-4414-8104-067311cb711d",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "9ad02aa3-9c22-4414-8104-067311cb711d",
        "outputId": "37ef8a47-f8a5-402b-c9c9-9a864b5c3087"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "ec.biom  ko.name   ogu.info  ogu.tax\t   pathway.name\n",
            "ko.biom  ogu.biom  ogu.nwk   pathway.biom\n"
          ]
        }
      ],
      "source": [
        "!ls $HOME/shotgun"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "bd917cb2-cce0-47d0-8956-6e5d741013d0",
      "metadata": {
        "id": "bd917cb2-cce0-47d0-8956-6e5d741013d0"
      },
      "source": [
        "The basic data table is `ogu.biom`. This table was generated by aligning sequencing reads against the Web of Life 2 (**WoL2**) reference genome database ([Zhu et al., 2019](https://www.nature.com/articles/s41467-019-13443-4)), and quantifying the number of reads mapped to each reference genome. The database files are available at:\n",
        "\n",
        "- https://ftp.microbio.me/pub/wol2/"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "f64c34b0-560b-4159-8b23-af94a978742a",
      "metadata": {
        "id": "f64c34b0-560b-4159-8b23-af94a978742a"
      },
      "source": [
        "Read the BIOM table. It contains 801 samples and 14,933 features."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 38,
      "id": "491a3f24-a25e-4a6e-9613-101b4e3ea680",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "491a3f24-a25e-4a6e-9613-101b4e3ea680",
        "outputId": "335a3cc9-b824-433d-fcbd-e673e5fbcc74"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "14933 x 801 <class 'biom.table.Table'> with 1344863 nonzero entries (11% dense)"
            ]
          },
          "metadata": {},
          "execution_count": 38
        }
      ],
      "source": [
        "table = Table.read(f'{HOME}/shotgun/ogu.biom')\n",
        "table"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "2c770e48-0751-4ec1-adf0-1f57b1e3434e",
      "metadata": {
        "id": "2c770e48-0751-4ec1-adf0-1f57b1e3434e"
      },
      "source": [
        "Each feature represents a reference genome. They are referred to as the operational genomic units (**OGUs**) ([Zhu et al., 2022](https://journals.asm.org/doi/full/10.1128/msystems.00167-22)), a term indicating that they are not the exact genomes that are present in the sample, but are reference genomes to which sequencing data can be mapped to. The true members of the microbial community may be closely related to those reference strains."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 39,
      "id": "8c93ee4b-d476-4b25-b8bc-3aa4ba76d851",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8c93ee4b-d476-4b25-b8bc-3aa4ba76d851",
        "outputId": "d92d3d2b-40df-45f8-e87c-357e98eb0053"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['G000005825', 'G000006175', 'G000006605', 'G000006725',\n",
              "       'G000006745', 'G000006785', 'G000006845', 'G000006865',\n",
              "       'G000006925', 'G000006965'], dtype='<U10')"
            ]
          },
          "metadata": {},
          "execution_count": 39
        }
      ],
      "source": [
        "table.ids('observation')[:10]"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "e3e3f63b-d4c7-4ff0-a381-456755c15d34",
      "metadata": {
        "id": "e3e3f63b-d4c7-4ff0-a381-456755c15d34"
      },
      "source": [
        "The information of the OGUs is provided in `ogu.info`."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 40,
      "id": "5f5cb766-2fcc-4365-af1e-6287a62cf370",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5f5cb766-2fcc-4365-af1e-6287a62cf370",
        "outputId": "8309fbbb-3f03-4cee-bd2c-cabbd1774bdd"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "#genome\tassembly_accession\tbioproject\tbiosample\twgs_master\trefseq_category\ttaxid\tspecies_taxid\torgan\n",
            "G000005825\tGCF_000005825.2\tPRJNA224116\tSAMN02603086\t\trepresentative genome\t398511\t79885\tBacillus pse\n",
            "G000006175\tGCF_000006175.1\tPRJNA224116\tSAMN00000040\t\trepresentative genome\t456320\t2188\tMethanococcus\n",
            "G000006605\tGCF_000006605.1\tPRJNA224116\tSAMEA3283089\t\trepresentative genome\t306537\t38289\tCorynebacter\n",
            "G000006725\tGCF_000006725.1\tPRJNA224116\tSAMN02603773\t\trepresentative genome\t160492\t2371\tXylella fasti\n",
            "G000006745\tGCF_000006745.1\tPRJNA224116\tSAMN02603969\t\tna\t243277\t666\tVibrio cholerae O1 biovar El Tor \n",
            "G000006785\tGCF_000006785.2\tPRJNA224116\tSAMN02604089\t\tna\t160490\t1314\tStreptococcus pyogenes M1 GAS\tst\n",
            "G000006845\tGCF_000006845.1\tPRJNA224116\tSAMN02604088\t\tna\t242231\t485\tNeisseria gonorrhoeae FA 1090\tstr\n",
            "G000006865\tGCF_000006865.1\tPRJNA224116\tSAMN02603339\t\tna\t272623\t1358\tLactococcus lactis subsp. lactis\n",
            "G000006925\tGCF_000006925.2\tPRJNA62907\tSAMN02603599\t\treference genome\t198214\t623\tShigella flexneri 2a\n"
          ]
        }
      ],
      "source": [
        "!head $HOME/shotgun/ogu.info | cut -c1-100"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "3b858077-3132-453f-8e6e-b16d8b926b60",
      "metadata": {
        "id": "3b858077-3132-453f-8e6e-b16d8b926b60"
      },
      "source": [
        "Like the 16S rRNA data, the OGU table is also sparse (11% of the cell values are non-zero). An additional challenge in read-based metagenomics is that false positive assignments are prevalent. Therefore, the researcher needs to be scrutious."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 41,
      "id": "0c48c36d-67fa-418c-aeb4-a10c8406e42d",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0c48c36d-67fa-418c-aeb4-a10c8406e42d",
        "outputId": "860d5c7c-5d8e-4ac1-a0f1-33131304e8f5"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "0.11243420779272678"
            ]
          },
          "metadata": {},
          "execution_count": 41
        }
      ],
      "source": [
        "table.get_table_density()"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "998f00ef-24f5-4ecb-81ac-e42a59983c4c",
      "metadata": {
        "id": "998f00ef-24f5-4ecb-81ac-e42a59983c4c"
      },
      "source": [
        "The taxonomic assignments of the OGUs are provided in `ogu.tax`."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 42,
      "id": "4552692b-1a6a-41f0-84f9-77d93aa49758",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        },
        "id": "4552692b-1a6a-41f0-84f9-77d93aa49758",
        "outputId": "afcf4a37-e07e-4cc7-adf7-61780db1d7cc"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "              Domain               Phylum                Class  \\\n",
              "G000005825  Bacteria           Firmicutes              Bacilli   \n",
              "G000006175   Archaea  Methanobacteriota_A         Methanococci   \n",
              "G000006605  Bacteria     Actinobacteriota        Actinomycetia   \n",
              "G000006725  Bacteria       Proteobacteria  Gammaproteobacteria   \n",
              "G000006745  Bacteria       Proteobacteria  Gammaproteobacteria   \n",
              "\n",
              "                       Order            Family            Genus  \\\n",
              "G000005825      Bacillales_H     Bacillaceae_D       Bacillus_S   \n",
              "G000006175   Methanococcales  Methanococcaceae    Methanococcus   \n",
              "G000006605   Mycobacteriales  Mycobacteriaceae  Corynebacterium   \n",
              "G000006725   Xanthomonadales  Xanthomonadaceae          Xylella   \n",
              "G000006745  Enterobacterales      Vibrionaceae           Vibrio   \n",
              "\n",
              "                             Species  \n",
              "G000005825   Bacillus_S pseudofirmus  \n",
              "G000006175    Methanococcus voltae_A  \n",
              "G000006605  Corynebacterium jeikeium  \n",
              "G000006725        Xylella fastidiosa  \n",
              "G000006745           Vibrio cholerae  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-753318eb-b4d3-471e-a51e-27e0850d26c5\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Domain</th>\n",
              "      <th>Phylum</th>\n",
              "      <th>Class</th>\n",
              "      <th>Order</th>\n",
              "      <th>Family</th>\n",
              "      <th>Genus</th>\n",
              "      <th>Species</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>G000005825</th>\n",
              "      <td>Bacteria</td>\n",
              "      <td>Firmicutes</td>\n",
              "      <td>Bacilli</td>\n",
              "      <td>Bacillales_H</td>\n",
              "      <td>Bacillaceae_D</td>\n",
              "      <td>Bacillus_S</td>\n",
              "      <td>Bacillus_S pseudofirmus</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>G000006175</th>\n",
              "      <td>Archaea</td>\n",
              "      <td>Methanobacteriota_A</td>\n",
              "      <td>Methanococci</td>\n",
              "      <td>Methanococcales</td>\n",
              "      <td>Methanococcaceae</td>\n",
              "      <td>Methanococcus</td>\n",
              "      <td>Methanococcus voltae_A</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>G000006605</th>\n",
              "      <td>Bacteria</td>\n",
              "      <td>Actinobacteriota</td>\n",
              "      <td>Actinomycetia</td>\n",
              "      <td>Mycobacteriales</td>\n",
              "      <td>Mycobacteriaceae</td>\n",
              "      <td>Corynebacterium</td>\n",
              "      <td>Corynebacterium jeikeium</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>G000006725</th>\n",
              "      <td>Bacteria</td>\n",
              "      <td>Proteobacteria</td>\n",
              "      <td>Gammaproteobacteria</td>\n",
              "      <td>Xanthomonadales</td>\n",
              "      <td>Xanthomonadaceae</td>\n",
              "      <td>Xylella</td>\n",
              "      <td>Xylella fastidiosa</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>G000006745</th>\n",
              "      <td>Bacteria</td>\n",
              "      <td>Proteobacteria</td>\n",
              "      <td>Gammaproteobacteria</td>\n",
              "      <td>Enterobacterales</td>\n",
              "      <td>Vibrionaceae</td>\n",
              "      <td>Vibrio</td>\n",
              "      <td>Vibrio cholerae</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-753318eb-b4d3-471e-a51e-27e0850d26c5')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-753318eb-b4d3-471e-a51e-27e0850d26c5 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-753318eb-b4d3-471e-a51e-27e0850d26c5');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-0fcbae51-d3f2-43c4-b018-f3d2f64941f7\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-0fcbae51-d3f2-43c4-b018-f3d2f64941f7')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-0fcbae51-d3f2-43c4-b018-f3d2f64941f7 button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "taxonomy",
              "summary": "{\n  \"name\": \"taxonomy\",\n  \"rows\": 14933,\n  \"fields\": [\n    {\n      \"column\": \"Domain\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 2,\n        \"samples\": [\n          \"Archaea\",\n          \"Bacteria\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Phylum\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 121,\n        \"samples\": [\n          \"Nitrospirota_A\",\n          \"Atribacterota\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Class\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 309,\n        \"samples\": [\n          \"UBA796\",\n          \"Fusobacteriia\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Order\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 884,\n        \"samples\": [\n          \"Brevibacillales\",\n          \"Thermosulfidibacterales\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Family\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 1978,\n        \"samples\": [\n          \"Jiangellaceae\",\n          \"JA-3-3Ab\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Genus\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 6580,\n        \"samples\": [\n          \"Verrucomicrobium\",\n          \"FW602-bin19\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Species\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 11854,\n        \"samples\": [\n          \"SZUA-47 sp003231925\",\n          \"JACRCG01 sp009684515\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 42
        }
      ],
      "source": [
        "taxonomy = {}\n",
        "with open(f'{HOME}/shotgun/ogu.tax', 'r') as f:\n",
        "    for line in f:\n",
        "        id_, lineage = line.rstrip().split('\\t')\n",
        "        ranked = [x.split('__')[1] for x in lineage.split('; ')]\n",
        "        taxonomy[id_] = ranked\n",
        "taxonomy = pd.DataFrame.from_dict(taxonomy, orient='index', columns=ranks).replace('', np.nan)\n",
        "taxonomy.head()"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "2c9797c4-6a8b-4fae-b84b-2a9313b987db",
      "metadata": {
        "id": "2c9797c4-6a8b-4fae-b84b-2a9313b987db"
      },
      "source": [
        "### Phylogeny"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "9025bcc0-7577-4fb8-a1eb-3154cce8b85c",
      "metadata": {
        "id": "9025bcc0-7577-4fb8-a1eb-3154cce8b85c"
      },
      "source": [
        "As a classification method, taxonomy is descriptive, operational but coarse. The 7-8 standard taxonomic ranks may fall short in resolving the ever increasing number of discovered organisms (there are now more than 2 million microbial genomes in NCBI). To capture the relationships among microbes with finer resolution, one may resort to phylogeny instead.\n",
        "\n",
        "A [**phylogenetic tree**](https://en.wikipedia.org/wiki/Phylogenetic_tree) is a graph that describes the evolutionary relationships among organisms. All (cellular) organisms descend from a common ancestor. Therefore a tree structure from a root to many tips can unite them all. The phylogenetic tree not only contains the branching pattern (**topology**) indicating the order of speciation, but also provides **branch lengths** indicating the amount of evolution accumulated within each lineage. Therefore, phylogeny is suitable for modeling microbial feature relations in a quantitative manner."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "14f54c97-1e0c-4285-80c6-0869b226b388",
      "metadata": {
        "id": "14f54c97-1e0c-4285-80c6-0869b226b388"
      },
      "source": [
        "The WoL2 reference phylogeny is provided by `ogu.nwk`. It is a [Newick](https://en.wikipedia.org/wiki/Newick_format)-formatted file. One can read it using scikit-bio's [`TreeNode`](https://scikit.bio/docs/latest/generated/skbio.tree.TreeNode.html) class."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 43,
      "id": "8b2ed7d7-a6a9-450f-8da7-ae07ee2ddbfe",
      "metadata": {
        "id": "8b2ed7d7-a6a9-450f-8da7-ae07ee2ddbfe"
      },
      "outputs": [],
      "source": [
        "from skbio import TreeNode"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 44,
      "id": "5251d087-e907-4567-9236-ecf48a2252fc",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5251d087-e907-4567-9236-ecf48a2252fc",
        "outputId": "76fec222-85f0-4a54-b9ae-0425b7e8d337"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<TreeNode, name: N1, internal node count: 15245, tips count: 15953>"
            ]
          },
          "metadata": {},
          "execution_count": 44
        }
      ],
      "source": [
        "tree = TreeNode.read(f'{HOME}/shotgun/ogu.nwk')\n",
        "tree"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "292c90e3-4f04-4274-baa0-e85f7ec5f03c",
      "metadata": {
        "id": "292c90e3-4f04-4274-baa0-e85f7ec5f03c"
      },
      "source": [
        "Because the tree is large, we cannot display the entirety of it. But let's locate an arbitrary tip (the first OGU in the list, `G000005825`), trace up to its grandparent, and display a very small clade descending from it."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 45,
      "id": "500a09c9-d238-4739-9ea8-bcda5415beb9",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "500a09c9-d238-4739-9ea8-bcda5415beb9",
        "outputId": "615835fd-612b-4f10-e3e6-7d794d8c8add"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "                    /-G000005825\n",
            "          /N8376---|\n",
            "         |         |          /-G000513115\n",
            "         |          \\N8873---|\n",
            "         |                   |          /-G000513095\n",
            "-N7879---|                    \\N9376---|\n",
            "         |                              \\-G000513135\n",
            "         |\n",
            "         |          /-G001310555\n",
            "          \\N8377---|\n",
            "                   |          /-G001590785\n",
            "                    \\N8874---|\n",
            "                             |          /-G000292245\n",
            "                              \\N9377---|\n",
            "                                        \\-G001038565\n"
          ]
        }
      ],
      "source": [
        "print(tree.find('G000005825').parent.parent.ascii_art())"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "d1e7e61f-40ff-4d02-97cb-1fd38e20abbb",
      "metadata": {
        "id": "d1e7e61f-40ff-4d02-97cb-1fd38e20abbb"
      },
      "source": [
        "This structure enables a series of analyses, which will be discussed in later sections."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "96f3ea30-c708-41be-b64c-97052fdde8bf",
      "metadata": {
        "id": "96f3ea30-c708-41be-b64c-97052fdde8bf"
      },
      "source": [
        "Note: The 16S rRNA data also comes with a reference phylogeny: `16s.nwk` under the `amplicon` directory. We skipped it because it will be discussed in a later section."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "3bb5a97d-225c-4632-a219-25da55485d0e",
      "metadata": {
        "id": "3bb5a97d-225c-4632-a219-25da55485d0e"
      },
      "source": [
        "### Function"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "f006786f-4982-4bc4-8b1c-147cb397af76",
      "metadata": {
        "id": "f006786f-4982-4bc4-8b1c-147cb397af76"
      },
      "source": [
        "One major advantage of shotgun metagenomics over amplicon sequencing is that it provides information of whole microbial genomes. Therefore, one can identify the functional genes present in those genomes, and subsequently predict the [**biological function**](https://en.wikipedia.org/wiki/Function_(biology)) of the metagenome."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "44c2d6e4-c2cf-4d16-810f-d5965c1f8f3f",
      "metadata": {
        "id": "44c2d6e4-c2cf-4d16-810f-d5965c1f8f3f"
      },
      "source": [
        "The WoL2 database and the [Woltka](https://github.com/qiyunzhu/woltka/) program enabled identification of genes from read-to-genome alignments, and collapsing the results according to a variety of functional ontologies. In the demo dataset, results were collapsed according to the [KEGG](https://www.genome.jp/kegg/) ontology."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "ac4324cd-33b1-4f78-94da-1476d7cd00a3",
      "metadata": {
        "id": "ac4324cd-33b1-4f78-94da-1476d7cd00a3"
      },
      "source": [
        "`ko.biom` records the number of reads per sample that can be assigned to each KEGG ortholog (KO)."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 46,
      "id": "188fdcd2-e5d6-4e51-a068-bd7ecd085b02",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "188fdcd2-e5d6-4e51-a068-bd7ecd085b02",
        "outputId": "c671d5b5-0d17-40b4-e336-3ca1d04bc8d4"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "9319 x 799 <class 'biom.table.Table'> with 1674550 nonzero entries (22% dense)"
            ]
          },
          "metadata": {},
          "execution_count": 46
        }
      ],
      "source": [
        "table = Table.read(f'{HOME}/shotgun/ko.biom')\n",
        "table"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 47,
      "id": "e7605030-280c-44be-8009-6441c83235b1",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "e7605030-280c-44be-8009-6441c83235b1",
        "outputId": "fa21761b-5997-4d63-db28-ecf2f8ae329e"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['K00001', 'K00003', 'K00004', 'K00005', 'K00007', 'K00008',\n",
              "       'K00009', 'K00010', 'K00012', 'K00013'], dtype='<U6')"
            ]
          },
          "metadata": {},
          "execution_count": 47
        }
      ],
      "source": [
        "table.ids('observation')[:10]"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "79a5055c-0d81-421a-92d5-218fa1cb1c67",
      "metadata": {
        "id": "79a5055c-0d81-421a-92d5-218fa1cb1c67"
      },
      "source": [
        "The description of each KO is provided in `ko.name`."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 48,
      "id": "1e4c162a-bf63-4dbc-8441-92abcef1b6da",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "1e4c162a-bf63-4dbc-8441-92abcef1b6da",
        "outputId": "4bc45d23-f1f1-487f-b0a1-a7d833322a60"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "K00001\tE1.1.1.1, adh; alcohol dehydrogenase [EC:1.1.1.1]\n",
            "K00003\thom; homoserine dehydrogenase [EC:1.1.1.3]\n",
            "K00004\tBDH, butB; (R,R)-butanediol dehydrogenase / meso-butanediol dehydrogenase / diacetyl reductase [EC:1.1.1.4 1.1.1.- 1.1.1.303]\n",
            "K00005\tgldA; glycerol dehydrogenase [EC:1.1.1.6]\n",
            "K00007\tdalD; D-arabinitol 4-dehydrogenase [EC:1.1.1.11]\n"
          ]
        }
      ],
      "source": [
        "!head -n5 $HOME/shotgun/ko.name"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "cac32134-a2d1-4a8b-bf99-af36b164ba84",
      "metadata": {
        "id": "cac32134-a2d1-4a8b-bf99-af36b164ba84"
      },
      "source": [
        "The KOs were grouped by their [Enzyme Commission (EC) numbers](https://enzyme.expasy.org/), as provided in `ec.biom`."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 49,
      "id": "0aea2356-5262-4298-86a7-8e363a9f3726",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0aea2356-5262-4298-86a7-8e363a9f3726",
        "outputId": "c8437d6b-aff7-4446-aebd-9fb0b0928938"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "3326 x 797 <class 'biom.table.Table'> with 784055 nonzero entries (29% dense)"
            ]
          },
          "metadata": {},
          "execution_count": 49
        }
      ],
      "source": [
        "table = Table.read(f'{HOME}/shotgun/ec.biom')\n",
        "table"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 50,
      "id": "addcdcd0-c8c1-493c-b114-bde6af04dbf9",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "addcdcd0-c8c1-493c-b114-bde6af04dbf9",
        "outputId": "45259924-86e7-409e-9e4c-d187e8d08746"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['1.-.-.-', '1.1.-.-', '1.1.1.-', '1.1.1.1', '1.1.1.10',\n",
              "       '1.1.1.100', '1.1.1.102', '1.1.1.103', '1.1.1.107', '1.1.1.108'],\n",
              "      dtype='<U11')"
            ]
          },
          "metadata": {},
          "execution_count": 50
        }
      ],
      "source": [
        "table.ids('observation')[:10]"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "3329fd8e-90c0-4b72-a973-78f2c0272809",
      "metadata": {
        "id": "3329fd8e-90c0-4b72-a973-78f2c0272809"
      },
      "source": [
        "The KOs were collapsed by the metabolic pathways they participate, as provided in `pathway.biom`."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 51,
      "id": "ee225ea4-9278-49ba-9c0e-c7028318945d",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ee225ea4-9278-49ba-9c0e-c7028318945d",
        "outputId": "2dc9ea79-9a26-48ff-90a6-60158f288c7b"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "110 x 787 <class 'biom.table.Table'> with 59624 nonzero entries (68% dense)"
            ]
          },
          "metadata": {},
          "execution_count": 51
        }
      ],
      "source": [
        "table = Table.read(f'{HOME}/shotgun/pathway.biom')\n",
        "table"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 52,
      "id": "7444b9ec-9da1-44f9-ad1b-1e63e19ae3db",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7444b9ec-9da1-44f9-ad1b-1e63e19ae3db",
        "outputId": "61477bac-370b-4803-9cbd-633a9d714bd2"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['ko00010', 'ko00020', 'ko00030', 'ko00040', 'ko00051', 'ko00052',\n",
              "       'ko00053', 'ko00061', 'ko00062', 'ko00071'], dtype='<U7')"
            ]
          },
          "metadata": {},
          "execution_count": 52
        }
      ],
      "source": [
        "table.ids('observation')[:10]"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "c8b0d18e-7851-4e81-b447-6a089cbbfdeb",
      "metadata": {
        "id": "c8b0d18e-7851-4e81-b447-6a089cbbfdeb"
      },
      "source": [
        "The description of each pathway is provided in `pathway.name`."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 53,
      "id": "ad52bfb5-de9d-470c-90ae-bbcd1ff8d4e7",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ad52bfb5-de9d-470c-90ae-bbcd1ff8d4e7",
        "outputId": "38632f45-b0e1-4773-dad8-ca0c1c11bf23"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "ko00010\tGlycolysis / Gluconeogenesis\n",
            "ko00020\tCitrate cycle (TCA cycle)\n",
            "ko00030\tPentose phosphate pathway\n",
            "ko00040\tPentose and glucuronate interconversions\n",
            "ko00051\tFructose and mannose metabolism\n"
          ]
        }
      ],
      "source": [
        "!head -n5 $HOME/shotgun/pathway.name"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "3725d00c-cdea-49c5-a53f-6b8de3f338e0",
      "metadata": {
        "id": "3725d00c-cdea-49c5-a53f-6b8de3f338e0"
      },
      "source": [
        "Shotgun metagenomics captures a static image of microbiome function -- the presence of functional genes in microbial genomes. However, it falls short in informing the expression of those genes. A complementary technique -- [**metatranscriptomics**](https://en.wikipedia.org/wiki/Metatranscriptomics), quantifies the mRNAs (transcripts) of a biological sample, thereby providing such information. This method has seen increasing popularity in microbiome research."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "94026ec5-ea73-4e95-b40b-df9ddecb82da",
      "metadata": {
        "id": "94026ec5-ea73-4e95-b40b-df9ddecb82da"
      },
      "source": [
        "## Metagenomic assembly"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "90261e94-c648-4498-8057-5f29fc8dceac",
      "metadata": {
        "id": "90261e94-c648-4498-8057-5f29fc8dceac"
      },
      "source": [
        "Shotgun metagenomic sequencing data can be bioinformatically [**assembed**](https://en.wikipedia.org/wiki/Sequence_assembly) into longer sequences (contigs or scaffolds), and [**binned**](https://en.wikipedia.org/wiki/Binning_(metagenomics)) into putative draft genomes, a.k.a., metagenome-assembled genomes (**MAGs**). This process allows researchers to directly characterize microbes present in a sample, rather than relying on a reference as a proxy.\n",
        "\n",
        "Assembly-based metagenomics provides more compelling evidence and more comprehensive view of individual microbes, as compared with read-based metagenomics. However, assembly and binning are bioinformatically challenging (i.e., MAGs may not be perfectly recovered), and less sensitive with low-abundance microbes. Therefore both strategies are complementary to each other in a metagenomic study."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "e6dc245b-c516-452b-ac91-e367cf03f256",
      "metadata": {
        "id": "e6dc245b-c516-452b-ac91-e367cf03f256"
      },
      "source": [
        "The results of metagenomic assembly are provided under the `assembly` directory."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 54,
      "id": "6158ec2c-f7f1-44a6-9e19-5c6fa409946c",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "6158ec2c-f7f1-44a6-9e19-5c6fa409946c",
        "outputId": "d98e3527-a684-4d67-c496-f39a01217d63"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "fasta  mag.biom  mag.tax  prokka\n"
          ]
        }
      ],
      "source": [
        "!ls $HOME/assembly"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "d97d32e7-571b-4671-855a-19b14a6ec6f9",
      "metadata": {
        "id": "d97d32e7-571b-4671-855a-19b14a6ec6f9"
      },
      "source": [
        "A total of **811** high-quality MAGs were recovered from the EMP500 samples. Their sequences (scaffolds of individual bins) are provided under `assembly/fasta`."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 55,
      "id": "79314a9e-b9f9-4a93-9fb2-89e9fe411bb1",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "79314a9e-b9f9-4a93-9fb2-89e9fe411bb1",
        "outputId": "7a3ad136-1cc7-48fc-d26e-f224ffb52780"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Bioreactor.Metabat.Bin.10.fa\n",
            "Bioreactor.Metabat.Bin.11.fa\n",
            "Bioreactor.Metabat.Bin.14.fa\n",
            "Bioreactor.Metabat.Bin.16.fa\n",
            "Bioreactor.Metabat.Bin.19.fa\n"
          ]
        }
      ],
      "source": [
        "!basename -a $HOME/assembly/fasta/*.fa | head -n5"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 56,
      "id": "a4b223c3-eb6d-4ae9-9e37-52ea595faab2",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "a4b223c3-eb6d-4ae9-9e37-52ea595faab2",
        "outputId": "07fd8a3a-4da8-4f0f-a10a-a730aa58995c"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            ">NODE_8_length_861446_cov_17.890671\n",
            "CGTTTTGTTTAATTTCCAAATTAAATTGAAATTTTTTTCGTTTAGATTATGTAAATAACT\n",
            "GATAATAAGATGGAAAAATTTCACTAATTATTATAAAGTAAATTCTAAAAATATTTACGA\n",
            "ATGTTTTCTATAGCCCCGATTGTAAAGGAAATCCTTTTTTGCGAGATGGCTTGCCGAGCA\n",
            "AAAAAGATTGTAGTGGAAAGCGGGAAATAGCTTCTAATTTTCTTTATAAATATTGATTAA\n"
          ]
        }
      ],
      "source": [
        "!head -n5 $HOME/assembly/fasta/Bioreactor.Metabat.Bin.10.fa"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "8f1e7bf5-f284-4109-81a6-4344e06d9c69",
      "metadata": {
        "id": "8f1e7bf5-f284-4109-81a6-4344e06d9c69"
      },
      "source": [
        "The abundance of each MAG in each sample was quantifying by aligning sequencing reads back to MAG sequences. The results are provided in `mag.biom`."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 57,
      "id": "ea304efc-1c92-4813-b237-071e88159b4e",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ea304efc-1c92-4813-b237-071e88159b4e",
        "outputId": "37c7db2e-9c80-462f-dbdc-7d200a222448"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "811 x 813 <class 'biom.table.Table'> with 249292 nonzero entries (37% dense)"
            ]
          },
          "metadata": {},
          "execution_count": 57
        }
      ],
      "source": [
        "table = Table.read(f'{HOME}/assembly/mag.biom')\n",
        "table"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 58,
      "id": "2efb11f0-0fb5-4256-af0d-96681cf5d8da",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2efb11f0-0fb5-4256-af0d-96681cf5d8da",
        "outputId": "f3ad11ec-a534-4ff8-e5d7-467e2ba167e2"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['SW.Metabat.Bin.241.fa', 'Song_Terr.Metabat.Bin.486.fa',\n",
              "       'Song_Terr.Metabat.Bin.574.fa', 'Song_Terr.Metabat.Bin.455.fa',\n",
              "       'Song_Terr.Metabat.Bin.534.fa', 'Song_Terr.Metabat.Bin.529.fa',\n",
              "       'Myrold59_soil.Metabat.Bin.1.fa', 'SaltMarsh.Metabat.Bin.33.fa',\n",
              "       'Bird_feces.Metabat.Bin.58.fa', 'Myrold59_soil.Metabat.Bin.7.fa'],\n",
              "      dtype='<U35')"
            ]
          },
          "metadata": {},
          "execution_count": 58
        }
      ],
      "source": [
        "table.ids('observation')[:10]"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "1360e0b8-31cc-4337-92fb-0786980c78c9",
      "metadata": {
        "id": "1360e0b8-31cc-4337-92fb-0786980c78c9"
      },
      "source": [
        "The MAGs were classified using [GTDB-Tk](https://github.com/Ecogenomics/GTDBTk). The results are provided in `mag.tax`."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 59,
      "id": "83af5000-6169-45c7-abae-3be44f55a3b9",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "83af5000-6169-45c7-abae-3be44f55a3b9",
        "outputId": "7896eaee-c1ed-412f-e6f9-5cb536c1e429"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Bioreactor.Metabat.Bin.10\td__Bacteria; p__Bacteroidota; c__Bacteroidia; o__Flavobacteriales; f__Weeksellaceae; g__Cloacibacterium; s__\n",
            "Bioreactor.Metabat.Bin.11\td__Bacteria; p__Patescibacteria; c__Paceibacteria; o__UBA9973; f__UBA918; g__OLB19; s__\n",
            "Bioreactor.Metabat.Bin.14\td__Bacteria; p__Bacteroidota; c__Bacteroidia; o__Chitinophagales; f__Chitinophagaceae; g__UTBCD1; s__UTBCD1 sp002066995\n",
            "Bioreactor.Metabat.Bin.16\td__Bacteria; p__Bacteroidota; c__Ignavibacteria; o__SJA-28; f__B-1AR; g__OLB4; s__OLB4 sp002050165\n",
            "Bioreactor.Metabat.Bin.19\td__Bacteria; p__Pseudomonadota; c__Alphaproteobacteria; o__Caulobacterales; f__Caulobacteraceae; g__Brevundimonas; s__\n"
          ]
        }
      ],
      "source": [
        "!head -n5 $HOME/assembly/mag.tax"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "5c48cbbd-2c74-4ef1-8984-98a8740b6657",
      "metadata": {
        "id": "5c48cbbd-2c74-4ef1-8984-98a8740b6657"
      },
      "source": [
        "### Annotation"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "0bf0953e-33d4-41e0-bca5-f31cc6ca4fdc",
      "metadata": {
        "id": "0bf0953e-33d4-41e0-bca5-f31cc6ca4fdc"
      },
      "source": [
        "Each MAG was [**annotated**](https://en.wikipedia.org/wiki/DNA_annotation) using [Prokka](https://github.com/tseemann/prokka). This process identifies functional features, such as protein-coding genes, tRNAs, rRNAs and more, on each genome. The outputs are provided under `assembly/prokka`. Specifically, there are protein sequences (`faa`), their corresponding coding sequences (CDSes) (`ffn`) and the annotation tables (`gff`)."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 60,
      "id": "ced13d08-6bf1-482d-9d3c-dcf7c2d99f5a",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ced13d08-6bf1-482d-9d3c-dcf7c2d99f5a",
        "outputId": "7372c0da-58f3-4f2e-e8c5-5a7d1ee5673e"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "faa  ffn  gff  stats.tsv\n"
          ]
        }
      ],
      "source": [
        "!ls $HOME/assembly/prokka"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "b45f1058-73c3-441b-b919-e4ffeb20bf75",
      "metadata": {
        "id": "b45f1058-73c3-441b-b919-e4ffeb20bf75"
      },
      "source": [
        "The predicted protein sequences and their descriptions are like:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 61,
      "id": "80639fa6-3463-4dc0-b933-459dfc98c0bb",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "80639fa6-3463-4dc0-b933-459dfc98c0bb",
        "outputId": "cee86fc0-b3f8-45c6-ab51-3920edc72ce5"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/bin/bash: line 1: emp500/assembly/prokka/faa/Bioreactor.Metabat.Bin.10.faa.gz: No such file or directory\n"
          ]
        }
      ],
      "source": [
        "!zcat < $HOME/assembly/prokka/faa/Bioreactor.Metabat.Bin.10.faa.gz | head"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "4395bada-c962-4eea-99f3-de2258094600",
      "metadata": {
        "id": "4395bada-c962-4eea-99f3-de2258094600"
      },
      "source": [
        "Each [**GFF3**](https://en.wikipedia.org/wiki/General_feature_format)-formatted annotation table records the start and end coordinates of each feature and their descriptions. scikit-bio's [`IntervalMetadata`](https://scikit.bio/docs/latest/generated/skbio.metadata.IntervalMetadata.html) class can process such data."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 62,
      "id": "f67595be-cc01-414e-9601-e46960a99446",
      "metadata": {
        "id": "f67595be-cc01-414e-9601-e46960a99446"
      },
      "outputs": [],
      "source": [
        "from skbio.metadata import IntervalMetadata"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "c9eefb09-231f-4fba-8b35-b141b9cdefe1",
      "metadata": {
        "id": "c9eefb09-231f-4fba-8b35-b141b9cdefe1"
      },
      "source": [
        "Let's read the annotation of the first scaffold sequence of a MAG."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 63,
      "id": "28dc6561-3530-40f2-8953-c48632f1a2cf",
      "metadata": {
        "id": "28dc6561-3530-40f2-8953-c48632f1a2cf"
      },
      "outputs": [],
      "source": [
        "fp = f'{HOME}/assembly/prokka/gff/Bioreactor.Metabat.Bin.10.gff'"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 64,
      "id": "7947b922-8cee-4267-8310-0fec76593484",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7947b922-8cee-4267-8310-0fec76593484",
        "outputId": "df4bd479-8b0b-4470-9010-8f6dc8803a5d"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "840 interval features\n",
              "---------------------\n",
              "Interval(interval_metadata=<136076787586720>, bounds=[(213, 885)], fuzzy=[(False, False)], metadata={'source': 'Prodigal:002006', 'type': 'CDS', 'score': '.', 'strand': '-', 'phase': 0, 'ID': 'GHKEOJEL_00001', 'inference': 'ab initio prediction:Prodigal:002006', 'locus_tag': 'GHKEOJEL_00001', 'product': 'hypothetical protein'})\n",
              "Interval(interval_metadata=<136076787586720>, bounds=[(896, 1751)], fuzzy=[(False, False)], metadata={'source': 'Prodigal:002006', 'type': 'CDS', 'score': '.', 'strand': '-', 'phase': 0, 'ID': 'GHKEOJEL_00002', 'eC_number': '1.1.1.133', 'Name': 'rmlD', 'gene': 'rmlD', 'inference': 'ab initio prediction:Prodigal:002006,similar to AA sequence:UniProtKB:O66251', 'locus_tag': 'GHKEOJEL_00002', 'product': 'dTDP-4-dehydrorhamnose reductase'})\n",
              "...\n",
              "Interval(interval_metadata=<136076787586720>, bounds=[(860451, 860829)], fuzzy=[(False, False)], metadata={'source': 'Prodigal:002006', 'type': 'CDS', 'score': '.', 'strand': '+', 'phase': 0, 'ID': 'GHKEOJEL_00839', 'Name': 'gcvH_1', 'gene': 'gcvH_1', 'inference': 'ab initio prediction:Prodigal:002006,similar to AA sequence:UniProtKB:P64214', 'locus_tag': 'GHKEOJEL_00839', 'product': 'Glycine cleavage system H protein'})\n",
              "Interval(interval_metadata=<136076787586720>, bounds=[(860857, 861184)], fuzzy=[(False, False)], metadata={'source': 'Prodigal:002006', 'type': 'CDS', 'score': '.', 'strand': '+', 'phase': 0, 'ID': 'GHKEOJEL_00840', 'inference': 'ab initio prediction:Prodigal:002006', 'locus_tag': 'GHKEOJEL_00840', 'product': 'hypothetical protein'})"
            ]
          },
          "metadata": {},
          "execution_count": 64
        }
      ],
      "source": [
        "anno = IntervalMetadata.read(fp, seq_id='NODE_8_length_861446_cov_17.890671')\n",
        "anno"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "ec9b2838-546f-4609-86c8-a601920e575e",
      "metadata": {
        "id": "ec9b2838-546f-4609-86c8-a601920e575e"
      },
      "source": [
        "## Untargeted metabolomics"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "a516daa8-f0a7-45c0-8b82-e771e7f3d0e6",
      "metadata": {
        "id": "a516daa8-f0a7-45c0-8b82-e771e7f3d0e6"
      },
      "source": [
        "The chemistry of a biological sample can be assessed by [mass spectrometry](https://en.wikipedia.org/wiki/Mass_spectrometry) (**MS**), a technique that breaks molecules into ions and quantifies the mass-to-charge ratio of ions. The information can be analyzed bioinformatically to recover the chemical profile of the sample.\n",
        "\n",
        "[**Metabolomics**](https://en.wikipedia.org/wiki/Metabolomics) refers to the analysis of _small molecules_ (usually <1.5 kDa) in a sample. These molecules are usually (but not always) metabolites involved in a biological system. Therefore, metabolomics describes the causations and consequences of biological functions. Metabolomics is usually achieved through MS. The term \"metabolomics\" seemingly indicates microbiome (\"meta-)\", but in fact it covers all small molecules, microbe- or host-related, or ones related to both, a sample.\n",
        "\n",
        "- The analysis of _large molecules_, typically proteins, is referred to as [proteomics](https://en.wikipedia.org/wiki/Proteomics) (or [**metaproteomics**](https://en.wikipedia.org/wiki/Metaproteomics) when it concerns microbiomes). It is also (usually) achieved using MS, however with different target molecular weight range. It is useful for analyzing the translation of microbial genes (in addition to transcription, as metatranscriptomics does).\n",
        "\n",
        "Just like amplicon and shotgun sequencing methods, metabolomics can be categorize into targetted and **untargeted**, the latter captures any molecule (without a specific target) in a biological system."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "939ce645-3ffb-4b55-aa95-6192473671bc",
      "metadata": {
        "id": "939ce645-3ffb-4b55-aa95-6192473671bc"
      },
      "source": [
        "The mass spectrometry data are provided under the `masspec` directory."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 65,
      "id": "b2b12444-c3c2-4c67-97e0-566087c88add",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "b2b12444-c3c2-4c67-97e0-566087c88add",
        "outputId": "aad3de15-33e4-4778-d31e-0f6f28f827a6"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "canopus.tsv  fbmn.tsv  is.microbial  lcms.biom\n"
          ]
        }
      ],
      "source": [
        "!ls $HOME/masspec"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "63149cd1-ef4c-425c-8eac-5d638a213f25",
      "metadata": {
        "id": "63149cd1-ef4c-425c-8eac-5d638a213f25"
      },
      "source": [
        "The file `lcms.biom` is a data table derived from the liquid chromatography-tandem mass spectrometry ([**LC-MS/MS**](https://en.wikipedia.org/wiki/Liquid_chromatography%E2%80%93mass_spectrometry)). The spectrum data were processed using [MZmine](https://mzmine.github.io/download.html). Each cell value represents the ion intensity (i.e., peak area) of a spectrum.\n",
        "\n",
        "- Complementarily, the EMP500 team performed gas chromatography–mass spectrometry ([**GC–MS**](https://en.wikipedia.org/wiki/Gas_chromatography%E2%80%93mass_spectrometry)), a technique suitable for identifying substances with high volatility (whereas LC-MS is broad-spectrum). The GC-MS data is not included in the demo dataset, but you can try to find and analyze it for practice."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 66,
      "id": "5e565743-564c-4785-94cd-4cb6d5ab382c",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5e565743-564c-4785-94cd-4cb6d5ab382c",
        "outputId": "8fc3caf2-8ad5-4977-b803-61a8fc8ca62b"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "52496 x 618 <class 'biom.table.Table'> with 13252433 nonzero entries (40% dense)"
            ]
          },
          "metadata": {},
          "execution_count": 66
        }
      ],
      "source": [
        "table = Table.read(f'{HOME}/masspec/lcms.biom')\n",
        "table"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "7b09d989-96b4-48fb-90be-7ad09a2a081b",
      "metadata": {
        "id": "7b09d989-96b4-48fb-90be-7ad09a2a081b"
      },
      "source": [
        "Unlike DNA sequencing data, the cell values are not integers."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 67,
      "id": "ed7ce179-4640-48ba-a625-ba1329285391",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 256
        },
        "id": "ed7ce179-4640-48ba-a625-ba1329285391",
        "outputId": "6eb8c39d-bfd1-48ba-ece2-783543bd1fbb"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "    13114.mayer.34.s008  13114.king.27.s003  13114.metcalf.40.s014  \\\n",
              "48        708502.552009       742278.375775          780958.569711   \n",
              "66        756467.905517       181889.791978           113203.48584   \n",
              "72       1032820.099247       481232.796289         1552998.498345   \n",
              "74        156171.224081        54795.567999                      0   \n",
              "78                    0       114801.646948           862922.71701   \n",
              "\n",
              "    13114.berry.2.s009  13114.mayer.34.s006  13114.berry.2.s016  \\\n",
              "48       525207.300634        837844.689714      1053890.955027   \n",
              "66        75658.911859         70067.723199       112611.885491   \n",
              "72                   0        614104.200347       117087.877593   \n",
              "74        51718.926669          58285.43987         79254.90734   \n",
              "78                   0        619526.558313       154669.075313   \n",
              "\n",
              "    13114.mayer.34.s004  13114.berry.2.s013  13114.mayer.34.s005  \\\n",
              "48       2277550.765829       552157.680682       1179237.016635   \n",
              "66        594923.758322       174490.808163        195012.211269   \n",
              "72        547994.579832      1381665.580329        222175.789584   \n",
              "74        173580.544466        623636.91613        248736.367788   \n",
              "78         66938.299592                   0        902987.716157   \n",
              "\n",
              "    13114.metcalf.40.s011  ...  13114.pinto.62.s012  13114.pinto.62.s013  \\\n",
              "48         2748297.819676  ...         47982.975828         19442.191001   \n",
              "66         2008954.672368  ...         20829.936317         16638.240278   \n",
              "72          518249.531565  ...        894220.947705        671855.752303   \n",
              "74                      0  ...           43242.9141                    0   \n",
              "78                      0  ...                    0        125630.772463   \n",
              "\n",
              "    13114.pinto.62.s014  13114.pinto.62.s015  13114.pinto.62.s016  \\\n",
              "48         12980.440753        119126.009914        118102.826454   \n",
              "66         30219.774487         23545.752273         14887.642869   \n",
              "72                    0         137499.18635         188622.31547   \n",
              "74         45532.972947          60441.97725                    0   \n",
              "78        411793.312374                    0        150690.051884   \n",
              "\n",
              "    13114.pinto.62.s020  13114.pinto.62.s021  13114.pinto.62.s022  \\\n",
              "48        776581.281141         50417.260057          49672.51069   \n",
              "66         95195.165107          8013.130592         13010.343786   \n",
              "72        649648.468435        394061.155994         142007.36747   \n",
              "74                    0         42873.470147         63038.621463   \n",
              "78         64049.512684        105533.343132                    0   \n",
              "\n",
              "    13114.pinto.62.s023  13114.pinto.62.s024  \n",
              "48         14376.599412         16604.140035  \n",
              "66         13802.244664         16931.318684  \n",
              "72         45646.719534                    0  \n",
              "74                    0                    0  \n",
              "78        114020.088281         81276.930523  \n",
              "\n",
              "[5 rows x 618 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-5f14f3c7-2f92-4b71-bc85-d1038771c9a6\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>13114.mayer.34.s008</th>\n",
              "      <th>13114.king.27.s003</th>\n",
              "      <th>13114.metcalf.40.s014</th>\n",
              "      <th>13114.berry.2.s009</th>\n",
              "      <th>13114.mayer.34.s006</th>\n",
              "      <th>13114.berry.2.s016</th>\n",
              "      <th>13114.mayer.34.s004</th>\n",
              "      <th>13114.berry.2.s013</th>\n",
              "      <th>13114.mayer.34.s005</th>\n",
              "      <th>13114.metcalf.40.s011</th>\n",
              "      <th>...</th>\n",
              "      <th>13114.pinto.62.s012</th>\n",
              "      <th>13114.pinto.62.s013</th>\n",
              "      <th>13114.pinto.62.s014</th>\n",
              "      <th>13114.pinto.62.s015</th>\n",
              "      <th>13114.pinto.62.s016</th>\n",
              "      <th>13114.pinto.62.s020</th>\n",
              "      <th>13114.pinto.62.s021</th>\n",
              "      <th>13114.pinto.62.s022</th>\n",
              "      <th>13114.pinto.62.s023</th>\n",
              "      <th>13114.pinto.62.s024</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>48</th>\n",
              "      <td>708502.552009</td>\n",
              "      <td>742278.375775</td>\n",
              "      <td>780958.569711</td>\n",
              "      <td>525207.300634</td>\n",
              "      <td>837844.689714</td>\n",
              "      <td>1053890.955027</td>\n",
              "      <td>2277550.765829</td>\n",
              "      <td>552157.680682</td>\n",
              "      <td>1179237.016635</td>\n",
              "      <td>2748297.819676</td>\n",
              "      <td>...</td>\n",
              "      <td>47982.975828</td>\n",
              "      <td>19442.191001</td>\n",
              "      <td>12980.440753</td>\n",
              "      <td>119126.009914</td>\n",
              "      <td>118102.826454</td>\n",
              "      <td>776581.281141</td>\n",
              "      <td>50417.260057</td>\n",
              "      <td>49672.51069</td>\n",
              "      <td>14376.599412</td>\n",
              "      <td>16604.140035</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>66</th>\n",
              "      <td>756467.905517</td>\n",
              "      <td>181889.791978</td>\n",
              "      <td>113203.48584</td>\n",
              "      <td>75658.911859</td>\n",
              "      <td>70067.723199</td>\n",
              "      <td>112611.885491</td>\n",
              "      <td>594923.758322</td>\n",
              "      <td>174490.808163</td>\n",
              "      <td>195012.211269</td>\n",
              "      <td>2008954.672368</td>\n",
              "      <td>...</td>\n",
              "      <td>20829.936317</td>\n",
              "      <td>16638.240278</td>\n",
              "      <td>30219.774487</td>\n",
              "      <td>23545.752273</td>\n",
              "      <td>14887.642869</td>\n",
              "      <td>95195.165107</td>\n",
              "      <td>8013.130592</td>\n",
              "      <td>13010.343786</td>\n",
              "      <td>13802.244664</td>\n",
              "      <td>16931.318684</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>72</th>\n",
              "      <td>1032820.099247</td>\n",
              "      <td>481232.796289</td>\n",
              "      <td>1552998.498345</td>\n",
              "      <td>0</td>\n",
              "      <td>614104.200347</td>\n",
              "      <td>117087.877593</td>\n",
              "      <td>547994.579832</td>\n",
              "      <td>1381665.580329</td>\n",
              "      <td>222175.789584</td>\n",
              "      <td>518249.531565</td>\n",
              "      <td>...</td>\n",
              "      <td>894220.947705</td>\n",
              "      <td>671855.752303</td>\n",
              "      <td>0</td>\n",
              "      <td>137499.18635</td>\n",
              "      <td>188622.31547</td>\n",
              "      <td>649648.468435</td>\n",
              "      <td>394061.155994</td>\n",
              "      <td>142007.36747</td>\n",
              "      <td>45646.719534</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>74</th>\n",
              "      <td>156171.224081</td>\n",
              "      <td>54795.567999</td>\n",
              "      <td>0</td>\n",
              "      <td>51718.926669</td>\n",
              "      <td>58285.43987</td>\n",
              "      <td>79254.90734</td>\n",
              "      <td>173580.544466</td>\n",
              "      <td>623636.91613</td>\n",
              "      <td>248736.367788</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>43242.9141</td>\n",
              "      <td>0</td>\n",
              "      <td>45532.972947</td>\n",
              "      <td>60441.97725</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>42873.470147</td>\n",
              "      <td>63038.621463</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>78</th>\n",
              "      <td>0</td>\n",
              "      <td>114801.646948</td>\n",
              "      <td>862922.71701</td>\n",
              "      <td>0</td>\n",
              "      <td>619526.558313</td>\n",
              "      <td>154669.075313</td>\n",
              "      <td>66938.299592</td>\n",
              "      <td>0</td>\n",
              "      <td>902987.716157</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>125630.772463</td>\n",
              "      <td>411793.312374</td>\n",
              "      <td>0</td>\n",
              "      <td>150690.051884</td>\n",
              "      <td>64049.512684</td>\n",
              "      <td>105533.343132</td>\n",
              "      <td>0</td>\n",
              "      <td>114020.088281</td>\n",
              "      <td>81276.930523</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows × 618 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-5f14f3c7-2f92-4b71-bc85-d1038771c9a6')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-5f14f3c7-2f92-4b71-bc85-d1038771c9a6 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-5f14f3c7-2f92-4b71-bc85-d1038771c9a6');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-07c4baa5-34fb-4926-821c-b6d9259909ca\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-07c4baa5-34fb-4926-821c-b6d9259909ca')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-07c4baa5-34fb-4926-821c-b6d9259909ca button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe"
            }
          },
          "metadata": {},
          "execution_count": 67
        }
      ],
      "source": [
        "table.to_dataframe().head()"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "7fa17ab9-890c-43b4-abf5-8fafe3801a7b",
      "metadata": {
        "id": "7fa17ab9-890c-43b4-abf5-8fafe3801a7b"
      },
      "source": [
        "The spectrum data were subject to the feature-based molecular networking workflow ([**FBMN**](https://lfnothias.github.io/GNPSDocumentation/featurebasedmolecularnetworking/)) implemented on the [GNPS](https://gnps.ucsd.edu/ProteoSAFe/static/gnps-splash.jsp) server ([Wang et al., 2016](https://www.nature.com/articles/nbt.3597)) in order to group and identify them and to identify their source molecules. The latter task was achieve by searching reference databases (just like with sequence data)."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 68,
      "id": "14672757-4203-4d15-a834-3edcb22194fb",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 324
        },
        "id": "14672757-4203-4d15-a834-3edcb22194fb",
        "outputId": "4384a09a-910e-4bd0-8a49-4ac77809e672"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "     mqscore            spectrum  \\\n",
              "id                                 \n",
              "48  0.905755  CCMSLIB00003303197   \n",
              "66  0.953557  CCMSLIB00003303202   \n",
              "72  0.811642  CCMSLIB00005761025   \n",
              "74  0.761311  CCMSLIB00005761025   \n",
              "78  0.706896  CCMSLIB00005764095   \n",
              "\n",
              "                                              library  \\\n",
              "id                                                      \n",
              "48            2,6-Di-tert-butyl-4-hydroxymethylphenol   \n",
              "66            2,6-Di-tert-butyl-4-hydroxymethylphenol   \n",
              "72  Massbank:LU030804 Lauryldiethanolamine|2-[dode...   \n",
              "74  Massbank:LU030804 Lauryldiethanolamine|2-[dode...   \n",
              "78  Massbank:LU030803 Lauryldiethanolamine|2-[dode...   \n",
              "\n",
              "                                smiles  \\\n",
              "id                                       \n",
              "48  CC(C)(C)C1=CC(=CC(=C1O)C(C)(C)C)CO   \n",
              "66  CC(C)(C)C1=CC(=CC(=C1O)C(C)(C)C)CO   \n",
              "72               CCCCCCCCCCCCN(CCO)CCO   \n",
              "74               CCCCCCCCCCCCN(CCO)CCO   \n",
              "78               CCCCCCCCCCCCN(CCO)CCO   \n",
              "\n",
              "                                                inchi  \n",
              "id                                                     \n",
              "48  InChI=1S/C15H24O2/c1-14(2,3)11-7-10(9-16)8-12(...  \n",
              "66  InChI=1S/C15H24O2/c1-14(2,3)11-7-10(9-16)8-12(...  \n",
              "72  1S/C16H35NO2/c1-2-3-4-5-6-7-8-9-10-11-12-17(13...  \n",
              "74  1S/C16H35NO2/c1-2-3-4-5-6-7-8-9-10-11-12-17(13...  \n",
              "78  1S/C16H35NO2/c1-2-3-4-5-6-7-8-9-10-11-12-17(13...  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-d53447bb-384e-4f6e-b49d-5ba0705f4dcd\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>mqscore</th>\n",
              "      <th>spectrum</th>\n",
              "      <th>library</th>\n",
              "      <th>smiles</th>\n",
              "      <th>inchi</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>id</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>48</th>\n",
              "      <td>0.905755</td>\n",
              "      <td>CCMSLIB00003303197</td>\n",
              "      <td>2,6-Di-tert-butyl-4-hydroxymethylphenol</td>\n",
              "      <td>CC(C)(C)C1=CC(=CC(=C1O)C(C)(C)C)CO</td>\n",
              "      <td>InChI=1S/C15H24O2/c1-14(2,3)11-7-10(9-16)8-12(...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>66</th>\n",
              "      <td>0.953557</td>\n",
              "      <td>CCMSLIB00003303202</td>\n",
              "      <td>2,6-Di-tert-butyl-4-hydroxymethylphenol</td>\n",
              "      <td>CC(C)(C)C1=CC(=CC(=C1O)C(C)(C)C)CO</td>\n",
              "      <td>InChI=1S/C15H24O2/c1-14(2,3)11-7-10(9-16)8-12(...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>72</th>\n",
              "      <td>0.811642</td>\n",
              "      <td>CCMSLIB00005761025</td>\n",
              "      <td>Massbank:LU030804 Lauryldiethanolamine|2-[dode...</td>\n",
              "      <td>CCCCCCCCCCCCN(CCO)CCO</td>\n",
              "      <td>1S/C16H35NO2/c1-2-3-4-5-6-7-8-9-10-11-12-17(13...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>74</th>\n",
              "      <td>0.761311</td>\n",
              "      <td>CCMSLIB00005761025</td>\n",
              "      <td>Massbank:LU030804 Lauryldiethanolamine|2-[dode...</td>\n",
              "      <td>CCCCCCCCCCCCN(CCO)CCO</td>\n",
              "      <td>1S/C16H35NO2/c1-2-3-4-5-6-7-8-9-10-11-12-17(13...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>78</th>\n",
              "      <td>0.706896</td>\n",
              "      <td>CCMSLIB00005764095</td>\n",
              "      <td>Massbank:LU030803 Lauryldiethanolamine|2-[dode...</td>\n",
              "      <td>CCCCCCCCCCCCN(CCO)CCO</td>\n",
              "      <td>1S/C16H35NO2/c1-2-3-4-5-6-7-8-9-10-11-12-17(13...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-d53447bb-384e-4f6e-b49d-5ba0705f4dcd')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-d53447bb-384e-4f6e-b49d-5ba0705f4dcd button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-d53447bb-384e-4f6e-b49d-5ba0705f4dcd');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-d8ca7360-1e03-4b4a-a2d7-b7f26f3fa156\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-d8ca7360-1e03-4b4a-a2d7-b7f26f3fa156')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-d8ca7360-1e03-4b4a-a2d7-b7f26f3fa156 button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "fbmn",
              "summary": "{\n  \"name\": \"fbmn\",\n  \"rows\": 52496,\n  \"fields\": [\n    {\n      \"column\": \"id\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 16416,\n        \"min\": 48,\n        \"max\": 58035,\n        \"num_unique_values\": 52496,\n        \"samples\": [\n          22741,\n          30387,\n          41024\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"mqscore\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.12045737518630009,\n        \"min\": 0.600166,\n        \"max\": 0.999783,\n        \"num_unique_values\": 6174,\n        \"samples\": [\n          0.969604,\n          0.916387,\n          0.6851659999999999\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"spectrum\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 3857,\n        \"samples\": [\n          \"CCMSLIB00000486712\",\n          \"CCMSLIB00003418894\",\n          \"CCMSLIB00003213500\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"library\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 2539,\n        \"samples\": [\n          \"Taurocholic acid\",\n          \"Leu Lys Val\",\n          \"Atractylenolide III\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"smiles\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 1686,\n        \"samples\": [\n          \"CCC=CCC(C=CC1C2CC(C1CC=CCCCC(=O)O)CO2)O\",\n          \"CC(C(=O)NC(CC1=CNC2=CC=CC=C21)C(=O)O)N\",\n          \"CCCCCCCCC=CCCCCCCCC(=O)NCCS(=O)(=O)O\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"inchi\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 1797,\n        \"samples\": [\n          \"InChI=1S/C6H6O4/c7-2-4-1-5(8)6(9)3-10-4/h1,3,7,9H,2H2\",\n          \"InChI=1S/C21H40O4/c1-2-3-4-5-6-7-8-9-10-11-12-13-14-15-16-17-21(24)25-19-20(23)18-22/h9-10,20,22-23H,2-8,11-19H2,1H3/b10-9-\",\n          \"InChI=1S/C7H9N/c1-6-3-7(2)5-8-4-6/h3-5H,1-2H3\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 68
        }
      ],
      "source": [
        "fbmn = pd.read_table(f'{HOME}/masspec/fbmn.tsv', index_col=0)\n",
        "fbmn.head()"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "f6626c53-597b-4b86-b510-c6df8567f352",
      "metadata": {
        "id": "f6626c53-597b-4b86-b510-c6df8567f352"
      },
      "source": [
        "Among a total of 52,496 spectra, only 6,513 (12%) were annotated. This implicates the challenges of analyzing mass spectrum data."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 69,
      "id": "d5723bf6-cfa9-4b01-8841-16cf6f706b43",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "d5723bf6-cfa9-4b01-8841-16cf6f706b43",
        "outputId": "9c3aabcb-6cea-42c7-c0f6-5d92a1976f99"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "52496"
            ]
          },
          "metadata": {},
          "execution_count": 69
        }
      ],
      "source": [
        "fbmn.shape[0]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 70,
      "id": "687ad9db-dcae-4f72-a23f-509d46175d34",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "687ad9db-dcae-4f72-a23f-509d46175d34",
        "outputId": "7c55264c-8c33-4475-dc5d-a4a458d989d7"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "6513"
            ]
          },
          "metadata": {},
          "execution_count": 70
        }
      ],
      "source": [
        "fbmn.dropna(how='all').shape[0]"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "9de45287-db4c-4f05-8bc5-cf936edb10ed",
      "metadata": {
        "id": "9de45287-db4c-4f05-8bc5-cf936edb10ed"
      },
      "source": [
        "In parallel, the data were analyzed using **CANOPUS**, a deep neural network framework for classifying molecules ([Dührkop et al., 2021](https://www.nature.com/articles/s41587-020-0740-8)). The classification follows the hierarchy of formula - class - superclass - pathway (just like taxonomy)."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 71,
      "id": "f5a8887a-08cb-42d2-8e6e-3535ec964fd2",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 238
        },
        "id": "f5a8887a-08cb-42d2-8e6e-3535ec964fd2",
        "outputId": "db8ee206-ce81-42e5-9de5-d0d17b79730f"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "      formula                       class                  superclass  \\\n",
              "id                                                                      \n",
              "48    C15H22O                         NaN            Sesquiterpenoids   \n",
              "66    C15H22O                         NaN            Sesquiterpenoids   \n",
              "72  C16H35NO2              Fatty alcohols                 Fatty acyls   \n",
              "74  C16H35NO2  Straight chain fatty acids  Fatty Acids and Conjugates   \n",
              "78  C16H35NO2  Straight chain fatty acids  Fatty Acids and Conjugates   \n",
              "\n",
              "        pathway  \n",
              "id               \n",
              "48   Terpenoids  \n",
              "66   Terpenoids  \n",
              "72  Fatty acids  \n",
              "74  Fatty acids  \n",
              "78  Fatty acids  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-da68e96c-d3a0-4b98-a3f1-ca339413f4eb\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>formula</th>\n",
              "      <th>class</th>\n",
              "      <th>superclass</th>\n",
              "      <th>pathway</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>id</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>48</th>\n",
              "      <td>C15H22O</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Sesquiterpenoids</td>\n",
              "      <td>Terpenoids</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>66</th>\n",
              "      <td>C15H22O</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Sesquiterpenoids</td>\n",
              "      <td>Terpenoids</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>72</th>\n",
              "      <td>C16H35NO2</td>\n",
              "      <td>Fatty alcohols</td>\n",
              "      <td>Fatty acyls</td>\n",
              "      <td>Fatty acids</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>74</th>\n",
              "      <td>C16H35NO2</td>\n",
              "      <td>Straight chain fatty acids</td>\n",
              "      <td>Fatty Acids and Conjugates</td>\n",
              "      <td>Fatty acids</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>78</th>\n",
              "      <td>C16H35NO2</td>\n",
              "      <td>Straight chain fatty acids</td>\n",
              "      <td>Fatty Acids and Conjugates</td>\n",
              "      <td>Fatty acids</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-da68e96c-d3a0-4b98-a3f1-ca339413f4eb')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-da68e96c-d3a0-4b98-a3f1-ca339413f4eb button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-da68e96c-d3a0-4b98-a3f1-ca339413f4eb');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-50b6e909-6305-4242-88e4-25156ef66f53\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-50b6e909-6305-4242-88e4-25156ef66f53')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-50b6e909-6305-4242-88e4-25156ef66f53 button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "canopus",
              "summary": "{\n  \"name\": \"canopus\",\n  \"rows\": 52496,\n  \"fields\": [\n    {\n      \"column\": \"id\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 16416,\n        \"min\": 48,\n        \"max\": 58035,\n        \"num_unique_values\": 52496,\n        \"samples\": [\n          22741,\n          30387,\n          41024\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"formula\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 14921,\n        \"samples\": [\n          \"C19H34O6\",\n          \"C33H41N3O5\",\n          \"C28H42O11\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"class\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 270,\n        \"samples\": [\n          \"Simple indole alkaloids\",\n          \"Lanostane, Tirucallane and Euphane triterpenoids\",\n          \"Cyclophytane diterpenoids\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"superclass\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 60,\n        \"samples\": [\n          \"Sesquiterpenoids\",\n          \"Fatty amides\",\n          \"Lysine alkaloids\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"pathway\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 7,\n        \"samples\": [\n          \"Terpenoids\",\n          \"Fatty acids\",\n          \"Alkaloids\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 71
        }
      ],
      "source": [
        "canopus = pd.read_table(f'{HOME}/masspec/canopus.tsv', index_col=0)\n",
        "canopus.head()"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "c37d5957-7632-4091-b6bd-226c2839b445",
      "metadata": {
        "id": "c37d5957-7632-4091-b6bd-226c2839b445"
      },
      "source": [
        "A much larges number (43,072, or 82%) of spectra were classified."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 72,
      "id": "1e9df7ca-bad8-4f74-ade7-c1cc4841c440",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "1e9df7ca-bad8-4f74-ade7-c1cc4841c440",
        "outputId": "5e8d7d9e-0e2a-4fe4-b328-616e8703a349"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "43072"
            ]
          },
          "metadata": {},
          "execution_count": 72
        }
      ],
      "source": [
        "canopus.dropna(how='all').shape[0]"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "9480d7ba-562a-4392-815a-bdf052416d31",
      "metadata": {
        "id": "9480d7ba-562a-4392-815a-bdf052416d31"
      },
      "source": [
        "Among the 52.5k spectra recovered from the samples, only a small proportion can be related to microbes. The EMP500 team ensembled annotation results by multiple workflows, and obtained a consensus list of microbially related secondary metabolites. The results (0 - no, 1 - yes) are provided in `is.microbial`. This file indicates that 6588 (12.5%) spectra are microbial."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 73,
      "id": "6694d54c-430a-4b67-8788-f2be0a2c6ac0",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "6694d54c-430a-4b67-8788-f2be0a2c6ac0",
        "outputId": "83984680-acd3-428c-9770-40bb0d688aa7"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "48\t0\n",
            "66\t0\n",
            "72\t0\n",
            "74\t0\n",
            "78\t0\n"
          ]
        }
      ],
      "source": [
        "!head -n5 $HOME/masspec/is.microbial"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 74,
      "id": "be3d6037-763e-4a8b-a940-f8e3f41e4034",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 178
        },
        "id": "be3d6037-763e-4a8b-a940-f8e3f41e4034",
        "outputId": "42f64558-3369-4ea6-b06b-3811d0bfd8ea"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "Microbial\n",
              "False    45908\n",
              "True      6588\n",
              "Name: count, dtype: int64"
            ],
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>count</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Microbial</th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>False</th>\n",
              "      <td>45908</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>True</th>\n",
              "      <td>6588</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div><br><label><b>dtype:</b> int64</label>"
            ]
          },
          "metadata": {},
          "execution_count": 74
        }
      ],
      "source": [
        "pd.read_table(f'{HOME}/masspec/is.microbial', index_col=0, names=['Microbial']).squeeze().astype(bool).value_counts()"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "10e53dc5-ada8-4746-851b-145d9a47e33e",
      "metadata": {
        "id": "10e53dc5-ada8-4746-851b-145d9a47e33e"
      },
      "source": [
        "## Summary"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "d4e7f75c-d713-4325-90e8-bc265d4f6bae",
      "metadata": {
        "id": "d4e7f75c-d713-4325-90e8-bc265d4f6bae"
      },
      "source": [
        "This section navigated the components of EMP500, a multi-omic dataset: sample metadata, data tables, and feature metadata (taxonomy, phylogeny, functional ontology, etc.). It discussed the various omic data types included in this demo dataset and beyond. It demonstrated how to use scikit-bio and other Python libraries to parse and explore these files. We will perform analysis of these data using scikit-bio in the following sections."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 74,
      "id": "4ea53712-9a29-4574-be46-ef990f647a9d",
      "metadata": {
        "id": "4ea53712-9a29-4574-be46-ef990f647a9d"
      },
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python [conda env:skbio-070]",
      "language": "python",
      "name": "conda-env-skbio-070-py"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.13.5"
    },
    "colab": {
      "provenance": [],
      "include_colab_link": true
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}